<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pyGenClean.Misc.compare_gold_standard &mdash; pyGenClean 1.8.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.8.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.8.2 documentation" href="../../../index.html" />
    <link rel="up" title="pyGenClean" href="../../pyGenClean.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="../../../index.html"><img src="../../../_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" accesskey="U">pyGenClean</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for pyGenClean.Misc.compare_gold_standard</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python2.7</span>

<span class="c1"># This file is part of pyGenClean.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is free software: you can redistribute it and/or modify it under</span>
<span class="c1"># the terms of the GNU General Public License as published by the Free Software</span>
<span class="c1"># Foundation, either version 3 of the License, or (at your option) any later</span>
<span class="c1"># version.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is distributed in the hope that it will be useful, but WITHOUT ANY</span>
<span class="c1"># WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR</span>
<span class="c1"># A PARTICULAR PURPOSE.  See the GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1"># You should have received a copy of the GNU General Public License along with</span>
<span class="c1"># pyGenClean.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>


<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">__version__</span>
<span class="kn">from</span> <span class="nn">..PlinkUtils</span> <span class="kn">import</span> <span class="n">createRowFromPlinkSpacedOutput</span>
<span class="kn">from</span> <span class="nn">..DupSNPs.duplicated_snps</span> <span class="kn">import</span> <span class="n">flipGenotype</span>


<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;compare_gold_standard&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="c1"># Getting and checking the options</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>
    <span class="n">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Options used:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  --{} {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">),</span> <span class="n">value</span><span class="p">))</span>

    <span class="c1"># Reading the same sample file</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading the same samples file&quot;</span><span class="p">)</span>
    <span class="n">same_samples</span> <span class="o">=</span> <span class="n">read_same_samples_file</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">same_samples</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
    <span class="n">required_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">same_samples</span><span class="p">]</span> <span class="o">+</span>
                           <span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">same_samples</span><span class="p">])</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  - Found a total of {} unique &quot;</span>
                <span class="s2">&quot;samples&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">required_samples</span><span class="p">)))</span>

    <span class="c1"># Check the fam files</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Checking in fam files for required samples&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">check_fam_for_samples</span><span class="p">(</span><span class="n">required_samples</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">bfile</span> <span class="o">+</span> <span class="s2">&quot;.fam&quot;</span><span class="p">,</span>
                                 <span class="n">args</span><span class="o">.</span><span class="n">gold_bfile</span> <span class="o">+</span> <span class="s2">&quot;.fam&quot;</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  - Didn&#39;t find all required samples... STOP&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Find overlap markers with the gold standard file</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finding overlapping SNPs with gold standard&quot;</span><span class="p">)</span>
    <span class="n">findOverlappingSNPsWithGoldStandard</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bfile</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">gold_bfile</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
                                        <span class="n">args</span><span class="o">.</span><span class="n">use_marker_names</span><span class="p">)</span>

    <span class="c1"># Extract the required SNPs using Plink</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Extracting overlapping SNPs from the gold standard and the &quot;</span>
                <span class="s2">&quot;source panel&quot;</span><span class="p">)</span>
    <span class="n">extractSNPs</span><span class="p">([</span><span class="n">args</span><span class="o">.</span><span class="n">gold_bfile</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">bfile</span><span class="p">],</span>
                <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_snp_to_extract&quot;</span><span class="p">,</span>
                 <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_snp_to_extract&quot;</span><span class="p">],</span>
                <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_standard&quot;</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel&quot;</span><span class="p">],</span>
                <span class="n">args</span><span class="o">.</span><span class="n">sge</span><span class="p">)</span>

    <span class="n">gold_input_prefix</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_standard&quot;</span>
    <span class="c1"># Renaiming the reference file, so that the SNP names are the same</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">use_marker_names</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Renaming gold standard&#39;s SNPs to match source panel&quot;</span><span class="p">)</span>
        <span class="n">renameSNPs</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_standard&quot;</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.update_names&quot;</span><span class="p">,</span>
                   <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_standard.rename&quot;</span><span class="p">)</span>
        <span class="n">gold_input_prefix</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_standard.rename&quot;</span>

    <span class="c1"># Finding the SNPs to flip and flip them in the gold standard</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">do_not_flip</span><span class="p">:</span>
        <span class="c1"># Computing the frequency</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing gold standard frequencies&quot;</span><span class="p">)</span>
        <span class="n">computeFrequency</span><span class="p">(</span><span class="n">gold_input_prefix</span><span class="p">,</span>
                         <span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.frequency&quot;</span><span class="p">)</span>

        <span class="n">source_alleles</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">source_manifest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading the source manifest&quot;</span><span class="p">)</span>
            <span class="n">source_alleles</span> <span class="o">=</span> <span class="n">read_source_manifest</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_manifest</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading the allele file&quot;</span><span class="p">)</span>
            <span class="n">source_alleles</span> <span class="o">=</span> <span class="n">read_source_alleles</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_alleles</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finding SNPs to flip or to exclude from gold standard&quot;</span><span class="p">)</span>
        <span class="n">findFlippedSNPs</span><span class="p">(</span><span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.frequency.frq&quot;</span><span class="p">,</span> <span class="n">source_alleles</span><span class="p">,</span>
                        <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

        <span class="c1"># Excluding SNPs</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Excluding SNPs and samples from the gold standard&quot;</span><span class="p">)</span>
        <span class="n">exclude_SNPs_samples</span><span class="p">(</span><span class="n">gold_input_prefix</span><span class="p">,</span>
                             <span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.cleaned&quot;</span><span class="p">,</span>
                             <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.snp_to_remove&quot;</span><span class="p">,</span>
                             <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_samples2keep&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Excluding SNPs and samples from source panel&quot;</span><span class="p">)</span>
        <span class="n">exclude_SNPs_samples</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel&quot;</span><span class="p">,</span>
                             <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel.cleaned&quot;</span><span class="p">,</span>
                             <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.snp_to_remove&quot;</span><span class="p">,</span>
                             <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel_samples2keep&quot;</span><span class="p">)</span>

        <span class="c1"># Flipping the SNP that need to be flip in the gold standard</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Flipping SNPs in gold standard&quot;</span><span class="p">)</span>
        <span class="n">flipSNPs</span><span class="p">(</span><span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.cleaned&quot;</span><span class="p">,</span>
                 <span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.cleaned.flipped&quot;</span><span class="p">,</span>
                 <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.snp_to_flip_in_reference&quot;</span><span class="p">)</span>

        <span class="c1"># Splitting the files, and running the duplicated samples script</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing statistics&quot;</span><span class="p">)</span>
        <span class="n">compute_statistics</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.duplicated_samples&quot;</span><span class="p">,</span>
                           <span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.cleaned.flipped&quot;</span><span class="p">,</span>
                           <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel.cleaned&quot;</span><span class="p">,</span> <span class="n">same_samples</span><span class="p">,</span>
                           <span class="n">args</span><span class="o">.</span><span class="n">sge</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># We do not need to flip...</span>
        <span class="c1"># Splitting the files, and running the duplicated samples script</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Keeping samples from the gold standard&quot;</span><span class="p">)</span>
        <span class="n">exclude_SNPs_samples</span><span class="p">(</span><span class="n">gold_input_prefix</span><span class="p">,</span>
                             <span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.cleaned&quot;</span><span class="p">,</span>
                             <span class="n">keepSample</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.gold_samples2keep&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Keeping samples from source panel&quot;</span><span class="p">)</span>
        <span class="n">exclude_SNPs_samples</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel&quot;</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel.cleaned&quot;</span><span class="p">,</span>
            <span class="n">keepSample</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel_samples2keep&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing statistics&quot;</span><span class="p">)</span>
        <span class="n">compute_statistics</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.duplicated_samples&quot;</span><span class="p">,</span>
                           <span class="n">gold_input_prefix</span> <span class="o">+</span> <span class="s2">&quot;.cleaned&quot;</span><span class="p">,</span>
                           <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.source_panel.cleaned&quot;</span><span class="p">,</span> <span class="n">same_samples</span><span class="p">,</span>
                           <span class="n">args</span><span class="o">.</span><span class="n">sge</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span></div>


<div class="viewcode-block" id="read_source_manifest"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.read_source_manifest">[docs]</a><span class="k">def</span> <span class="nf">read_source_manifest</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads Illumina manifest.&quot;&quot;&quot;</span>
    <span class="n">alleles</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">open_function</span> <span class="o">=</span> <span class="nb">open</span>
    <span class="k">if</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">):</span>
        <span class="n">open_function</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span>
    <span class="k">with</span> <span class="n">open_function</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="n">header_index</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span>
            <span class="p">(</span><span class="n">col_name</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">col_name</span> <span class="ow">in</span>
            <span class="nb">enumerate</span><span class="p">(</span><span class="n">input_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>
        <span class="p">])</span>
        <span class="k">for</span> <span class="n">col_name</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="s2">&quot;IlmnStrand&quot;</span><span class="p">}:</span>
            <span class="k">if</span> <span class="n">col_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header_index</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: no column named {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">col_name</span><span class="p">)</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">marker_name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">header_index</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">]]</span>
            <span class="n">the_alleles</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">header_index</span><span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">]]</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">header_index</span><span class="p">[</span><span class="s2">&quot;IlmnStrand&quot;</span><span class="p">]]</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">the_alleles</span> <span class="o">==</span> <span class="s2">&quot;[D/I]&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">the_alleles</span> <span class="o">==</span> <span class="s2">&quot;[I/D]&quot;</span><span class="p">):</span>
                <span class="c1"># This is an indel, so we skip it</span>
                <span class="k">continue</span>
            <span class="n">top_alleles</span> <span class="o">=</span> <span class="n">illumina_to_snp</span><span class="p">(</span><span class="n">strand</span><span class="p">,</span> <span class="n">the_alleles</span><span class="p">)</span>
            <span class="n">allele1</span><span class="p">,</span> <span class="n">allele2</span> <span class="o">=</span> <span class="n">top_alleles</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
            <span class="n">alleles</span><span class="p">[</span><span class="n">marker_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">allele1</span><span class="p">,</span> <span class="n">allele2</span><span class="p">}</span>

    <span class="k">return</span> <span class="n">alleles</span></div>


<div class="viewcode-block" id="read_source_alleles"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.read_source_alleles">[docs]</a><span class="k">def</span> <span class="nf">read_source_alleles</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads an allele file.&quot;&quot;&quot;</span>
    <span class="n">alleles</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">open_function</span> <span class="o">=</span> <span class="nb">open</span>
    <span class="k">if</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">):</span>
        <span class="n">open_function</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span>
    <span class="k">with</span> <span class="n">open_function</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">marker_name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">allele1</span><span class="p">,</span> <span class="n">allele2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
            <span class="n">alleles</span><span class="p">[</span><span class="n">marker_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">allele1</span><span class="p">,</span> <span class="n">allele2</span><span class="p">}</span>

    <span class="k">return</span> <span class="n">alleles</span></div>


<div class="viewcode-block" id="illumina_to_snp"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.illumina_to_snp">[docs]</a><span class="k">def</span> <span class="nf">illumina_to_snp</span><span class="p">(</span><span class="n">strand</span><span class="p">,</span> <span class="n">snp</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the TOP strand of the marker.</span>

<span class="sd">    Function that takes a strand (TOP or BOT) and a SNP (e.g. : [A/C]) and</span>
<span class="sd">    returns a space separated AlleleA[space]AlleleB string.</span>

<span class="sd">    :param strand: Either &quot;TOP&quot; or &quot;BOT&quot;</span>
<span class="sd">    :type strand: str</span>

<span class="sd">    :param snp: [A/C], [A/T], [G/C], [T/C], [A/G], [C/G], [T/A] or [T/G].</span>
<span class="sd">    :type snp: str</span>

<span class="sd">    :returns: The nucleotide for allele A and the nucleotide for allele B</span>
<span class="sd">              (space separated)</span>
<span class="sd">    :rtype: str</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># This correspondence matrix converts the alleles</span>
    <span class="c1"># from the annotation file to the appropriate</span>
    <span class="c1"># alleles for TOP strand.</span>
    <span class="c1"># This dictates most of the behavior for this script.</span>
    <span class="c1"># Modify it to customize allele conversion.</span>
    <span class="n">illumina_correspondance</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;[A/C]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;TOP&quot;</span><span class="p">:</span> <span class="s2">&quot;A C&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;[A/G]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;TOP&quot;</span><span class="p">:</span> <span class="s2">&quot;A G&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;[T/C]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;BOT&quot;</span><span class="p">:</span> <span class="s2">&quot;A G&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;[T/G]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;BOT&quot;</span><span class="p">:</span> <span class="s2">&quot;A C&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;[A/T]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;TOP&quot;</span><span class="p">:</span> <span class="s2">&quot;A T&quot;</span><span class="p">,</span>
            <span class="s2">&quot;BOT&quot;</span><span class="p">:</span> <span class="s2">&quot;T A&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;[C/G]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;TOP&quot;</span><span class="p">:</span> <span class="s2">&quot;C G&quot;</span><span class="p">,</span>
            <span class="s2">&quot;BOT&quot;</span><span class="p">:</span> <span class="s2">&quot;G C&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;[G/C]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;TOP&quot;</span><span class="p">:</span> <span class="s2">&quot;G C&quot;</span><span class="p">,</span>
            <span class="s2">&quot;BOT&quot;</span><span class="p">:</span> <span class="s2">&quot;C G&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;[T/A]&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;TOP&quot;</span><span class="p">:</span> <span class="s2">&quot;T A&quot;</span><span class="p">,</span>
            <span class="s2">&quot;BOT&quot;</span><span class="p">:</span> <span class="s2">&quot;A T&quot;</span><span class="p">,</span>
        <span class="p">},</span>
    <span class="p">}</span>

    <span class="k">return</span> <span class="n">illumina_correspondance</span><span class="p">[</span><span class="n">snp</span><span class="p">][</span><span class="n">strand</span><span class="p">]</span></div>


<div class="viewcode-block" id="compute_statistics"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.compute_statistics">[docs]</a><span class="k">def</span> <span class="nf">compute_statistics</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">gold_prefix</span><span class="p">,</span> <span class="n">source_prefix</span><span class="p">,</span> <span class="n">same_samples</span><span class="p">,</span>
                       <span class="n">use_sge</span><span class="p">,</span> <span class="n">final_out_prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Compute the statistics.&quot;&quot;&quot;</span>
    <span class="c1"># Now, creating a temporary directory</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">out_dir</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: file exists&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># The out prefix</span>
    <span class="n">out_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="s2">&quot;tmp&quot;</span><span class="p">)</span>

    <span class="c1"># Subsetting the files</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  - Subsetting the files&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="p">(</span><span class="n">gold_sample</span><span class="p">,</span> <span class="n">source_sample</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">same_samples</span><span class="p">):</span>
        <span class="c1"># Preparing the files</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_gold.samples&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output_file</span><span class="p">:</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gold_sample</span><span class="p">)</span>

            <span class="n">filename</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_source.samples&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output_file</span><span class="p">:</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">source_sample</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;can&#39;t write in dir {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Preparing the files</span>
    <span class="n">nb</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">same_samples</span><span class="p">)</span>
    <span class="n">keepSamples</span><span class="p">(</span>
        <span class="p">[</span><span class="n">gold_prefix</span><span class="p">]</span><span class="o">*</span><span class="n">nb</span> <span class="o">+</span> <span class="p">[</span><span class="n">source_prefix</span><span class="p">]</span><span class="o">*</span><span class="n">nb</span><span class="p">,</span>
        <span class="p">[</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_gold.samples&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nb</span><span class="p">)]</span> <span class="o">+</span>
        <span class="p">[</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_source.samples&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nb</span><span class="p">)],</span>
        <span class="p">[</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_gold&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nb</span><span class="p">)]</span> <span class="o">+</span>
        <span class="p">[</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_source&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nb</span><span class="p">)],</span>
        <span class="n">use_sge</span><span class="p">,</span>
        <span class="n">transpose</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Creating reports</span>
    <span class="c1"># The diff file</span>
    <span class="n">diff_file</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">diff_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">final_out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.diff&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: can&#39;t write file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">final_out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.diff&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="c1"># The diff file header</span>
    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">diff_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;famID_source&quot;</span><span class="p">,</span> <span class="s2">&quot;indID_source&quot;</span><span class="p">,</span>
                                  <span class="s2">&quot;famID_gold&quot;</span><span class="p">,</span> <span class="s2">&quot;indID_gold&quot;</span><span class="p">,</span> <span class="s2">&quot;geno_source&quot;</span><span class="p">,</span>
                                  <span class="s2">&quot;geno_gold&quot;</span><span class="p">])</span>

    <span class="c1"># The concordance file</span>
    <span class="n">concordance_file</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">concordance_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">final_out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.concordance&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: can&#39;t write file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">final_out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.concordance&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="c1"># The concordance file header</span>
    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">concordance_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;famID_source&quot;</span><span class="p">,</span> <span class="s2">&quot;indID_source&quot;</span><span class="p">,</span>
                                         <span class="s2">&quot;famID_gold&quot;</span><span class="p">,</span> <span class="s2">&quot;indID_gold&quot;</span><span class="p">,</span>
                                         <span class="s2">&quot;nb_geno&quot;</span><span class="p">,</span> <span class="s2">&quot;nb_diff&quot;</span><span class="p">,</span>
                                         <span class="s2">&quot;concordance&quot;</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nb</span><span class="p">):</span>
        <span class="c1"># The samples</span>
        <span class="n">gold_sample</span><span class="p">,</span> <span class="n">source_sample</span> <span class="o">=</span> <span class="n">same_samples</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>

        <span class="c1"># Reading the source freq file</span>
        <span class="n">source_genotypes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_source.tped&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">marker_name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="n">geno</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">geno</span> <span class="o">!=</span> <span class="s2">&quot;0 0&quot;</span><span class="p">:</span>
                        <span class="n">geno</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">))</span>
                        <span class="n">source_genotypes</span><span class="p">[</span><span class="n">marker_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">geno</span>
        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="c1"># Reading the gold freq file</span>
        <span class="n">gold_genotypes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;_{}_gold.tped&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">marker_name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="n">geno</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">geno</span> <span class="o">!=</span> <span class="s2">&quot;0 0&quot;</span><span class="p">:</span>
                        <span class="n">geno</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">))</span>
                        <span class="n">gold_genotypes</span><span class="p">[</span><span class="n">marker_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">geno</span>
        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="c1"># The genotyped markers in both, with their number</span>
        <span class="n">genotyped_markers</span> <span class="o">=</span> <span class="n">source_genotypes</span><span class="o">.</span><span class="n">viewkeys</span><span class="p">()</span> <span class="o">&amp;</span> \
            <span class="n">gold_genotypes</span><span class="o">.</span><span class="n">viewkeys</span><span class="p">()</span>
        <span class="n">nb_genotyped</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">genotyped_markers</span><span class="p">)</span>

        <span class="c1"># Finding the number of differences, and creating the diff file</span>
        <span class="n">nb_diff</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">marker_name</span> <span class="ow">in</span> <span class="n">genotyped_markers</span><span class="p">:</span>
            <span class="c1"># Getting the genotypes</span>
            <span class="n">source_geno</span> <span class="o">=</span> <span class="n">source_genotypes</span><span class="p">[</span><span class="n">marker_name</span><span class="p">]</span>
            <span class="n">gold_geno</span> <span class="o">=</span> <span class="n">gold_genotypes</span><span class="p">[</span><span class="n">marker_name</span><span class="p">]</span>

            <span class="c1"># Comparing the genotypes</span>
            <span class="k">if</span> <span class="n">source_geno</span> <span class="o">!=</span> <span class="n">gold_geno</span><span class="p">:</span>
                <span class="n">nb_diff</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">source_geno_print</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">source_geno</span> <span class="o">-</span> <span class="p">{</span><span class="s2">&quot;0&quot;</span><span class="p">})</span>
                <span class="n">source_geno_print</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">source_geno_print</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">source_geno_print</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">source_geno_print</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">gold_geno_print</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gold_geno</span> <span class="o">-</span> <span class="p">{</span><span class="s2">&quot;0&quot;</span><span class="p">})</span>
                <span class="n">gold_geno_print</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gold_geno_print</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">gold_geno_print</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gold_geno_print</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="c1"># We print in the diff file</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">diff_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">marker_name</span><span class="p">,</span>
                                              <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">source_sample</span><span class="p">),</span>
                                              <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gold_sample</span><span class="p">),</span>
                                              <span class="s2">&quot;/&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">source_geno_print</span><span class="p">),</span>
                                              <span class="s2">&quot;/&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gold_geno_print</span><span class="p">)])</span>

        <span class="c1"># Creating the concordance file</span>
        <span class="n">concordance</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">if</span> <span class="n">nb_genotyped</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">concordance</span> <span class="o">=</span> <span class="p">(</span><span class="n">nb_genotyped</span> <span class="o">-</span> <span class="n">nb_diff</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">nb_genotyped</span><span class="p">)</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">concordance_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">source_sample</span><span class="p">),</span>
                                             <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gold_sample</span><span class="p">),</span>
                                             <span class="nb">str</span><span class="p">(</span><span class="n">nb_genotyped</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">nb_diff</span><span class="p">),</span>
                                             <span class="nb">str</span><span class="p">(</span><span class="n">concordance</span><span class="p">)])</span>

    <span class="c1"># Closing the output files</span>
    <span class="n">diff_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">concordance_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Deleating the temporary directory</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">,</span> <span class="s2">&quot;   - Can&#39;t delete {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span></div>


<div class="viewcode-block" id="check_fam_for_samples"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.check_fam_for_samples">[docs]</a><span class="k">def</span> <span class="nf">check_fam_for_samples</span><span class="p">(</span><span class="n">required_samples</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="n">gold</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check fam files for required_samples.&quot;&quot;&quot;</span>
    <span class="c1"># Checking the source panel</span>
    <span class="n">source_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[:</span><span class="mi">2</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">required_samples</span><span class="p">:</span>
                <span class="n">source_samples</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

    <span class="c1"># Checking the gold standard</span>
    <span class="n">gold_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gold</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[:</span><span class="mi">2</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">required_samples</span><span class="p">:</span>
                <span class="n">gold_samples</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

    <span class="c1"># Checking if we got everything</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  - Found {} samples in source panel&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">source_samples</span><span class="p">),</span>
    <span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  - Found {} samples in gold standard&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">gold_samples</span><span class="p">),</span>
    <span class="p">))</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">required_samples</span> <span class="o">-</span> <span class="p">(</span><span class="n">source_samples</span> <span class="o">|</span> <span class="n">gold_samples</span><span class="p">))</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">False</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">True</span></div>


<div class="viewcode-block" id="read_same_samples_file"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.read_same_samples_file">[docs]</a><span class="k">def</span> <span class="nf">read_same_samples_file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads a file containing same samples.&quot;&quot;&quot;</span>
    <span class="c1"># The same samples</span>
    <span class="n">same_samples</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Creating the extraction files</span>
    <span class="n">gold_file</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">gold_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.gold_samples2keep&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: can&#39;t create file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.gold_samples2keep&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">source_file</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">source_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.source_panel_samples2keep&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;{}: can&#39;t create &quot;</span>
               <span class="s2">&quot;file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.source_panel_samples2keep&quot;</span><span class="p">))</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Getting the samples</span>
            <span class="n">gold_sample</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">row</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">source_sample</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">:])</span>
            <span class="n">same_samples</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">gold_sample</span><span class="p">,</span> <span class="n">source_sample</span><span class="p">))</span>

            <span class="c1"># Printing files</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">gold_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gold_sample</span><span class="p">)</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">source_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">source_sample</span><span class="p">)</span>

    <span class="c1"># Closing the files</span>
    <span class="n">gold_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">source_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">same_samples</span></div>


<div class="viewcode-block" id="flipSNPs"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.flipSNPs">[docs]</a><span class="k">def</span> <span class="nf">flipSNPs</span><span class="p">(</span><span class="n">inPrefix</span><span class="p">,</span> <span class="n">outPrefix</span><span class="p">,</span> <span class="n">flipFileName</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Flip SNPs using Plink.&quot;&quot;&quot;</span>
    <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">inPrefix</span><span class="p">,</span> <span class="s2">&quot;--flip&quot;</span><span class="p">,</span>
                    <span class="n">flipFileName</span><span class="p">,</span> <span class="s2">&quot;--make-bed&quot;</span><span class="p">,</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="n">outPrefix</span><span class="p">]</span>
    <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span></div>


<div class="viewcode-block" id="exclude_SNPs_samples"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.exclude_SNPs_samples">[docs]</a><span class="k">def</span> <span class="nf">exclude_SNPs_samples</span><span class="p">(</span><span class="n">inPrefix</span><span class="p">,</span> <span class="n">outPrefix</span><span class="p">,</span> <span class="n">exclusionSNP</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                         <span class="n">keepSample</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">transpose</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Exclude some SNPs and keep some samples using Plink.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">exclusionSNP</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">keepSample</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Something wront with development... work on that source code...&quot;</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">inPrefix</span><span class="p">,</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span>
                    <span class="n">outPrefix</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">exclusionSNP</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="c1"># We want to exclude SNP</span>
        <span class="n">plinkCommand</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;--exclude&quot;</span><span class="p">,</span> <span class="n">exclusionSNP</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">keepSample</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="c1"># We want to keep samples</span>
        <span class="n">plinkCommand</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;--keep&quot;</span><span class="p">,</span> <span class="n">keepSample</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">transpose</span><span class="p">:</span>
        <span class="c1"># We want a transpose ped file</span>
        <span class="n">plinkCommand</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;--recode&quot;</span><span class="p">,</span> <span class="s2">&quot;--transpose&quot;</span><span class="p">,</span> <span class="s2">&quot;--tab&quot;</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># We want a binary file</span>
        <span class="n">plinkCommand</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--make-bed&quot;</span><span class="p">)</span>
    <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span></div>


<div class="viewcode-block" id="renameSNPs"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.renameSNPs">[docs]</a><span class="k">def</span> <span class="nf">renameSNPs</span><span class="p">(</span><span class="n">inPrefix</span><span class="p">,</span> <span class="n">updateFileName</span><span class="p">,</span> <span class="n">outPrefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Updates the name of the SNPs using Plink.&quot;&quot;&quot;</span>
    <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">inPrefix</span><span class="p">,</span> <span class="s2">&quot;--update-map&quot;</span><span class="p">,</span>
                    <span class="n">updateFileName</span><span class="p">,</span> <span class="s2">&quot;--update-name&quot;</span><span class="p">,</span> <span class="s2">&quot;--make-bed&quot;</span><span class="p">,</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span>
                    <span class="n">outPrefix</span><span class="p">]</span>
    <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span></div>


<div class="viewcode-block" id="findFlippedSNPs"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.findFlippedSNPs">[docs]</a><span class="k">def</span> <span class="nf">findFlippedSNPs</span><span class="p">(</span><span class="n">goldFrqFile1</span><span class="p">,</span> <span class="n">sourceAlleles</span><span class="p">,</span> <span class="n">outPrefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Find flipped SNPs and flip them in the data1.&quot;&quot;&quot;</span>
    <span class="n">goldAlleles</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">goldFrqFile1</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
        <span class="n">headerIndex</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">inputFile</span><span class="p">):</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">createRowFromPlinkSpacedOutput</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># This is the header</span>
                <span class="n">headerIndex</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span>
                    <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">))</span>
                <span class="p">])</span>

                <span class="c1"># Checking the columns</span>
                <span class="k">for</span> <span class="n">columnName</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="s2">&quot;A1&quot;</span><span class="p">,</span> <span class="s2">&quot;A2&quot;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">columnName</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">headerIndex</span><span class="p">:</span>
                        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(fileName)s</span><span class="s2">: no column named &quot;</span> \
                                <span class="s2">&quot;</span><span class="si">%(columnName)s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
                        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">snpName</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">]]</span>
                <span class="n">allele1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;A1&quot;</span><span class="p">]]</span>
                <span class="n">allele2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;A2&quot;</span><span class="p">]]</span>

                <span class="n">alleles</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">allele1</span><span class="p">,</span> <span class="n">allele2</span><span class="p">])</span>
                <span class="k">if</span> <span class="s2">&quot;0&quot;</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
                    <span class="n">alleles</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span>

                <span class="n">goldAlleles</span><span class="p">[</span><span class="n">snpName</span><span class="p">]</span> <span class="o">=</span> <span class="n">alleles</span>

    <span class="c1"># Finding the SNPs to flip</span>
    <span class="n">toFlipOutputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">toFlipOutputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outPrefix</span> <span class="o">+</span> <span class="s2">&quot;.snp_to_flip_in_reference&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(outPrefix)s</span><span class="s2">.snp_to_flip_in_reference: can&#39;t write &quot;</span> \
              <span class="s2">&quot;file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">toRemoveOutputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">toRemoveOutputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outPrefix</span> <span class="o">+</span> <span class="s2">&quot;.snp_to_remove&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(outPrefix)s</span><span class="s2">.snp_to_remove: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">toRemoveOutputFileExplanation</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">toRemoveOutputFileExplanation</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span>
            <span class="n">outPrefix</span> <span class="o">+</span> <span class="s2">&quot;.snp_to_remove.explanation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;w&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toRemoveOutputFileExplanation</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;Reason&quot;</span><span class="p">,</span>
                                                          <span class="s2">&quot;Alleles 1&quot;</span><span class="p">,</span>
                                                          <span class="s2">&quot;Alleles 2&quot;</span><span class="p">])</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(outPrefix)s</span><span class="s2">.snp_to_remove: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">snpName</span> <span class="ow">in</span> <span class="n">goldAlleles</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
        <span class="n">alleles1</span> <span class="o">=</span> <span class="n">goldAlleles</span><span class="p">[</span><span class="n">snpName</span><span class="p">]</span>
        <span class="n">alleles2</span> <span class="o">=</span> <span class="n">sourceAlleles</span><span class="p">[</span><span class="n">snpName</span><span class="p">]</span>

        <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">):</span>
            <span class="c1"># Both are heterozygous</span>
            <span class="k">if</span> <span class="p">(({</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">}</span> <span class="o">==</span> <span class="n">alleles1</span> <span class="ow">and</span> <span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">}</span> <span class="o">==</span> <span class="n">alleles2</span><span class="p">)</span> <span class="ow">or</span>
                    <span class="p">({</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">}</span> <span class="o">==</span> <span class="n">alleles1</span> <span class="ow">and</span> <span class="p">{</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">}</span> <span class="o">==</span> <span class="n">alleles2</span><span class="p">)):</span>
                <span class="c1"># We can&#39;t flip those..., so we remove them</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toRemoveOutputFile</span><span class="p">,</span> <span class="n">snpName</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toRemoveOutputFileExplanation</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
                    <span class="n">snpName</span><span class="p">,</span>
                    <span class="s2">&quot;Undetermined&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alleles1</span><span class="p">),</span>
                    <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alleles2</span><span class="p">),</span>
                <span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">alleles1</span> <span class="o">!=</span> <span class="n">alleles2</span><span class="p">:</span>
                    <span class="c1"># Let&#39;s try the flip one</span>
                    <span class="k">if</span> <span class="n">flipGenotype</span><span class="p">(</span><span class="n">alleles1</span><span class="p">)</span> <span class="o">==</span> <span class="n">alleles2</span><span class="p">:</span>
                        <span class="c1"># We need to flip it</span>
                        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toFlipOutputFile</span><span class="p">,</span> <span class="n">snpName</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Those SNP are discordant...</span>
                        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toRemoveOutputFile</span><span class="p">,</span> <span class="n">snpName</span>
                        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toRemoveOutputFileExplanation</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
                            <span class="n">snpName</span><span class="p">,</span>
                            <span class="s2">&quot;Invalid&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alleles1</span><span class="p">),</span>
                            <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alleles2</span><span class="p">),</span>
                        <span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># We want to remove this SNP, because there is at least one</span>
            <span class="c1"># homozygous individual</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toRemoveOutputFile</span><span class="p">,</span> <span class="n">snpName</span>
            <span class="n">tmp_allele1</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alleles1</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">tmp_allele1</span> <span class="o">+=</span> <span class="n">tmp_allele1</span>
            <span class="n">tmp_allele2</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alleles1</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">tmp_allele2</span> <span class="o">+=</span> <span class="n">tmp_allele2</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">toRemoveOutputFileExplanation</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">snpName</span><span class="p">,</span>
                                                              <span class="s2">&quot;Homozygous&quot;</span><span class="p">,</span>
                                                              <span class="n">tmp_allele1</span><span class="p">,</span>
                                                              <span class="n">tmp_allele2</span><span class="p">])</span>

    <span class="c1"># Closing output files</span>
    <span class="n">toFlipOutputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">toRemoveOutputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">toRemoveOutputFileExplanation</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="computeFrequency"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.computeFrequency">[docs]</a><span class="k">def</span> <span class="nf">computeFrequency</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">outPrefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Compute the frequency using Plink.&quot;&quot;&quot;</span>
    <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="s2">&quot;--freq&quot;</span><span class="p">,</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span>
                    <span class="n">outPrefix</span><span class="p">]</span>
    <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span></div>


<div class="viewcode-block" id="findOverlappingSNPsWithGoldStandard"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.findOverlappingSNPsWithGoldStandard">[docs]</a><span class="k">def</span> <span class="nf">findOverlappingSNPsWithGoldStandard</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">gold_prefixe</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">,</span>
                                        <span class="n">use_marker_names</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Find the overlapping SNPs in 4 different data sets.&quot;&quot;&quot;</span>
    <span class="c1"># Reading the main file</span>
    <span class="n">sourceSnpToExtract</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">use_marker_names</span><span class="p">:</span>
        <span class="n">sourceSnpToExtract</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">duplicates</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.bim&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inputFile</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">chromosome</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">position</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
                <span class="n">snpName</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">use_marker_names</span><span class="p">:</span>
                    <span class="n">sourceSnpToExtract</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">snpName</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sourceSnpToExtract</span><span class="p">:</span>
                        <span class="n">sourceSnpToExtract</span><span class="p">[(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">)]</span> <span class="o">=</span> <span class="n">snpName</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># It&#39;s a duplicate</span>
                        <span class="n">duplicates</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">))</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">.bim: no such file&quot;</span> <span class="o">%</span> <span class="n">prefix</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Removing duplicates from the list</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">use_marker_names</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  - There are {} duplicated markers &quot;</span>
                    <span class="s2">&quot;in {};&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">duplicates</span><span class="p">),</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.bim&quot;</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  - removing them for simplicity...&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">snpID</span> <span class="ow">in</span> <span class="n">duplicates</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">sourceSnpToExtract</span><span class="p">[</span><span class="n">snpID</span><span class="p">]</span>

    <span class="c1"># Reading the Gold standard files</span>
    <span class="n">goldSnpToExtract</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">use_marker_names</span><span class="p">:</span>
        <span class="n">goldSnpToExtract</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gold_prefixe</span> <span class="o">+</span> <span class="s2">&quot;.bim&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inputFile</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">chromosome</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">position</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">snpName</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">use_marker_names</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">snpName</span> <span class="ow">in</span> <span class="n">sourceSnpToExtract</span><span class="p">:</span>
                    <span class="n">goldSnpToExtract</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">snpName</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sourceSnpToExtract</span><span class="p">:</span>
                    <span class="c1"># We want this SNP</span>
                    <span class="n">goldSnpToExtract</span><span class="p">[(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">)]</span> <span class="o">=</span> <span class="n">snpName</span>

    <span class="c1"># Printing the names of the SNPs to extract</span>
    <span class="n">goldOutputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">goldOutputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.gold_snp_to_extract&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(out_prefix)s</span><span class="s2">.goldSnpToExtract: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">sourceOutputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">sourceOutputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.source_snp_to_extract&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(out_prefix)s</span><span class="s2">.sourceSnpToExtract: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">changeNameOutputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">changeNameOutputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.update_names&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(out_prefix)s</span><span class="s2">.updateNames: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Writing the file</span>
    <span class="k">if</span> <span class="n">use_marker_names</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">snpName</span> <span class="ow">in</span> <span class="n">goldSnpToExtract</span><span class="p">:</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">sourceOutputFile</span><span class="p">,</span> <span class="n">snpName</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">goldOutputFile</span><span class="p">,</span> <span class="n">snpName</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">snpID</span> <span class="ow">in</span> <span class="n">goldSnpToExtract</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">sourceOutputFile</span><span class="p">,</span> <span class="n">sourceSnpToExtract</span><span class="p">[</span><span class="n">snpID</span><span class="p">]</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">goldOutputFile</span><span class="p">,</span> <span class="n">goldSnpToExtract</span><span class="p">[</span><span class="n">snpID</span><span class="p">]</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">changeNameOutputFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
                <span class="n">goldSnpToExtract</span><span class="p">[</span><span class="n">snpID</span><span class="p">],</span>
                <span class="n">sourceSnpToExtract</span><span class="p">[</span><span class="n">snpID</span><span class="p">],</span>
            <span class="p">])</span>

    <span class="c1"># Closing the output file</span>
    <span class="n">goldOutputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">sourceOutputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">changeNameOutputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="extractSNPs"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.extractSNPs">[docs]</a><span class="k">def</span> <span class="nf">extractSNPs</span><span class="p">(</span><span class="n">prefixes</span><span class="p">,</span> <span class="n">snpToExtractFileNames</span><span class="p">,</span> <span class="n">outPrefixes</span><span class="p">,</span> <span class="n">runSGE</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a list of SNPs using Plink.&quot;&quot;&quot;</span>
    <span class="n">s</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">jobIDs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">jobTemplates</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">runSGE</span><span class="p">:</span>
        <span class="c1"># Add the environment variable for DRMAA package</span>
        <span class="k">if</span> <span class="s2">&quot;DRMAA_LIBRARY_PATH&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">:</span>
            <span class="n">t</span> <span class="o">=</span> <span class="s2">&quot;/shares/data/sge/lib/lx24-amd64/libdrmaa.so.1.0&quot;</span>
            <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;DRMAA_LIBRARY_PATH&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span>

        <span class="c1"># Import the python drmaa library</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="kn">import</span> <span class="nn">drmaa</span>
        <span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s2">&quot;drmaa is not install, install drmaa&quot;</span><span class="p">)</span>

        <span class="c1"># Initializing a session</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">drmaa</span><span class="o">.</span><span class="n">Session</span><span class="p">()</span>
        <span class="n">s</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">prefixes</span><span class="p">):</span>
        <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="s2">&quot;--extract&quot;</span><span class="p">,</span>
                        <span class="n">snpToExtractFileNames</span><span class="p">[</span><span class="n">k</span><span class="p">],</span> <span class="s2">&quot;--make-bed&quot;</span><span class="p">,</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span>
                        <span class="n">outPrefixes</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>

        <span class="k">if</span> <span class="n">runSGE</span><span class="p">:</span>
            <span class="c1"># We run using SGE</span>
            <span class="c1"># Creating the job template</span>
            <span class="n">jt</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">createJobTemplate</span><span class="p">()</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">remoteCommand</span> <span class="o">=</span> <span class="n">plinkCommand</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">workingDirectory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">jobEnvironment</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">args</span> <span class="o">=</span> <span class="n">plinkCommand</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">jobName</span> <span class="o">=</span> <span class="s2">&quot;_plink_extract_snps&quot;</span>

            <span class="c1"># Running the job</span>
            <span class="n">jobID</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">runJob</span><span class="p">(</span><span class="n">jt</span><span class="p">)</span>

            <span class="c1"># Storing the job template and the job ID</span>
            <span class="n">jobTemplates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">jt</span><span class="p">)</span>
            <span class="n">jobIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">jobID</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># We run normal</span>
            <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">runSGE</span><span class="p">:</span>
        <span class="c1"># We wait for all the jobs to be over</span>
        <span class="n">hadProblems</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">jobID</span> <span class="ow">in</span> <span class="n">jobIDs</span><span class="p">:</span>
            <span class="n">retVal</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">wait</span><span class="p">(</span><span class="n">jobID</span><span class="p">,</span> <span class="n">drmaa</span><span class="o">.</span><span class="n">Session</span><span class="o">.</span><span class="n">TIMEOUT_WAIT_FOREVER</span><span class="p">)</span>
            <span class="n">hadProblems</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">retVal</span><span class="o">.</span><span class="n">exitStatus</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>

        <span class="c1"># The jobs should be finished, so we clean everything</span>
        <span class="c1"># Deleating the job template, and exiting the session</span>
        <span class="k">for</span> <span class="n">jt</span> <span class="ow">in</span> <span class="n">jobTemplates</span><span class="p">:</span>
            <span class="n">s</span><span class="o">.</span><span class="n">deleteJobTemplate</span><span class="p">(</span><span class="n">jt</span><span class="p">)</span>

        <span class="c1"># Closing the connection</span>
        <span class="n">s</span><span class="o">.</span><span class="n">exit</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">hadProblem</span> <span class="ow">in</span> <span class="n">hadProblems</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">hadProblem</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Some SGE jobs had errors...&quot;</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="keepSamples"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.keepSamples">[docs]</a><span class="k">def</span> <span class="nf">keepSamples</span><span class="p">(</span><span class="n">prefixes</span><span class="p">,</span> <span class="n">samplesToExtractFileNames</span><span class="p">,</span> <span class="n">outPrefixes</span><span class="p">,</span>
                <span class="n">runSGE</span><span class="p">,</span> <span class="n">transpose</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract a list of SNPs using Plink.&quot;&quot;&quot;</span>
    <span class="n">s</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">jobIDs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">jobTemplates</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">runSGE</span><span class="p">:</span>
        <span class="c1"># Add the environment variable for DRMAA package</span>
        <span class="k">if</span> <span class="s2">&quot;DRMAA_LIBRARY_PATH&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">:</span>
            <span class="n">t</span> <span class="o">=</span> <span class="s2">&quot;/shares/data/sge/lib/lx24-amd64/libdrmaa.so.1.0&quot;</span>
            <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;DRMAA_LIBRARY_PATH&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span>

        <span class="c1"># Import the python drmaa library</span>
        <span class="kn">import</span> <span class="nn">drmaa</span>

        <span class="c1"># Initializing a session</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">drmaa</span><span class="o">.</span><span class="n">Session</span><span class="p">()</span>
        <span class="n">s</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">prefixes</span><span class="p">):</span>
        <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="s2">&quot;--keep&quot;</span><span class="p">,</span>
                        <span class="n">samplesToExtractFileNames</span><span class="p">[</span><span class="n">k</span><span class="p">],</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="n">outPrefixes</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
        <span class="k">if</span> <span class="n">transpose</span><span class="p">:</span>
            <span class="n">plinkCommand</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;--recode&quot;</span><span class="p">,</span> <span class="s2">&quot;--transpose&quot;</span><span class="p">,</span> <span class="s2">&quot;--tab&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plinkCommand</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--make-bed&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">runSGE</span><span class="p">:</span>
            <span class="c1"># We run using SGE</span>
            <span class="c1"># Creating the job template</span>
            <span class="n">jt</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">createJobTemplate</span><span class="p">()</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">remoteCommand</span> <span class="o">=</span> <span class="n">plinkCommand</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">workingDirectory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">jobEnvironment</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">args</span> <span class="o">=</span> <span class="n">plinkCommand</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="n">jt</span><span class="o">.</span><span class="n">jobName</span> <span class="o">=</span> <span class="s2">&quot;_plink_keep_samples&quot;</span>

            <span class="c1"># Running the job</span>
            <span class="n">jobID</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">runJob</span><span class="p">(</span><span class="n">jt</span><span class="p">)</span>

            <span class="c1"># Storing the job template and the job ID</span>
            <span class="n">jobTemplates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">jt</span><span class="p">)</span>
            <span class="n">jobIDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">jobID</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># We run normal</span>
            <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">runSGE</span><span class="p">:</span>
        <span class="c1"># We wait for all the jobs to be over</span>
        <span class="n">hadProblems</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">jobID</span> <span class="ow">in</span> <span class="n">jobIDs</span><span class="p">:</span>
            <span class="n">retVal</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">wait</span><span class="p">(</span><span class="n">jobID</span><span class="p">,</span> <span class="n">drmaa</span><span class="o">.</span><span class="n">Session</span><span class="o">.</span><span class="n">TIMEOUT_WAIT_FOREVER</span><span class="p">)</span>
            <span class="n">hadProblems</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">retVal</span><span class="o">.</span><span class="n">exitStatus</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>

        <span class="c1"># The jobs should be finished, so we clean everything</span>
        <span class="c1"># Deleating the job template, and exiting the session</span>
        <span class="k">for</span> <span class="n">jt</span> <span class="ow">in</span> <span class="n">jobTemplates</span><span class="p">:</span>
            <span class="n">s</span><span class="o">.</span><span class="n">deleteJobTemplate</span><span class="p">(</span><span class="n">jt</span><span class="p">)</span>

        <span class="c1"># Closing the connection</span>
        <span class="n">s</span><span class="o">.</span><span class="n">exit</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">hadProblem</span> <span class="ow">in</span> <span class="n">hadProblems</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">hadProblem</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Some SGE jobs had errors...&quot;</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="runCommand"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.runCommand">[docs]</a><span class="k">def</span> <span class="nf">runCommand</span><span class="p">(</span><span class="n">command</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Run a command.&quot;&quot;&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">command</span><span class="p">,</span>
                                         <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;couldn&#39;t run command</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="checkArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.checkArgs">[docs]</a><span class="k">def</span> <span class="nf">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    :param args: a :py:class:`Namespace` object containing the options of the</span>
<span class="sd">                 program.</span>
<span class="sd">    :type args: :py:class:`argparse.Namespace`</span>

<span class="sd">    :returns: ``True`` if everything was OK.</span>

<span class="sd">    If there is a problem with an option, an exception is raised using the</span>
<span class="sd">    :py:class:`ProgramError` class, a message is printed</span>
<span class="sd">    to the :class:`sys.stderr` and the program exists with code 1.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check if we have the binary files</span>
    <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">bfile</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">gold_bfile</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;no input file&quot;</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">fileName</span> <span class="ow">in</span> <span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;.bim&quot;</span><span class="p">,</span> <span class="s2">&quot;.fam&quot;</span><span class="p">]]:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fileName</span><span class="p">):</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fileName</span><span class="p">)</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Check for the same sample file</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">same_samples</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">same_samples</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Check we have either a manifest or a allele file</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_manifest</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_alleles</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;need an allele file (either an Illumina manifest &quot;</span>
               <span class="s2">&quot;[--source-manifest] or a file containing alleles for each &quot;</span>
               <span class="s2">&quot;marker [--source-alleles]&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">((</span><span class="n">args</span><span class="o">.</span><span class="n">source_manifest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_alleles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">)):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;use either --source-manifest or --source-alleles, not both&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="c1"># Check for the manifests</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">source_manifest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_manifest</span><span class="p">):</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_manifest</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">source_alleles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_alleles</span><span class="p">):</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">source_alleles</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">return</span> <span class="bp">True</span></div>


<div class="viewcode-block" id="parseArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.parseArgs">[docs]</a><span class="k">def</span> <span class="nf">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>  <span class="c1"># pragma: no cover</span>
    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    :returns: A :py:class:`numpy.Namespace` object created by</span>
<span class="sd">              the :py:mod:`argparse` module. It contains the values of the</span>
<span class="sd">              different options.</span>

<span class="sd">    ====================== ======= ============================================</span>
<span class="sd">        Options              Type                   Description</span>
<span class="sd">    ====================== ======= ============================================</span>
<span class="sd">    ``--bfile``            string  The input file prefix (will find the plink</span>
<span class="sd">                                   binary files by appending the prefix to the</span>
<span class="sd">                                   ``.bim``, ``.bed`` and ``.fam`` files,</span>
<span class="sd">                                   respectively).</span>
<span class="sd">    ``--gold-bfile``       string  The input file prefix (will find the plink</span>
<span class="sd">                                   binary files by appending the prefix to the</span>
<span class="sd">                                   ``.bim``, ``.bed`` and ``.fam`` files,</span>
<span class="sd">                                   respectively) for the Gold Standard.</span>
<span class="sd">    ``--same-samples``     string  A file containing samples which are present</span>
<span class="sd">                                   in both the gold standard and the source</span>
<span class="sd">                                   panel. One line by identity and tab</span>
<span class="sd">                                   separated. For each row, first sample is</span>
<span class="sd">                                   Gold Standard, second is source panel.</span>
<span class="sd">    ``--source-manifest``  string  The illumina marker manifest. This file</span>
<span class="sd">                                   should have tabs as field separator. There</span>
<span class="sd">                                   should be no lines before the main header</span>
<span class="sd">                                   line. There should be no lines after the</span>
<span class="sd">                                   last data line.</span>
<span class="sd">    ``--source-alleles``   string  A file containing the source alleles (TOP).</span>
<span class="sd">                                   Two columns (separated by tabulation, one</span>
<span class="sd">                                   with the marker name, the other with the</span>
<span class="sd">                                   alleles (separated by space). No header.</span>
<span class="sd">    ``--sge``              boolean Use SGE for parallelization.</span>
<span class="sd">    ``--do-not-flip``      boolean Do not flip SNPs. WARNING: only use this</span>
<span class="sd">                                   option only if the Gold Standard was</span>
<span class="sd">                                   generated using the same chip (hence,</span>
<span class="sd">                                   flipping is unnecessary).</span>
<span class="sd">    ``--use-marker-names`` boolean Use marker names instead of (chr, position).</span>
<span class="sd">                                   WARNING: only use this options only if the</span>
<span class="sd">                                   Gold Standard was generated using the same</span>
<span class="sd">                                   chip (hence, they have the same marker</span>
<span class="sd">                                   names).</span>
<span class="sd">    ``--out``              string  The prefix of the output files.</span>
<span class="sd">    ====================== ======= ============================================</span>

<span class="sd">    .. note::</span>
<span class="sd">        No option check is done here (except for the one automatically done by</span>
<span class="sd">        argparse). Those need to be done elsewhere (see :py:func:`checkArgs`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">argString</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">args</span></div>


<div class="viewcode-block" id="ProgramError"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.ProgramError">[docs]</a><span class="k">class</span> <span class="nc">ProgramError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;An :py:class:`Exception` raised in case of a problem.</span>

<span class="sd">    :param msg: the message to print to the user before exiting.</span>
<span class="sd">    :type msg: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Construction of the :py:class:`ProgramError` class.</span>

<span class="sd">        :param msg: the message to print to the user</span>
<span class="sd">        :type msg: str</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">message</span></div>


<span class="c1"># The parser object</span>
<span class="n">pretty_name</span> <span class="o">=</span> <span class="s2">&quot;Gold standard comparison&quot;</span>
<span class="n">desc</span> <span class="o">=</span> <span class="s2">&quot;Compares the dataset with a gold standard.&quot;</span>
<span class="n">long_desc</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;The script compares two dataset (the study dataset and a gold &quot;</span>
             <span class="s2">&quot;standard) to find genotyping discordance.&quot;</span><span class="p">)</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-v&quot;</span><span class="p">,</span> <span class="s2">&quot;--version&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;version&quot;</span><span class="p">,</span>
                    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;pyGenClean version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>

<span class="c1"># The INPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Input File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The input file prefix (will find the plink binary &quot;</span>
                         <span class="s2">&quot;files by appending the prefix to the .bim, .bed and &quot;</span>
                         <span class="s2">&quot;.fam files, respectively).&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--gold-bfile&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The input file prefix (will find the plink binary &quot;</span>
                         <span class="s2">&quot;files by appending the prefix to the .bim, .bed and &quot;</span>
                         <span class="s2">&quot;.fam files, respectively) for the Gold Standard.&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--same-samples&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;A file containing samples which are present in both &quot;</span>
                         <span class="s2">&quot;the gold standard and the source panel. One line by &quot;</span>
                         <span class="s2">&quot;identity and tab separated. For each row, first &quot;</span>
                         <span class="s2">&quot;sample is Gold Standard, second is source panel.&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--source-manifest&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The illumina marker manifest. This file should have &quot;</span>
                         <span class="s2">&quot;tabs as field separator. There should be no lines &quot;</span>
                         <span class="s2">&quot;before the main header line. There should be no &quot;</span>
                         <span class="s2">&quot;lines after the last data line.&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--source-alleles&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;A file containing the source alleles (TOP). Two &quot;</span>
                         <span class="s2">&quot;columns (separated by tabulation, one with the &quot;</span>
                         <span class="s2">&quot;marker name, the other with the alleles (separated &quot;</span>
                         <span class="s2">&quot;by space). No header.&quot;</span><span class="p">))</span>

<span class="c1"># The options</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Options&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--sge&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Use SGE for parallelization.&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--do-not-flip&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;Do not flip SNPs. WARNING: only use this option &quot;</span>
                         <span class="s2">&quot;only if the Gold Standard was generated using the &quot;</span>
                         <span class="s2">&quot;same chip (hence, flipping is unnecessary).&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--use-marker-names&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;Use marker names instead of (chr, position). &quot;</span>
                         <span class="s2">&quot;WARNING: only use this options only if the Gold &quot;</span>
                         <span class="s2">&quot;Standard was generated using the same chip (hence, &quot;</span>
                         <span class="s2">&quot;they have the same marker names).&quot;</span><span class="p">))</span>
<span class="c1"># The OUTPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Output File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">default</span><span class="o">=</span><span class="s2">&quot;compare_with_gold&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The prefix of the output files. [default: &quot;</span>
                         <span class="s2">&quot;</span><span class="si">%(default)s</span><span class="s2">]&quot;</span><span class="p">))</span>


<div class="viewcode-block" id="safe_main"><a class="viewcode-back" href="../../../module_content/pyGenClean.Misc.html#pyGenClean.Misc.compare_gold_standard.safe_main">[docs]</a><span class="k">def</span> <span class="nf">safe_main</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;A safe version of the main function (that catches ProgramError).&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">main</span><span class="p">()</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span></div>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">safe_main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" >pyGenClean</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>