<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pyGenClean.DupSamples.duplicated_samples &mdash; pyGenClean 1.8.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.8.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.8.2 documentation" href="../../../index.html" />
    <link rel="up" title="pyGenClean" href="../../pyGenClean.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="../../../index.html"><img src="../../../_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" accesskey="U">pyGenClean</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for pyGenClean.DupSamples.duplicated_samples</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python2.7</span>

<span class="c1"># This file is part of pyGenClean.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is free software: you can redistribute it and/or modify it under</span>
<span class="c1"># the terms of the GNU General Public License as published by the Free Software</span>
<span class="c1"># Foundation, either version 3 of the License, or (at your option) any later</span>
<span class="c1"># version.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is distributed in the hope that it will be useful, but WITHOUT ANY</span>
<span class="c1"># WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR</span>
<span class="c1"># A PARTICULAR PURPOSE.  See the GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1"># You should have received a copy of the GNU General Public License along with</span>
<span class="c1"># pyGenClean.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">StringIO</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">__version__</span>


<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;duplicated_samples&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check for duplicated samples in a tfam/tped file.</span>

<span class="sd">    :param argString: the options</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    Here are the steps for the duplicated samples step.</span>

<span class="sd">    1.  Prints the options.</span>
<span class="sd">    2.  Reads the ``tfam`` file (:py:func:`readTFAM`).</span>
<span class="sd">    3.  Separate the duplicated samples from the unique samples</span>
<span class="sd">        (:py:func:`findDuplicates`).</span>
<span class="sd">    4.  Writes the unique samples into a file named</span>
<span class="sd">        ``prefix.unique_samples.tfam`` (:py:func:`printUniqueTFAM`).</span>
<span class="sd">    5.  Reads the ``tped`` file and write into ``prefix.unique_samples.tped``</span>
<span class="sd">        the pedigree file for the unique samples (:py:func:`processTPED`).</span>
<span class="sd">        Saves in memory the pedigree for the duplicated samples. Updates the</span>
<span class="sd">        indexes of the duplicated samples.</span>
<span class="sd">    6.  If there are no duplicated samples, simply copies the files</span>
<span class="sd">        ``prefix.unique_samples`` (``tped`` and ``tfam``) to</span>
<span class="sd">        ``prefix.final.tfam`` and ``prefix..final.tped``, respectively.</span>
<span class="sd">    7.  Computes the completion (for each of the duplicated samples) and the</span>
<span class="sd">        concordance of each sample pairs (:py:func:`computeStatistics`).</span>
<span class="sd">    8.  Prints statistics (concordance and completion)</span>
<span class="sd">        (:py:func:`printStatistics`).</span>
<span class="sd">    9.  We print the concordance matrix for each duplicated samples</span>
<span class="sd">        (:py:func:`printConcordance`).</span>
<span class="sd">    10. We print the ``tped`` and the ``tfam`` file for the duplicated samples</span>
<span class="sd">        (``prefix.duplicated_samples``)</span>
<span class="sd">        (:py:func:`printDuplicatedTPEDandTFAM`).</span>
<span class="sd">    11. Choose the best of each duplicates (to keep and to complete) according</span>
<span class="sd">        to completion and concordance (:py:func:`chooseBestDuplicates`).</span>
<span class="sd">    12. Creates a unique ``tped`` and ``tfam`` from the duplicated samples by</span>
<span class="sd">        completing the best chosen one with the other samples</span>
<span class="sd">        (:py:func:`createAndCleanTPED`).</span>
<span class="sd">    13. Merge the two tfiles together (``prefix.unique_samples`` and</span>
<span class="sd">        ``prefix.chosen_samples``) to create the final dataset</span>
<span class="sd">        (``prefix.final``) (:py:func:`addToTPEDandTFAM`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Getting and checking the options</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>
    <span class="n">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Options used:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  --{} {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">),</span> <span class="n">value</span><span class="p">))</span>

    <span class="c1"># Reading the tfam file</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading TFAM&quot;</span><span class="p">)</span>
    <span class="n">tfam</span> <span class="o">=</span> <span class="n">readTFAM</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">tfile</span> <span class="o">+</span> <span class="s2">&quot;.tfam&quot;</span><span class="p">)</span>

    <span class="c1"># Find duplicated samples</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finding duplicated samples&quot;</span><span class="p">)</span>
    <span class="n">uniqueSamples</span><span class="p">,</span> <span class="n">duplicatedSamples</span> <span class="o">=</span> <span class="n">findDuplicates</span><span class="p">(</span><span class="n">tfam</span><span class="p">)</span>

    <span class="c1"># Prints the unique tfam</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating TFAM for unique samples&quot;</span><span class="p">)</span>
    <span class="n">printUniqueTFAM</span><span class="p">(</span><span class="n">tfam</span><span class="p">,</span> <span class="n">uniqueSamples</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

    <span class="c1"># Process the TPED file</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading TPED (and creating TPED for unique samples)&quot;</span><span class="p">)</span>
    <span class="n">tped</span><span class="p">,</span> <span class="n">tpedSamples</span> <span class="o">=</span> <span class="n">processTPED</span><span class="p">(</span><span class="n">uniqueSamples</span><span class="p">,</span> <span class="n">duplicatedSamples</span><span class="p">,</span>
                                    <span class="n">args</span><span class="o">.</span><span class="n">tfile</span> <span class="o">+</span> <span class="s2">&quot;.tped&quot;</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">duplicatedSamples</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;There are no duplicates in {}.tfam&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">tfile</span><span class="p">))</span>
        <span class="c1"># There are no duplicated samples</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.unique_samples.tfam&quot;</span><span class="p">,</span>
                        <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.final.tfam&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">.unique_samples.tfam: can&#39;t copy to &quot;</span> \
                  <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">.final.tfam&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.unique_samples.tped&quot;</span><span class="p">,</span>
                        <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.final.tped&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">.unique_samples.tped: can&#39;t copy to &quot;</span> \
                  <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">.final.tped&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># We continue</span>
        <span class="c1"># Compute statistics</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing the completion and concordance of duplicated &quot;</span>
                    <span class="s2">&quot;samples&quot;</span><span class="p">)</span>
        <span class="n">completion</span><span class="p">,</span> <span class="n">concordance</span> <span class="o">=</span> <span class="n">computeStatistics</span><span class="p">(</span><span class="n">tped</span><span class="p">,</span> <span class="n">tfam</span><span class="p">,</span> <span class="n">tpedSamples</span><span class="p">,</span>
                                                    <span class="n">duplicatedSamples</span><span class="p">,</span>
                                                    <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

        <span class="c1"># Print the statistics</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Printing the statistics&quot;</span><span class="p">)</span>
        <span class="n">completion_percentage</span> <span class="o">=</span> <span class="n">printStatistics</span><span class="p">(</span><span class="n">completion</span><span class="p">,</span> <span class="n">concordance</span><span class="p">,</span>
                                                <span class="n">tpedSamples</span><span class="p">,</span> <span class="n">duplicatedSamples</span><span class="p">,</span>
                                                <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

        <span class="c1"># Print the concordance file</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Printing the concordance file&quot;</span><span class="p">)</span>
        <span class="n">concordance_percentage</span> <span class="o">=</span> <span class="n">printConcordance</span><span class="p">(</span><span class="n">concordance</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

        <span class="c1"># Print the duplicated TFAM and TPED</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating TPED and TFAM for duplicated samples&quot;</span><span class="p">)</span>
        <span class="n">printDuplicatedTPEDandTFAM</span><span class="p">(</span><span class="n">tped</span><span class="p">,</span> <span class="n">tfam</span><span class="p">,</span> <span class="n">tpedSamples</span><span class="p">,</span> <span class="n">duplicatedSamples</span><span class="p">,</span>
                                   <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

        <span class="c1"># Choose the best duplicates</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Choosing the best duplicates&quot;</span><span class="p">)</span>
        <span class="n">chosenSamples</span><span class="p">,</span> <span class="n">comp</span><span class="p">,</span> <span class="n">conc</span> <span class="o">=</span> <span class="n">chooseBestDuplicates</span><span class="p">(</span>
            <span class="n">tped</span><span class="p">,</span>
            <span class="n">tpedSamples</span><span class="p">,</span>
            <span class="n">duplicatedSamples</span><span class="p">,</span>
            <span class="n">completion_percentage</span><span class="p">,</span>
            <span class="n">concordance_percentage</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Clean the genotype of the chosen samples</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cleaning and creating unique TPED and TFAM from &quot;</span>
                    <span class="s2">&quot;duplicated samples&quot;</span><span class="p">)</span>
        <span class="n">newTPED</span><span class="p">,</span> <span class="n">newTFAM</span> <span class="o">=</span> <span class="n">createAndCleanTPED</span><span class="p">(</span>
            <span class="n">tped</span><span class="p">,</span>
            <span class="n">tfam</span><span class="p">,</span>
            <span class="n">tpedSamples</span><span class="p">,</span>
            <span class="n">duplicatedSamples</span><span class="p">,</span>
            <span class="n">chosenSamples</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
            <span class="n">comp</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">sample_completion_threshold</span><span class="p">,</span>
            <span class="n">conc</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">sample_concordance_threshold</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Add the chosen TPED and TFAM</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating final TPED and TFAM file&quot;</span><span class="p">)</span>
        <span class="n">addToTPEDandTFAM</span><span class="p">(</span><span class="n">newTPED</span><span class="p">,</span> <span class="n">newTFAM</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
                         <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.unique_samples&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="addToTPEDandTFAM"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.addToTPEDandTFAM">[docs]</a><span class="k">def</span> <span class="nf">addToTPEDandTFAM</span><span class="p">(</span><span class="n">tped</span><span class="p">,</span> <span class="n">tfam</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="n">toAddPrefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Append a tfile to another, creating a new one.</span>

<span class="sd">    :param tped: the ``tped`` that will be appended to the other one.</span>
<span class="sd">    :param tfam: the ``tfam`` that will be appended to the other one.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>
<span class="sd">    :param toAddPrefix: the prefix of the final file.</span>

<span class="sd">    :type tped: :py:class:`numpy.array`</span>
<span class="sd">    :type tfam: :py:class:`numpy.array`</span>
<span class="sd">    :type prefix: str</span>
<span class="sd">    :type toAddPrefix: str</span>

<span class="sd">    Here are the steps of this function:</span>

<span class="sd">    1. Writes the ``tped`` into ``prefix.chosen_samples.tped``.</span>
<span class="sd">    2. Writes the ``tfam`` into ``prefix.chosen_samples.tfam``.</span>
<span class="sd">    3. Copies the previous ``tfam`` (``toAddPrefix.tfam``) into the final</span>
<span class="sd">       ``tfam`` (``prefix.final.tfam``).</span>
<span class="sd">    4. Append the ``tfam`` to the final ``tfam`` (``prefix.final.tfam``).</span>
<span class="sd">    5. Reads the previous ``tped`` (``toAddPrefix.tped``) and append the new</span>
<span class="sd">       ``tped`` to it, writing the final one (``prefix.final.tped``).</span>

<span class="sd">    .. warning::</span>
<span class="sd">        The ``tped`` and ``tfam`` variables need to contain at least one</span>
<span class="sd">        sample.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First, writing the chosen tped</span>
    <span class="n">tpedFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">tpedFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.chosen_samples.tped&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.chosen_samples.tped: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tped</span><span class="p">:</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">tpedFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="n">tpedFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Second, writing the chosen tfam</span>
    <span class="n">tfamFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">tfamFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.chosen_samples.tfam&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.chosen_samples.tfam: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tfam</span><span class="p">:</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">tfamFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="n">tfamFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># The final TFAM file (copying and appending)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">toAddPrefix</span> <span class="o">+</span> <span class="s2">&quot;.tfam&quot;</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.final.tfam&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(toAddPrefix)s</span><span class="s2">.tfam: can&#39;t copy to &quot;</span> \
              <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.final.tfam&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="n">newTFAM</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">newTFAM</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.final.tfam&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.final.tfam: can&#39;t append&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tfam</span><span class="p">:</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">newTFAM</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="n">newTFAM</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># The final TPED file</span>
    <span class="n">newTPED</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">newTPED</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.final.tped&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.final.tped: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">toAddPrefix</span> <span class="o">+</span> <span class="s2">&quot;.tped&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">inputFile</span><span class="p">):</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">originalSNP</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">toAdd</span> <span class="o">=</span> <span class="n">tped</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">originalSNP</span> <span class="o">!=</span> <span class="n">toAdd</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(toAddPrefix)s</span><span class="s2">.tped: not good sort&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
                <span class="n">row</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">toAdd</span><span class="p">[</span><span class="mi">4</span><span class="p">:]))</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">newTPED</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(toAddPrefix)s</span><span class="s2">.tped: no such file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="n">newTPED</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="createAndCleanTPED"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.createAndCleanTPED">[docs]</a><span class="k">def</span> <span class="nf">createAndCleanTPED</span><span class="p">(</span><span class="n">tped</span><span class="p">,</span> <span class="n">tfam</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">oldSamples</span><span class="p">,</span> <span class="n">chosenSamples</span><span class="p">,</span>
                       <span class="n">prefix</span><span class="p">,</span> <span class="n">completion</span><span class="p">,</span> <span class="n">completionT</span><span class="p">,</span> <span class="n">concordance</span><span class="p">,</span>
                       <span class="n">concordanceT</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Complete a TPED for duplicate samples.</span>

<span class="sd">    :param tped: the ``tped`` containing the duplicated samples.</span>
<span class="sd">    :param tfam: the ``tfam`` containing the duplicated samples.</span>
<span class="sd">    :param samples: the updated position of the samples in the ``tped``</span>
<span class="sd">                    containing only duplicated samples.</span>
<span class="sd">    :param oldSamples: the original duplicated sample positions.</span>
<span class="sd">    :param chosenSamples: the position of the chosen samples.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>
<span class="sd">    :param completion: the completion of each of the duplicated samples.</span>
<span class="sd">    :param completionT: the completion threshold.</span>
<span class="sd">    :param concordance: the pairwise concordance of each of the duplicated</span>
<span class="sd">                        samples.</span>
<span class="sd">    :param concordanceT: the concordance threshold.</span>

<span class="sd">    :type tped: :py:class:`numpy.array`</span>
<span class="sd">    :type tfam: :py:class:`numpy.array`</span>
<span class="sd">    :type samples: dict</span>
<span class="sd">    :type oldSamples: dict</span>
<span class="sd">    :type chosenSamples: dict</span>
<span class="sd">    :type prefix: str</span>
<span class="sd">    :type completion: :py:class:`numpy.array`</span>
<span class="sd">    :type completionT: float</span>
<span class="sd">    :type concordance: dict</span>
<span class="sd">    :type concordanceT: float</span>

<span class="sd">    Using a ``tped`` containing duplicated samples, it creates a ``tped``</span>
<span class="sd">    containing unique samples by completing a chosen sample with the other</span>
<span class="sd">    replicates.</span>

<span class="sd">    .. note::</span>
<span class="sd">        A chosen sample is not completed using bad replicates (those that</span>
<span class="sd">        don&#39;t have a concordance or a completion higher than a certain</span>
<span class="sd">        threshold). The bad replicates are written in the file</span>
<span class="sd">        ``prefix.not_good_enough``.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">zeroedOutFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">zeroedOutFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.zeroed_out&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.zeroed_out: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">zeroedOutFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;famID&quot;</span><span class="p">,</span> <span class="s2">&quot;indID&quot;</span><span class="p">,</span> <span class="s2">&quot;snpID&quot;</span><span class="p">])</span>

    <span class="n">notGoodEnoughFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">notGoodEnoughFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.not_good_enough&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.not_good_enough: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">notGoodEnoughFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;origIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;dupIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;famID&quot;</span><span class="p">,</span>
                                          <span class="s2">&quot;indID&quot;</span><span class="p">,</span> <span class="s2">&quot;reason&quot;</span><span class="p">])</span>
    <span class="n">notGoodEnoughSamples</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

    <span class="c1"># Split the tped in &#39;snpInfo&#39; and &#39;genotypes&#39;</span>
    <span class="n">snpInfo</span> <span class="o">=</span> <span class="n">tped</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">]</span>
    <span class="n">genotypes</span> <span class="o">=</span> <span class="n">tped</span><span class="p">[:,</span> <span class="mi">4</span><span class="p">:]</span>

    <span class="c1"># Getting the completion and concordance indexes</span>
    <span class="n">completionSampleID</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">concordanceSampleID</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">indexes</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="c1"># Checking the completion</span>
        <span class="n">completionToKeep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">completion</span><span class="p">[</span><span class="n">indexes</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">completionT</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">completionToRemove</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">completion</span><span class="p">[</span><span class="n">indexes</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">completionT</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">completionToRemove</span><span class="p">:</span>
            <span class="n">notGoodEnoughSamples</span><span class="p">[(</span><span class="nb">str</span><span class="p">(</span><span class="n">oldSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">][</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>
                                  <span class="nb">str</span><span class="p">(</span><span class="n">indexes</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">sampleID</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                  <span class="n">sampleID</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;completion&quot;</span><span class="p">)</span>

        <span class="c1"># Checking the concordance</span>
        <span class="n">concordanceToKeep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">concordance</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">concordanceT</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">concordanceToRemove</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="n">concordance</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">concordanceT</span>
        <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">concordanceToRemove</span><span class="p">:</span>
            <span class="n">notGoodEnoughSamples</span><span class="p">[(</span><span class="nb">str</span><span class="p">(</span><span class="n">oldSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">][</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>
                                  <span class="nb">str</span><span class="p">(</span><span class="n">indexes</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">sampleID</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                  <span class="n">sampleID</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;concordance&quot;</span><span class="p">)</span>

        <span class="n">completionSampleID</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span> <span class="o">=</span> <span class="n">completionToKeep</span>
        <span class="n">concordanceSampleID</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span> <span class="o">=</span> <span class="n">concordanceToKeep</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">curSNPgenotypes</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genotypes</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">indexes</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="c1"># Getting the concordance and the completion</span>
            <span class="n">concordanceToKeep</span> <span class="o">=</span> <span class="n">concordanceSampleID</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span>
            <span class="n">completionToKeep</span> <span class="o">=</span> <span class="n">completionSampleID</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span>

            <span class="c1"># Getting the chosen sample index</span>
            <span class="n">chosenOne</span> <span class="o">=</span> <span class="n">chosenSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span>

            <span class="c1"># Getting the duplicates&#39; genotypes</span>
            <span class="n">curGenotypes</span> <span class="o">=</span> <span class="n">curSNPgenotypes</span><span class="p">[</span><span class="n">indexes</span><span class="p">]</span>

            <span class="c1"># Subsetting for the good completion and concordance</span>
            <span class="n">curGenotypes</span> <span class="o">=</span> <span class="n">curGenotypes</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">concordanceToKeep</span><span class="p">,</span>
                                                       <span class="n">completionToKeep</span><span class="p">)]</span>

            <span class="c1"># Removing the no call from the genotypes</span>
            <span class="n">cleanedCurGenotype</span> <span class="o">=</span> <span class="n">curGenotypes</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">curGenotypes</span> <span class="o">!=</span> <span class="s2">&quot;0 0&quot;</span><span class="p">)]</span>

            <span class="c1"># Checking the number of unique genotypes</span>
            <span class="n">uniqueGenotypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">cleanedCurGenotype</span><span class="p">)</span>

            <span class="c1"># Checking the number of unique genotypes (except 0 0)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">uniqueGenotypes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># There are more than one unique genotypes (except 0 0)</span>
                <span class="c1"># len = 0 means all were 0 0</span>
                <span class="c1"># len = 1 means all the same</span>
                <span class="c1"># len &gt; 1 means more than one unique genotypes</span>
                <span class="n">possibleAlleles</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="nb">set</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">uniqueGenotypes</span><span class="p">))</span>
                <span class="p">]</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">geno</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">uniqueGenotypes</span><span class="p">):</span>
                    <span class="n">possibleAlleles</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">|=</span> <span class="nb">set</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">))</span>
                <span class="n">allEqual</span> <span class="o">=</span> <span class="bp">True</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">possibleAlleles</span><span class="p">)):</span>
                    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">possibleAlleles</span><span class="p">)):</span>
                        <span class="k">if</span> <span class="n">possibleAlleles</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">!=</span> <span class="n">possibleAlleles</span><span class="p">[</span><span class="n">l</span><span class="p">]:</span>
                            <span class="n">allEqual</span> <span class="o">=</span> <span class="bp">False</span>
                            <span class="k">break</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">allEqual</span><span class="p">:</span>
                        <span class="k">break</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">allEqual</span><span class="p">:</span>
                    <span class="c1"># Changing the chosen sample&#39;s genotype to no call</span>
                    <span class="n">tped</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">chosenOne</span><span class="o">+</span><span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;0 0&quot;</span>
                    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">zeroedOutFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">sampleID</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">sampleID</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                                      <span class="n">snpInfo</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Do we want to complete the chosen sample&#39;s genotype???</span>
                <span class="k">pass</span>

    <span class="c1"># Writing the not good enough file</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">notGoodEnoughSamples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">notGoodEnoughFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">value</span><span class="p">)])</span>

    <span class="c1"># Closing the output files</span>
    <span class="n">zeroedOutFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">notGoodEnoughFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Getting the indexes to keep</span>
    <span class="n">sampleToKeep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">chosenSamples</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">+</span> <span class="mi">4</span>
    <span class="n">sampleToKeep</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">toKeep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">)),</span> <span class="n">sampleToKeep</span><span class="p">)</span>

    <span class="c1"># Subsetting the tped</span>
    <span class="n">newTPED</span> <span class="o">=</span> <span class="n">tped</span><span class="p">[:,</span> <span class="n">toKeep</span><span class="p">]</span>

    <span class="c1"># Getting the indexes of the tfam</span>
    <span class="n">toKeep</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">indexes</span> <span class="ow">in</span> <span class="n">oldSamples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chosenSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">])</span>
        <span class="n">toKeep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indexes</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">toKeep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">toKeep</span><span class="p">)</span>
    <span class="n">toKeep</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

    <span class="c1"># Subsetting the tfam</span>
    <span class="n">newTFAM</span> <span class="o">=</span> <span class="n">tfam</span><span class="p">[</span><span class="n">toKeep</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">newTPED</span><span class="p">,</span> <span class="n">newTFAM</span></div>


<div class="viewcode-block" id="chooseBestDuplicates"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.chooseBestDuplicates">[docs]</a><span class="k">def</span> <span class="nf">chooseBestDuplicates</span><span class="p">(</span><span class="n">tped</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">oldSamples</span><span class="p">,</span> <span class="n">completion</span><span class="p">,</span>
                         <span class="n">concordance_all</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Choose the best duplicates according to the completion rate.</span>

<span class="sd">    :param tped: the ``tped`` containing the duplicated samples.</span>
<span class="sd">    :param samples: the updated position of the samples in the tped containing</span>
<span class="sd">                    only duplicated samples.</span>
<span class="sd">    :param oldSamples: the original duplicated sample positions.</span>
<span class="sd">    :param completion: the completion of each of the duplicated samples.</span>
<span class="sd">    :param concordance_all: the concordance of every duplicated samples.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>

<span class="sd">    :type tped: :py:class:`numpy.array`</span>
<span class="sd">    :type samples: dict</span>
<span class="sd">    :type oldSamples: dict</span>
<span class="sd">    :type completion: :py:class:`numpy.array`</span>
<span class="sd">    :type concordance_all: dict</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    :returns: a tuple where the first element is a list of the chosen samples&#39;</span>
<span class="sd">              indexes, the second on is the completion and the last one is the</span>
<span class="sd">              concordance (a map).</span>

<span class="sd">    These are the steps to find the best duplicated sample:</span>

<span class="sd">    1. Sort the list of concordances.</span>
<span class="sd">    2. Sort the list of completions.</span>
<span class="sd">    3. Choose the best of the concordance and put in a set.</span>
<span class="sd">    4. Choose the best of the completion and put it in a set.</span>
<span class="sd">    5. Compute the intersection of the two sets. If there is one sample or</span>
<span class="sd">       more, then randomly choose one sample.</span>
<span class="sd">    6. If the intersection doesn&#39;t contain at least one sample, redo steps 3</span>
<span class="sd">       and 4, but increase the number of chosen best by one. Redo step 5 and 6</span>
<span class="sd">       (if required).</span>

<span class="sd">    The chosen samples are written in ``prefix.chosen_samples.info``. The rest</span>
<span class="sd">    are written in ``prefix.excluded_samples.info``.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The output files</span>
    <span class="n">chosenFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">chosenFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.chosen_samples.info&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.chosen_samples.info: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">chosenFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;origIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;dupIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;famID&quot;</span><span class="p">,</span> <span class="s2">&quot;indID&quot;</span><span class="p">])</span>

    <span class="n">excludedFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">excludedFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.excluded_samples.info&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.excluded_samples.info: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">excludedFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;origIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;dupIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;famID&quot;</span><span class="p">,</span>
                                     <span class="s2">&quot;indID&quot;</span><span class="p">])</span>

    <span class="c1"># For each duplicated sample</span>
    <span class="n">chosenIndexes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">sampleConcordance</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">indexes</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="c1"># Getting the completion for those duplicated samples</span>
        <span class="n">currCompletion</span> <span class="o">=</span> <span class="n">completion</span><span class="p">[</span><span class="n">indexes</span><span class="p">]</span>

        <span class="c1"># Sorting those completion</span>
        <span class="n">sortedCompletionIndexes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">currCompletion</span><span class="p">)</span>

        <span class="c1"># Getting the concordance</span>
        <span class="n">concordance</span> <span class="o">=</span> <span class="n">concordance_all</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span>
        <span class="n">currConcordance</span> <span class="o">=</span> <span class="p">[[]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indexes</span><span class="p">))]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indexes</span><span class="p">)):</span>
            <span class="n">indexToKeep</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indexes</span><span class="p">)))</span> <span class="o">-</span> <span class="nb">set</span><span class="p">([</span><span class="n">i</span><span class="p">]))</span>
            <span class="n">currConcordance</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">concordance</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">indexToKeep</span><span class="p">])</span>
        <span class="n">currConcordance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">currConcordance</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">sample</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sampleConcordance</span><span class="p">:</span>
            <span class="n">sampleConcordance</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">currConcordance</span>

        <span class="c1"># Sorting the concordance</span>
        <span class="n">sortedConcordanceIndexes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">currConcordance</span><span class="p">)</span>

        <span class="c1"># Trying to find the best duplicate to keep</span>
        <span class="n">nbToCheck</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">chosenIndex</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">while</span> <span class="n">nbToCheck</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">indexes</span><span class="p">):</span>
            <span class="c1"># Getting the `nbToCheck` best value (higher to lower)</span>
            <span class="n">completionValue</span> <span class="o">=</span> <span class="n">currCompletion</span><span class="p">[</span>
                <span class="n">sortedCompletionIndexes</span><span class="p">[</span><span class="n">nbToCheck</span><span class="o">*-</span><span class="mi">1</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="n">concordanceValue</span> <span class="o">=</span> <span class="n">currConcordance</span><span class="p">[</span>
                <span class="n">sortedConcordanceIndexes</span><span class="p">[</span><span class="n">nbToCheck</span><span class="o">*-</span><span class="mi">1</span><span class="p">]</span>
            <span class="p">]</span>

            <span class="c1"># Getting the indexes to consider</span>
            <span class="n">completionToConsider</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">currCompletion</span> <span class="o">&gt;=</span> <span class="n">completionValue</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">concordanceToConsider</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">currConcordance</span> <span class="o">&gt;=</span> <span class="n">concordanceValue</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

            <span class="c1"># Getting the intersection of the indexes</span>
            <span class="n">toConsider</span> <span class="o">=</span> <span class="n">concordanceToConsider</span> <span class="o">&amp;</span> <span class="n">completionToConsider</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">toConsider</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">chosenIndex</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">toConsider</span><span class="p">))</span>
                <span class="k">break</span>
            <span class="n">nbToCheck</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">chosenIndex</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Could not choose the best sample ID for {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="c1"># Printing the chosen samples</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">chosenFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">oldSamples</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">chosenIndex</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>
                                       <span class="nb">str</span><span class="p">(</span><span class="n">indexes</span><span class="p">[</span><span class="n">chosenIndex</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">sample</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                       <span class="n">sample</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span>

        <span class="c1"># Printing the excluded samples</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indexes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">chosenIndex</span><span class="p">:</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">excludedFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">oldSamples</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>
                                                 <span class="nb">str</span><span class="p">(</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">sample</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                                 <span class="n">sample</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span>

        <span class="n">chosenIndexes</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">indexes</span><span class="p">[</span><span class="n">chosenIndex</span><span class="p">]</span>

    <span class="c1"># Closing the output files</span>
    <span class="n">chosenFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">excludedFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">chosenIndexes</span><span class="p">,</span> <span class="n">completion</span><span class="p">,</span> <span class="n">sampleConcordance</span></div>


<div class="viewcode-block" id="printDuplicatedTPEDandTFAM"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.printDuplicatedTPEDandTFAM">[docs]</a><span class="k">def</span> <span class="nf">printDuplicatedTPEDandTFAM</span><span class="p">(</span><span class="n">tped</span><span class="p">,</span> <span class="n">tfam</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">oldSamples</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Print the TPED and TFAM of the duplicated samples.</span>

<span class="sd">    :param tped: the ``tped`` containing duplicated samples.</span>
<span class="sd">    :param tfam: the ``tfam`` containing duplicated samples.</span>
<span class="sd">    :param samples: the updated position of the samples in the tped containing</span>
<span class="sd">                    only duplicated samples.</span>
<span class="sd">    :param oldSamples: the original duplicated sample positions.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>

<span class="sd">    :type tped: :py:class:`numpy.array`</span>
<span class="sd">    :type tfam: :py:class:`numpy.array`</span>
<span class="sd">    :type samples: dict</span>
<span class="sd">    :type oldSamples: dict</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    The ``tped`` and ``tfam`` files are written in</span>
<span class="sd">    ``prefix.duplicated_samples.tped`` and ``prefix.duplicated_samples.tfam``,</span>
<span class="sd">    respectively.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Print the TPED</span>
    <span class="n">outputTPED</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">outputTPED</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.duplicated_samples.tped&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.duplicated_samples.tped: can&#39;t write &quot;</span> \
              <span class="s2">&quot;file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tped</span><span class="p">:</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outputTPED</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="n">outputTPED</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Print the TFAM</span>
    <span class="n">nbSamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tped</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">4</span><span class="p">:])</span>
    <span class="n">newTFAM</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">nbSamples</span><span class="p">)]</span>
    <span class="k">for</span> <span class="n">samples</span><span class="p">,</span> <span class="n">indexes</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">oldIndexes</span> <span class="o">=</span> <span class="n">oldSamples</span><span class="p">[</span><span class="n">samples</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indexes</span><span class="p">):</span>
            <span class="n">oldIndex</span> <span class="o">=</span> <span class="n">oldIndexes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">newTFAM</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">tfam</span><span class="p">[</span><span class="n">oldIndex</span><span class="p">]</span>
    <span class="n">outputTFAM</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">outputTFAM</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.duplicated_samples.tfam&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.duplicated_samples.tfam: can&#39;t write &quot;</span> \
              <span class="s2">&quot;file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">newTFAM</span><span class="p">:</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outputTFAM</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="n">outputTFAM</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="printConcordance"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.printConcordance">[docs]</a><span class="k">def</span> <span class="nf">printConcordance</span><span class="p">(</span><span class="n">concordance</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Print the concordance.</span>

<span class="sd">    :param concordance: the concordance of each sample.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>

<span class="sd">    :type concordance: dict</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    :returns: the concordance percentage (dict)</span>

<span class="sd">    The concordance is the number of genotypes that are equal when comparing a</span>
<span class="sd">    duplicated samples with another one, divided by the total number of</span>
<span class="sd">    genotypes (excluding genotypes that are no call [*i.e.* ``0``]). If a</span>
<span class="sd">    duplicated sample has 100% of no calls, the concordance will be zero.</span>

<span class="sd">    The file ``prefix.concordance`` will contain :math:`N \\times N` matrices</span>
<span class="sd">    for each set of duplicated samples.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">outFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">outFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.concordance&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.concordance&quot;</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">concordance_percentage</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">concordance</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outFile</span><span class="p">,</span> <span class="s2">&quot;#</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">key</span>

        <span class="c1"># Doing the division</span>
        <span class="n">none_zero</span> <span class="o">=</span> <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span>
        <span class="n">true_concordance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="o">*</span><span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
        <span class="n">true_concordance</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span><span class="n">none_zero</span><span class="p">)]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">true_divide</span><span class="p">(</span>
            <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="n">none_zero</span><span class="p">],</span>
            <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="n">none_zero</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="n">true_concordance</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
        <span class="n">true_concordance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asmatrix</span><span class="p">(</span><span class="n">true_concordance</span><span class="p">)</span>
        <span class="n">concordance_percentage</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">true_concordance</span>

        <span class="n">output</span> <span class="o">=</span> <span class="n">StringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
        <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="n">true_concordance</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%.8f</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outFile</span><span class="p">,</span> <span class="n">output</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">outFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">concordance_percentage</span></div>


<div class="viewcode-block" id="printStatistics"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.printStatistics">[docs]</a><span class="k">def</span> <span class="nf">printStatistics</span><span class="p">(</span><span class="n">completion</span><span class="p">,</span> <span class="n">concordance</span><span class="p">,</span> <span class="n">tpedSamples</span><span class="p">,</span> <span class="n">oldSamples</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Print the statistics in a file.</span>

<span class="sd">    :param completion: the completion of each duplicated samples.</span>
<span class="sd">    :param concordance: the concordance of each duplicated samples.</span>
<span class="sd">    :param tpedSamples: the updated position of the samples in the tped</span>
<span class="sd">                        containing only duplicated samples.</span>
<span class="sd">    :param oldSamples: the original duplicated sample positions.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>

<span class="sd">    :type completion: :py:class:`numpy.array`</span>
<span class="sd">    :type concordance: dict</span>
<span class="sd">    :type tpedSamples: dict</span>
<span class="sd">    :type oldSamples: dict</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    :returns: the completion for each duplicated samples, as a</span>
<span class="sd">              :py:class:`numpy.array`.</span>

<span class="sd">    Prints the statistics (completion of each samples and pairwise concordance</span>
<span class="sd">    between duplicated samples) in a file (``prefix.summary``).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Compute the completion percentage on none zero values</span>
    <span class="n">none_zero_indexes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">completion</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">completionPercentage</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">completion</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">completionPercentage</span><span class="p">[</span><span class="n">none_zero_indexes</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">true_divide</span><span class="p">(</span>
        <span class="n">completion</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">none_zero_indexes</span><span class="p">],</span>
        <span class="n">completion</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="n">none_zero_indexes</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="c1"># The output file containing the summary statistics (for each of the</span>
    <span class="c1"># duplicated samples, print the mean concordance and the completion).</span>
    <span class="n">outputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">outputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.summary&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.summary: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outputFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;origIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;dupIndex&quot;</span><span class="p">,</span> <span class="s2">&quot;famID&quot;</span><span class="p">,</span> <span class="s2">&quot;indID&quot;</span><span class="p">,</span>
                                   <span class="s2">&quot;</span><span class="si">% c</span><span class="s2">ompletion&quot;</span><span class="p">,</span> <span class="s2">&quot;completion&quot;</span><span class="p">,</span>
                                   <span class="s2">&quot;mean concordance&quot;</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">indexes</span> <span class="ow">in</span> <span class="n">tpedSamples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indexes</span><span class="p">):</span>
            <span class="c1"># The indexes</span>
            <span class="n">toPrint</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">oldSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
            <span class="c1"># The samples</span>
            <span class="n">toPrint</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">sampleID</span><span class="p">))</span>

            <span class="c1"># The completion</span>
            <span class="n">toPrint</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%.8f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">completionPercentage</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
            <span class="n">toPrint</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">/</span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">completion</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">index</span><span class="p">],</span>
                                      <span class="n">completion</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">index</span><span class="p">]))</span>

            <span class="c1"># The concordance (not on total values = 0)</span>
            <span class="n">indexToKeep</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indexes</span><span class="p">)))</span> <span class="o">-</span> <span class="nb">set</span><span class="p">([</span><span class="n">i</span><span class="p">]))</span>
            <span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">concordance</span><span class="p">[</span><span class="n">sampleID</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="n">i</span><span class="p">,</span> <span class="n">indexToKeep</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="n">total_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">concordance</span><span class="p">[</span><span class="n">sampleID</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="n">i</span><span class="p">,</span> <span class="n">indexToKeep</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="n">currConcordance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indexToKeep</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
            <span class="n">none_zero_indexes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">total_values</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>
            <span class="n">currConcordance</span><span class="p">[</span><span class="n">none_zero_indexes</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">true_divide</span><span class="p">(</span>
                <span class="n">values</span><span class="p">[</span><span class="n">none_zero_indexes</span><span class="p">],</span>
                <span class="n">total_values</span><span class="p">[</span><span class="n">none_zero_indexes</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="n">currConcordance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">currConcordance</span><span class="p">)</span>
            <span class="n">toPrint</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%.8f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">currConcordance</span><span class="p">)</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outputFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">toPrint</span><span class="p">)</span>

    <span class="c1"># Closing the output file</span>
    <span class="n">outputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">completionPercentage</span></div>


<div class="viewcode-block" id="computeStatistics"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.computeStatistics">[docs]</a><span class="k">def</span> <span class="nf">computeStatistics</span><span class="p">(</span><span class="n">tped</span><span class="p">,</span> <span class="n">tfam</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">oldSamples</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the completion and concordance of each samples.</span>

<span class="sd">    :param tped: the ``tped`` containing duplicated samples.</span>
<span class="sd">    :param tfam: the ``tfam`` containing duplicated samples.</span>
<span class="sd">    :param samples: the updated position of the samples in the tped containing</span>
<span class="sd">                    only duplicated samples.</span>
<span class="sd">    :param oldSamples: the original duplicated sample positions.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>

<span class="sd">    :type tped: :py:class:`numpy.array`</span>
<span class="sd">    :type tfam: :py:class:`numpy.array`</span>
<span class="sd">    :type samples: dict</span>
<span class="sd">    :type oldSamples: dict</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    :returns: a tuple containing the completion (:py:class:`numpy.array`) as</span>
<span class="sd">              first element, and the concordance (:py:class:`dict`) as last</span>
<span class="sd">              element.</span>

<span class="sd">    Reads the ``tped`` file and compute the completion for each duplicated</span>
<span class="sd">    samples and the pairwise concordance between duplicated samples.</span>

<span class="sd">    .. note::</span>
<span class="sd">        The completion and concordance computation excludes a markers if it&#39;s</span>
<span class="sd">        on chromosome 24 and if the sample is a female.</span>

<span class="sd">    .. note::</span>
<span class="sd">        A missing genotype is encoded by ``0``.</span>

<span class="sd">    .. note::</span>
<span class="sd">        No percentage is computed here, only the numbers. Percentages are</span>
<span class="sd">        computing in other functions: :py:func:`printStatistics`, for</span>
<span class="sd">        completion, and :py:func:`printConcordance`, for concordance.</span>


<span class="sd">    **Completion**</span>

<span class="sd">    Computes the completion of none zero values (where all genotypes of at</span>
<span class="sd">    least one duplicated sample are no call [*i.e.* ``0``]). The completion of</span>
<span class="sd">    sample :math:`i` (*i.e.* :math:`Comp_i`) is the number of genotypes</span>
<span class="sd">    that have a call divided by the total number of genotypes (the set</span>
<span class="sd">    :math:`G_i`):</span>

<span class="sd">    .. math::</span>
<span class="sd">        Comp_i = \\frac{||g \\in G_i\\textrm{ where }g \\neq 0||}{||G_i||}</span>

<span class="sd">    .. note::</span>
<span class="sd">        We consider a genotype as being missing if the sample is a male and if</span>
<span class="sd">        a marker on chromosome 23 or 24 is heterozygous.</span>


<span class="sd">    **Concordance**</span>

<span class="sd">    Computes the pairwise concordance between duplicated samples. For each</span>
<span class="sd">    marker, if both genotypes are not missing, we add one to the total number</span>
<span class="sd">    of compared markers. If both genotypes are the same, we add one to the</span>
<span class="sd">    number of concordant calls. We write the observed genotype difference in</span>
<span class="sd">    the file ``prefix.diff``. The concordance between sample :math:`i` and</span>
<span class="sd">    :math:`j` (*i.e.* :math:`Concordance_{i,j}`) is the number of genotypes</span>
<span class="sd">    that are equal divided by the total number of genotypes (excluding the no</span>
<span class="sd">    calls):</span>

<span class="sd">    .. math::</span>
<span class="sd">        Concordance_{i,j} = \\frac{</span>
<span class="sd">            ||g \\in G_i \\cup G_j \\textrm{ where } g_i = g_j \\neq 0||</span>
<span class="sd">        }{</span>
<span class="sd">            ||g \\in G_i \\cup G_j \\textrm{ where } g \\neq 0||</span>
<span class="sd">        }</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The diff file containing the genotype differences between a pair of</span>
    <span class="c1"># duplicated samples</span>
    <span class="n">diffOutput</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">diffOutput</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.diff&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(prefix)s</span><span class="s2">.diff: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">diffOutput</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;famID&quot;</span><span class="p">,</span> <span class="s2">&quot;indID&quot;</span><span class="p">,</span> <span class="s2">&quot;dupIndex_1&quot;</span><span class="p">,</span>
                                   <span class="s2">&quot;dupIndex_2&quot;</span><span class="p">,</span> <span class="s2">&quot;genotype_1&quot;</span><span class="p">,</span> <span class="s2">&quot;genotype_2&quot;</span><span class="p">])</span>

    <span class="c1"># The completion data type</span>
    <span class="n">completion</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tped</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">4</span><span class="p">:]))],</span>
                           <span class="p">[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tped</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">4</span><span class="p">:]))]])</span>

    <span class="c1"># The concordance data type</span>
    <span class="n">concordance</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">sampleID</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">nbDup</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">])</span>
        <span class="n">concordance</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">np</span><span class="o">.</span><span class="n">asmatrix</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">nbDup</span><span class="p">,</span> <span class="n">nbDup</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)),</span>
            <span class="n">np</span><span class="o">.</span><span class="n">asmatrix</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">nbDup</span><span class="p">,</span> <span class="n">nbDup</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)),</span>
        <span class="p">]</span>

    <span class="c1">#####################################################################</span>
    <span class="c1"># Add options for concordance only for hetero SNPs... (all samples) #</span>
    <span class="c1">#####################################################################</span>

    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tped</span><span class="p">:</span>
        <span class="n">genotype</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>
        <span class="n">chromosome</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">snpName</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">indexes</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indexes</span><span class="p">):</span>
                <span class="n">sex</span> <span class="o">=</span> <span class="n">tfam</span><span class="p">[</span><span class="n">oldSamples</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">i</span><span class="p">]][</span><span class="mi">4</span><span class="p">]</span>
                <span class="n">genotype1</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">genotype</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">))</span>

                <span class="c1"># Updating the completion</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="p">((</span><span class="n">chromosome</span> <span class="o">==</span> <span class="s2">&quot;24&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">sex</span> <span class="o">==</span> <span class="s2">&quot;2&quot;</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">sex</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">chromosome</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;23&quot;</span><span class="p">,</span> <span class="s2">&quot;24&quot;</span><span class="p">]):</span>
                        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;0&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">genotype1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genotype1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">):</span>
                            <span class="n">completion</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">index</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">elif</span> <span class="s2">&quot;0&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">genotype1</span><span class="p">:</span>
                        <span class="n">completion</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">index</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">completion</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">index</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="c1"># Updating the concordance</span>
                    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">indexes</span><span class="p">)):</span>
                        <span class="n">genotype2</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">genotype</span><span class="p">[</span><span class="n">indexes</span><span class="p">[</span><span class="n">j</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">))</span>

<span class="c1"># This is for man on chr 23 and 24, if heterozygous!!! #</span>
<span class="c1"># if (sex == &quot;1&quot;) and (chromosome in [&quot;23&quot;, &quot;24&quot;]):</span>
<span class="c1">#     if (&quot;0&quot; not in genotype1) and (&quot;0&quot; not in genotype2):</span>
<span class="c1">#         if (len(genotype1) != 2) and (len(genotype2) != 2):</span>
<span class="c1">#             concordance[key][1][i,j] += 1</span>
<span class="c1">#             concordance[key][1][j,i] += 1</span>
<span class="c1">#             if genotype1 == genotype2:</span>
<span class="c1">#                 concordance[key][0][i,j] += 1</span>
<span class="c1">#                 concordance[key][0][j,i] += 1</span>
<span class="c1">#             else:</span>
<span class="c1">#                 # Print to diff file</span>
<span class="c1">#                 toPrint = [snpName, key[0], key[1],</span>
<span class="c1">#                            str(index+1),</span>
<span class="c1">#                            str(indexes[j]+1)]</span>
<span class="c1">#                 toPrint.append(&quot; &quot;.join(genotype1))</span>
<span class="c1">#                 if len(genotype1) == 1:</span>
<span class="c1">#                     toPrint[-1] += &quot; %s&quot; % toPrint[-1]</span>
<span class="c1">#                 toPrint.append(&quot; &quot;.join(genotype2))</span>
<span class="c1">#                 if len(genotype2) == 1:</span>
<span class="c1">#                     toPrint[-1] += &quot; %s&quot; % toPrint[-1]</span>
<span class="c1">#                 print &gt;&gt;diffOutput, &quot;\t&quot;.join(toPrint)</span>
<span class="c1">#</span>
<span class="c1"># elif (&quot;0&quot; not in genotype1) and (&quot;0&quot; not in genotype2):</span>

                        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;0&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">genotype1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;0&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">genotype2</span><span class="p">):</span>
                            <span class="c1"># Both have calls</span>
                            <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="k">if</span> <span class="n">genotype1</span> <span class="o">==</span> <span class="n">genotype2</span><span class="p">:</span>
                                <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                                <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="c1"># Print to diff file</span>
                                <span class="n">toPrint</span> <span class="o">=</span> <span class="p">[</span><span class="n">snpName</span><span class="p">,</span> <span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                           <span class="nb">str</span><span class="p">(</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">indexes</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
                                <span class="n">toPrint</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genotype1</span><span class="p">))</span>
                                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genotype1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                                    <span class="n">toPrint</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="s2">&quot; </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">toPrint</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                                <span class="n">toPrint</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genotype2</span><span class="p">))</span>
                                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genotype2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                                    <span class="n">toPrint</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="s2">&quot; </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">toPrint</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">diffOutput</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">toPrint</span><span class="p">)</span>

    <span class="n">diffOutput</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">concordance</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">])):</span>
            <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">concordance</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="n">completion</span><span class="p">,</span> <span class="n">concordance</span></div>


<div class="viewcode-block" id="processTPED"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.processTPED">[docs]</a><span class="k">def</span> <span class="nf">processTPED</span><span class="p">(</span><span class="n">uniqueSamples</span><span class="p">,</span> <span class="n">duplicatedSamples</span><span class="p">,</span> <span class="n">fileName</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Process the TPED file.</span>

<span class="sd">    :param uniqueSamples: the position of unique samples.</span>
<span class="sd">    :param duplicatedSamples: the position of duplicated samples.</span>
<span class="sd">    :param fileName: the name of the file.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>

<span class="sd">    :type uniqueSamples: dict</span>
<span class="sd">    :type duplicatedSamples: collections.defaultdict</span>
<span class="sd">    :type fileName: str</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    :returns: a tuple containing the ``tped`` (:py:class:`numpy.array`) as</span>
<span class="sd">              first element, and the updated positions of the duplicated</span>
<span class="sd">              samples (:py:class:`dict`)</span>

<span class="sd">    Reads the entire ``tped`` and prints another one containing only unique</span>
<span class="sd">    samples (``prefix.unique_samples.tped``). It then creates a</span>
<span class="sd">    :py:class:`numpy.array` containing the duplicated samples.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Getting the indexes</span>
    <span class="n">uI</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">uniqueSamples</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="n">dI</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">duplicatedSamples</span><span class="o">.</span><span class="n">itervalues</span><span class="p">():</span>
        <span class="n">dI</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
    <span class="n">dI</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

    <span class="n">tped</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">outputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">outputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.unique_samples.tped&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">: can&#39;t write to file&quot;</span> <span class="o">%</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.unique_samples.tped&quot;</span>
        <span class="k">raise</span> <span class="n">ProgramError</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fileName</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inputFile</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">snpInfo</span> <span class="o">=</span> <span class="n">row</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span>
            <span class="n">genotype</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>

            <span class="c1"># Printing the new TPED file (unique samples only)</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outputFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">snpInfo</span> <span class="o">+</span> <span class="p">[</span><span class="n">genotype</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">uI</span><span class="p">])</span>

            <span class="c1"># Saving the TPED file (duplicated samples only)</span>
            <span class="n">tped</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">snpInfo</span> <span class="o">+</span> <span class="p">[</span><span class="n">genotype</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">dI</span><span class="p">]))</span>

    <span class="c1"># Closing the output file</span>
    <span class="n">outputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Updating the positions</span>
    <span class="n">updatedSamples</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">sampleID</span> <span class="ow">in</span> <span class="n">duplicatedSamples</span><span class="p">:</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">duplicatedSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span>
        <span class="n">newPositions</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">positions</span><span class="p">):</span>
            <span class="n">newPositions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dI</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
        <span class="n">updatedSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span> <span class="o">=</span> <span class="n">newPositions</span>

    <span class="n">tped</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tped</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">tped</span><span class="p">,</span> <span class="n">updatedSamples</span></div>


<div class="viewcode-block" id="printUniqueTFAM"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.printUniqueTFAM">[docs]</a><span class="k">def</span> <span class="nf">printUniqueTFAM</span><span class="p">(</span><span class="n">tfam</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Prints a new TFAM with only unique samples.</span>

<span class="sd">    :param tfam: a representation of a TFAM file.</span>
<span class="sd">    :param samples: the position of the samples</span>
<span class="sd">    :param prefix: the prefix of the output file name</span>

<span class="sd">    :type tfam: list</span>
<span class="sd">    :type samples: dict</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fileName</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.unique_samples.tfam&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fileName</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outputFile</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outputFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tfam</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(fileName)s</span><span class="s2">: no such file&quot;</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="findDuplicates"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.findDuplicates">[docs]</a><span class="k">def</span> <span class="nf">findDuplicates</span><span class="p">(</span><span class="n">tfam</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Finds the duplicates in a TFAM.</span>

<span class="sd">    :param tfam: representation of a ``tfam`` file.</span>
<span class="sd">    :type tfam: list</span>

<span class="sd">    :returns: two :py:class:`dict`, containing unique and duplicated samples</span>
<span class="sd">              position.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">uSamples</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">dSamples</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tfam</span><span class="p">):</span>
        <span class="n">sampleID</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">row</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">sampleID</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">uSamples</span><span class="p">:</span>
            <span class="c1"># This is the first time we see this sample</span>
            <span class="n">uSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span> <span class="o">=</span> <span class="n">i</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># We have seen this sample at least once...</span>
            <span class="k">if</span> <span class="n">sampleID</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dSamples</span><span class="p">:</span>
                <span class="c1"># This is the second time we see this sample...</span>
                <span class="n">dSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">uSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">],</span> <span class="n">i</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># We have seen this sample multiple times</span>
                <span class="n">dSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="c1"># Removing the duplicates from the unique samples</span>
    <span class="k">for</span> <span class="n">sampleID</span> <span class="ow">in</span> <span class="n">dSamples</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">sampleID</span> <span class="ow">in</span> <span class="n">uSamples</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">uSamples</span><span class="p">[</span><span class="n">sampleID</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">uSamples</span><span class="p">,</span> <span class="n">dSamples</span></div>


<div class="viewcode-block" id="readTFAM"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.readTFAM">[docs]</a><span class="k">def</span> <span class="nf">readTFAM</span><span class="p">(</span><span class="n">fileName</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads the TFAM file.</span>

<span class="sd">    :param fileName: the name of the ``tfam`` file.</span>

<span class="sd">    :type fileName: str</span>

<span class="sd">    :returns: a list of tuples, representing the ``tfam`` file.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Saving the TFAM file</span>
    <span class="n">tfam</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fileName</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
        <span class="n">tfam</span> <span class="o">=</span> <span class="p">[</span>
            <span class="nb">tuple</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">inputFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="p">]</span>

    <span class="n">tfam</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tfam</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">tfam</span></div>


<div class="viewcode-block" id="checkArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.checkArgs">[docs]</a><span class="k">def</span> <span class="nf">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    :param args: a :py:class:`argparse.Namespace` object containing the options</span>
<span class="sd">                 of the program.</span>

<span class="sd">    :type args: :py:class:`argparse.Namespace`</span>

<span class="sd">    :returns: ``True`` if everything was OK.</span>

<span class="sd">    If there is a problem with an option, an exception is raised using the</span>
<span class="sd">    :py:class:`ProgramError` class, a message is printed to the</span>
<span class="sd">    :class:`sys.stderr` and the program exists with code 1.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Checking the input files</span>
    <span class="k">for</span> <span class="n">suffix</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;.tped&quot;</span><span class="p">,</span> <span class="s2">&quot;.tfam&quot;</span><span class="p">]:</span>
        <span class="n">fileName</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">tfile</span> <span class="o">+</span> <span class="n">suffix</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fileName</span><span class="p">):</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(fileName)s</span><span class="s2">: no such file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Checking the concordance threshold</span>
    <span class="k">if</span> <span class="p">((</span><span class="n">args</span><span class="o">.</span><span class="n">sample_concordance_threshold</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span>
            <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">sample_concordance_threshold</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;sample-concordance-threshold: must be between 0 and 1 &quot;</span> \
              <span class="s2">&quot;(not </span><span class="si">%f</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="n">args</span><span class="o">.</span><span class="n">sample_concordance_threshold</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Checking the completion threshold</span>
    <span class="k">if</span> <span class="p">((</span><span class="n">args</span><span class="o">.</span><span class="n">sample_completion_threshold</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span>
            <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">sample_completion_threshold</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;sample-completion-threshold: must be between 0 and 1 &quot;</span> \
              <span class="s2">&quot;(not </span><span class="si">%f</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="n">args</span><span class="o">.</span><span class="n">sample_completion_threshold</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">return</span> <span class="bp">True</span></div>


<div class="viewcode-block" id="parseArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.parseArgs">[docs]</a><span class="k">def</span> <span class="nf">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>  <span class="c1"># pragma: no cover</span>
    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    :param argString: the options</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    :returns: A :py:class:`argparse.Namespace` object created by</span>
<span class="sd">              the :py:mod:`argparse` module. It contains the values of the</span>
<span class="sd">              different options.</span>

<span class="sd">    ================================== ====== =================================</span>
<span class="sd">                  Options               Type             Description</span>
<span class="sd">    ================================== ====== =================================</span>
<span class="sd">    ``--tfile``                        string The input file prefix (of type</span>
<span class="sd">                                              ``tfile``).</span>
<span class="sd">    ``--sample-completion-threshold``  float  The completion threshold.</span>
<span class="sd">    ``--sample-concordance-threshold`` float  The concordance threshold.</span>
<span class="sd">    ``--out``                          string The prefix of the output files.</span>
<span class="sd">    ================================== ====== =================================</span>

<span class="sd">    .. note::</span>
<span class="sd">        No option check is done here (except for the one automatically done by</span>
<span class="sd">        argparse). Those need to be done elsewhere (see :py:func:`checkArgs`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">argString</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">args</span></div>


<div class="viewcode-block" id="ProgramError"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.ProgramError">[docs]</a><span class="k">class</span> <span class="nc">ProgramError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;An :py:class:`Exception` raised in case of a problem.</span>

<span class="sd">    :param msg: the message to print to the user before exiting.</span>

<span class="sd">    :type msg: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Construction of the :py:class:`ProgramError` class.</span>

<span class="sd">        :param msg: the message to print to the user</span>
<span class="sd">        :type msg: str</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">message</span></div>


<span class="c1"># The parser object</span>
<span class="n">pretty_name</span> <span class="o">=</span> <span class="s2">&quot;Duplicated samples&quot;</span>
<span class="n">desc</span> <span class="o">=</span> <span class="s2">&quot;Extracts and merges duplicated samples.&quot;</span>
<span class="n">long_desc</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;The script evaluates concordance and completion rate. If the &quot;</span>
             <span class="s2">&quot;thresholds are met, the script merges and completes the &quot;</span>
             <span class="s2">&quot;genotypes.&quot;</span><span class="p">)</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-v&quot;</span><span class="p">,</span> <span class="s2">&quot;--version&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;version&quot;</span><span class="p">,</span>
                    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;pyGenClean version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>

<span class="c1"># The INPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Input File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--tfile&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The input file prefix (will find the tped and tfam &quot;</span>
                         <span class="s2">&quot;file by appending the prefix to .tped and .tfam, &quot;</span>
                         <span class="s2">&quot;respectively.) The duplicated samples should have &quot;</span>
                         <span class="s2">&quot;the same identification numbers (both family and &quot;</span>
                         <span class="s2">&quot;individual ids.)&quot;</span><span class="p">))</span>
<span class="c1"># The options</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Options&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--sample-completion-threshold&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
                   <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FLOAT&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The completion threshold to consider a replicate &quot;</span>
                         <span class="s2">&quot;when choosing the best replicates and for creating &quot;</span>
                         <span class="s2">&quot;the composite samples. [default: </span><span class="si">%(default).1f</span><span class="s2">]&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--sample-concordance-threshold&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
                   <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FLOAT&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.97</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The concordance threshold to consider a replicate &quot;</span>
                         <span class="s2">&quot;when choosing the best replicates and for creating &quot;</span>
                         <span class="s2">&quot;the composite samples. [default: </span><span class="si">%(default).2f</span><span class="s2">]&quot;</span><span class="p">))</span>
<span class="c1"># The OUTPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Output File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">default</span><span class="o">=</span><span class="s2">&quot;dup_samples&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The prefix of the output files. [default: &quot;</span>
                         <span class="s2">&quot;</span><span class="si">%(default)s</span><span class="s2">]&quot;</span><span class="p">))</span>


<div class="viewcode-block" id="safe_main"><a class="viewcode-back" href="../../../module_content/pyGenClean.DupSamples.html#pyGenClean.DupSamples.duplicated_samples.safe_main">[docs]</a><span class="k">def</span> <span class="nf">safe_main</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;A safe version of the main function (that catches ProgramError).&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">main</span><span class="p">()</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span></div>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">safe_main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" >pyGenClean</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>