<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Ethnicity Module &mdash; pyGenClean 1.8.2 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.8.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.8.2 documentation" href="index.html" />
    <link rel="next" title="Minor Allele Frequency of Zero Module" href="flag_maf_zero.html" />
    <link rel="prev" title="Related Samples Module" href="find_related_samples.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="index.html"><img src="_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="flag_maf_zero.html" title="Minor Allele Frequency of Zero Module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="find_related_samples.html" title="Related Samples Module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Ethnicity Module</a><ul>
<li><a class="reference internal" href="#input-files">Input Files</a></li>
<li><a class="reference internal" href="#procedure">Procedure</a></li>
<li><a class="reference internal" href="#output-files">Output Files</a></li>
<li><a class="reference internal" href="#the-plots">The Plots</a><ul>
<li><a class="reference internal" href="#modifying-the-outlier-plot">Modifying The Outlier Plot</a></li>
</ul>
</li>
<li><a class="reference internal" href="#finding-outliers">Finding Outliers</a></li>
<li><a class="reference internal" href="#the-algorithm">The Algorithm</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="find_related_samples.html"
                        title="previous chapter">Related Samples Module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="flag_maf_zero.html"
                        title="next chapter">Minor Allele Frequency of Zero Module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/check_ethnicity.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="ethnicity-module">
<span id="ethnicity-module-label"></span><h1>Ethnicity Module<a class="headerlink" href="#ethnicity-module" title="Permalink to this headline">Â¶</a></h1>
<p>The usage of the standalone module is shown below:</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> pyGenClean_check_ethnicity --help
<span class="go">usage: pyGenClean_check_ethnicity [-h] [-v] --bfile FILE [--skip-ref-pops]</span>
<span class="go">                                  [--ceu-bfile FILE] [--yri-bfile FILE]</span>
<span class="go">                                  [--jpt-chb-bfile FILE] [--min-nb-snp INT]</span>
<span class="go">                                  [--indep-pairwise STR STR STR] [--maf FLOAT]</span>
<span class="go">                                  [--sge] [--sge-walltime TIME]</span>
<span class="go">                                  [--sge-nodes INT INT]</span>
<span class="go">                                  [--ibs-sge-walltime TIME]</span>
<span class="go">                                  [--ibs-sge-nodes INT INT]</span>
<span class="go">                                  [--line-per-file-for-sge INT]</span>
<span class="go">                                  [--nb-components INT] [--outliers-of POP]</span>
<span class="go">                                  [--multiplier FLOAT] [--xaxis COMPONENT]</span>
<span class="go">                                  [--yaxis COMPONENT] [--format FORMAT]</span>
<span class="go">                                  [--title STRING] [--xlabel STRING]</span>
<span class="go">                                  [--ylabel STRING] [--create-scree-plot]</span>
<span class="go">                                  [--scree-plot-title TITLE] [--out FILE]</span>

<span class="go">Checks samples&#39; ethnicity using reference populations and IBS.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>

<span class="go">Input File:</span>
<span class="go">  --bfile FILE          The input file prefix (will find the plink binary</span>
<span class="go">                        files by appending the prefix to the .bim, .bed and</span>
<span class="go">                        .fam files, respectively.</span>
<span class="go">  --skip-ref-pops       Perform the MDS computation, but skip the three</span>
<span class="go">                        reference panels.</span>
<span class="go">  --ceu-bfile FILE      The input file prefix (will find the plink binary</span>
<span class="go">                        files by appending the prefix to the .bim, .bed and</span>
<span class="go">                        .fam files, respectively.) for the CEU population</span>
<span class="go">  --yri-bfile FILE      The input file prefix (will find the plink binary</span>
<span class="go">                        files by appending the prefix to the .bim, .bed and</span>
<span class="go">                        .fam files, respectively.) for the YRI population</span>
<span class="go">  --jpt-chb-bfile FILE  The input file prefix (will find the plink binary</span>
<span class="go">                        files by appending the prefix to the .bim, .bed and</span>
<span class="go">                        .fam files, respectively.) for the JPT-CHB population</span>

<span class="go">Options:</span>
<span class="go">  --min-nb-snp INT      The minimum number of markers needed to compute IBS</span>
<span class="go">                        values. [Default: 8000]</span>
<span class="go">  --indep-pairwise STR STR STR</span>
<span class="go">                        Three numbers: window size, window shift and the r2</span>
<span class="go">                        threshold. [default: [&#39;50&#39;, &#39;5&#39;, &#39;0.1&#39;]]</span>
<span class="go">  --maf FLOAT           Restrict to SNPs with MAF &gt;= threshold. [default:</span>
<span class="go">                        0.05]</span>
<span class="go">  --sge                 Use SGE for parallelization.</span>
<span class="go">  --sge-walltime TIME   The walltime for the job to run on the cluster. Do not</span>
<span class="go">                        use if you are not required to specify a walltime for</span>
<span class="go">                        your jobs on your cluster (e.g. &#39;qsub</span>
<span class="go">                        -lwalltime=1:0:0&#39; on the cluster).</span>
<span class="go">  --sge-nodes INT INT   The number of nodes and the number of processor per</span>
<span class="go">                        nodes to use (e.g. &#39;qsub -lnodes=X:ppn=Y&#39; on the</span>
<span class="go">                        cluster, where X is the number of nodes and Y is the</span>
<span class="go">                        number of processor to use. Do not use if you are not</span>
<span class="go">                        required to specify the number of nodes for your jobs</span>
<span class="go">                        on the cluster.</span>
<span class="go">  --ibs-sge-walltime TIME</span>
<span class="go">                        The walltime for the IBS jobs to run on the cluster.</span>
<span class="go">                        Do not use if you are not required to specify a</span>
<span class="go">                        walltime for your jobs on your cluster (e.g. &#39;qsub</span>
<span class="go">                        -lwalltime=1:0:0&#39; on the cluster).</span>
<span class="go">  --ibs-sge-nodes INT INT</span>
<span class="go">                        The number of nodes and the number of processor per</span>
<span class="go">                        nodes to use for the IBS jobs (e.g. &#39;qsub</span>
<span class="go">                        -lnodes=X:ppn=Y&#39; on the cluster, where X is the number</span>
<span class="go">                        of nodes and Y is the number of processor to use. Do</span>
<span class="go">                        not use if you are not required to specify the number</span>
<span class="go">                        of nodes for your jobs on the cluster.</span>
<span class="go">  --line-per-file-for-sge INT</span>
<span class="go">                        The number of line per file for SGE task array for the</span>
<span class="go">                        IBS jobs. [default: 100]</span>
<span class="go">  --nb-components INT   The number of component to compute. [default: 10]</span>

<span class="go">Outlier Options:</span>
<span class="go">  --outliers-of POP     Finds the outliers of this population. [default: CEU]</span>
<span class="go">  --multiplier FLOAT    To find the outliers, we look for more than x times</span>
<span class="go">                        the cluster standard deviation. [default: 1.9]</span>
<span class="go">  --xaxis COMPONENT     The component to use for the X axis. [default: C1]</span>
<span class="go">  --yaxis COMPONENT     The component to use for the Y axis. [default: C2]</span>

<span class="go">MDS Plot Options:</span>
<span class="go">  --format FORMAT       The output file format (png, ps, pdf, or X11 formats</span>
<span class="go">                        are available). [default: png]</span>
<span class="go">  --title STRING        The title of the MDS plot. [default: C2 in function of</span>
<span class="go">                        C1 - MDS]</span>
<span class="go">  --xlabel STRING       The label of the X axis. [default: C1]</span>
<span class="go">  --ylabel STRING       The label of the Y axis. [default: C2]</span>

<span class="go">Scree Plot Options:</span>
<span class="go">  --create-scree-plot   Computes Eigenvalues and creates a scree plot.</span>
<span class="go">  --scree-plot-title TITLE</span>
<span class="go">                        The main title of the scree plot [EIGENSOFT results]</span>

<span class="go">Output File:</span>
<span class="go">  --out FILE            The prefix of the output files. [default: ethnicity]</span>
</pre></div>
</div>
<div class="section" id="input-files">
<h2>Input Files<a class="headerlink" href="#input-files" title="Permalink to this headline">Â¶</a></h2>
<p>This module uses PLINK&#8217;s binary file format (<code class="docutils literal"><span class="pre">bed</span></code>, <code class="docutils literal"><span class="pre">bim</span></code> and <code class="docutils literal"><span class="pre">fam</span></code> files)
for the source data set (the data of interest) and three sets of binary files
for the reference panels (CEU, YRI and JPG-CHB).</p>
</div>
<div class="section" id="procedure">
<h2>Procedure<a class="headerlink" href="#procedure" title="Permalink to this headline">Â¶</a></h2>
<p>Here are the steps performed by the module:</p>
<ol class="arabic simple">
<li>Finds the overlapping markers between the three reference panels and the
source panel.</li>
<li>Extracts the required markers from all the data sets (source and reference
panels).</li>
<li>Combines the three reference panels together to create a single data set.</li>
<li>Renames the reference panel&#8217;s markers so that they match the names of the
markers in the source panel.</li>
<li>Computes the frequency of all the markers from the reference and the source
panels.</li>
<li>Finds the markers to flip in the reference panel, to enable fast comparison
with the source panel.</li>
<li>Excludes markers that cannot be flip from the reference and the source
panels.</li>
<li>Flips the markers that need to be in the reference panel.</li>
<li>Combines the reference and the source panels.</li>
<li>Computes the IBS values (see Section <a class="reference internal" href="find_related_samples.html#related-samples-label"><span>Related Samples Module</span></a> for more
information).</li>
<li>Creates the MDS file from the combined data set and the IBS values.</li>
<li>Creates a population file for plotting purposes.</li>
<li>Plots the MDS values.</li>
<li>Finds the outliers of a given reference population (either CEU, YRI or
JPT-CHB).<ol class="lowerroman">
<li>Reads the population file.</li>
<li>Reads the MDS values.</li>
<li>Computes the three reference population clusters&#8217; center.</li>
<li>Computes three clusters according to the reference population clusters&#8217;
centers, and finds the outliers of a given reference population.</li>
<li>Writes the outliers in a file.</li>
</ol>
</li>
</ol>
</div>
<div class="section" id="output-files">
<h2>Output Files<a class="headerlink" href="#output-files" title="Permalink to this headline">Â¶</a></h2>
<p>The output files of each of the steps described above are as follow (note that
the output prefix shown is the one by default [<em>i.e</em> <code class="docutils literal"><span class="pre">ethnicity</span></code>]):</p>
<ol class="arabic simple">
<li>Three files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.ref_snp_to_extract</span></code>: a list of markers to extract from the
reference panels.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.source_snp_to_extract</span></code>: a list of markers to extract from
the source panel.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.update_names</span></code>: the updated names of the marker in the
reference panels, so that they match with the names in the source panel.</li>
</ul>
</li>
<li>Four sets of PLINK&#8217;s binary files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.CEU</span></code>: the data set containing the extracted
markers from the CEU reference population.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.YRI</span></code>: the data set containing the extracted
markers from the YRI reference population.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.JPT-CHB</span></code>: the data set containing the
extracted markers from the JPG-CHB reference population.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.source_panel.ALL</span></code>: the data set containing the extracted
markers from the source population.</li>
</ul>
</li>
<li>One required file and one set of PLINK&#8217;s binary files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.ALL.files_to_merge</span></code>: the file required by
Plink to merge more than two data sets together.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.ALL</span></code>: the data set containing the merged
data sets of the three reference population.</li>
</ul>
</li>
<li>One set of PLINK&#8217;s binary files is created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.ALL.rename</span></code>: the data set after markers
have been renamed in the reference panels.</li>
</ul>
</li>
<li>Two sets of PLINK&#8217;s result files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.ALL.rename.frequency</span></code>: the frequencies of
the markers in the reference panels.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.source_panel.ALL.frequency</span></code>: the frequencies of the markers
in the source panels.</li>
</ul>
</li>
<li>Two files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.snp_to_flip_in_reference</span></code>: the list of markers to flip in
the reference panels.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.snp_to_remove</span></code>: the list of markers to remove because they
are not comparable to the markers in the source panel, even after trying
to flip them.</li>
</ul>
</li>
<li>Two sets of PLINK&#8217;s binary files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.ALL.rename.cleaned</span></code>: the data set after the
markers found in the previous step are excluded from the reference panels.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.source_panel.ALL.cleaned</span></code>: the data set after the markers
found in the previous step are excluded from the source panel.</li>
</ul>
</li>
<li>One set of PLINK&#8217;s binary files is created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.reference_panel.ALL.rename.cleaned.flipped</span></code>: the data set
after markers from the reference panels were flipped so that they become
comparable with the source panel.</li>
</ul>
</li>
<li>One required file and one set of PLINK&#8217;s binary files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.final_dataset_for_genome.files_to_merge</span></code>: the file required
by Plink to merge more than two data sets together.</li>
<li><code class="docutils literal"><span class="pre">ethnicity.final_dataset_for_genome</span></code>: the data set containing the
merged reference and source panels.</li>
</ul>
</li>
<li>Multiple files are created after this step.<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.ibs</span></code> : for more information about those files, see Section
<a class="reference internal" href="find_related_samples.html#related-samples-label"><span>Related Samples Module</span></a>.</li>
</ul>
</li>
<li>One set of PLINK&#8217;s result files is created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.mds</span></code>: files containing the MDS values.</li>
</ul>
</li>
<li>One file is created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.population_file</span></code>: the population file required for MDS
value plotting.</li>
</ul>
</li>
<li>One file is created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.mds.png</span></code>: the plot of the MDS values (see Figure
<a class="reference internal" href="#mds-plot-figure"><span>Initial MDS plot</span></a>).</li>
</ul>
</li>
<li>Four files are created:<ul>
<li><code class="docutils literal"><span class="pre">ethnicity.before.png</span></code>: the MDS values before outliers detection (see
Figure <a class="reference internal" href="#mds-plot-before-figure"><span>MDS plot before outlier detection</span></a>).</li>
<li><code class="docutils literal"><span class="pre">ethnicity.after.png</span></code>: the MDS values after outliers detection for
each of the three reference populations. The shaded points are the
outliers (see Figure <a class="reference internal" href="#mds-plot-after-figure"><span>MDS plot after outlier detection</span></a>).</li>
<li><code class="docutils literal"><span class="pre">ethnicity.outliers.png</span></code>: the MDS values after outliers detection for
the selected reference population (default is CEU) (see Figure
<a class="reference internal" href="#ethnicity-outliers-figure"><span>Ethnic outliers</span></a>).</li>
<li><code class="docutils literal"><span class="pre">ethnicity.outliers</span></code>: the list of outliers (excluding the reference
populations).</li>
<li><code class="docutils literal"><span class="pre">ethnicity.population_file_outliers</span></code>: a population file containing the
outliers (to help creating a new MDS plot using
<a class="reference internal" href="module_content/pyGenClean.PlinkUtils.html#module-pyGenClean.PlinkUtils.plot_MDS_standalone" title="pyGenClean.PlinkUtils.plot_MDS_standalone"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.PlinkUtils.plot_MDS_standalone</span></code></a>).</li>
</ul>
</li>
</ol>
</div>
<div class="section" id="the-plots">
<span id="ethnicity-plots-label"></span><h2>The Plots<a class="headerlink" href="#the-plots" title="Permalink to this headline">Â¶</a></h2>
<p>Multiple plots are created by this module. The first one (Figure
<a class="reference internal" href="#mds-plot-figure"><span>Initial MDS plot</span></a>) is the MDS values right after they are computed by
Plink. There is one color per reference populations (CEU in blue, YRI in green
and JPT-CHB in purple). The source population is represented as red crosses.</p>
<div class="figure align-center" id="id1">
<span id="mds-plot-figure"></span><a class="reference internal image-reference" href="_images/ethnicity_mds.png"><img alt="Initial MDS plot" src="_images/ethnicity_mds.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-text">Initial MDS plot</span></p>
</div>
<p>The second one (Figure <a class="reference internal" href="#mds-plot-before-figure"><span>MDS plot before outlier detection</span></a>) is the MDS values before
outlier detection. Points in red, green and blue represent the individuals part
of the CEU, YRI and JPT-CHB clusters, respectively. The yellow points represent
the center of each of the cluster, when only considering the three reference
panels.</p>
<div class="figure align-center" id="id2">
<span id="mds-plot-before-figure"></span><a class="reference internal image-reference" href="_images/ethnicity_before.png"><img alt="MDS plot before" src="_images/ethnicity_before.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-text">MDS plot before outlier detection</span></p>
</div>
<p>The third plot (Figure <a class="reference internal" href="#mds-plot-after-figure"><span>MDS plot after outlier detection</span></a>) is the MDS values after
outlier detection. Points in red, green and blue represent the individuals part
of the CEU, YRI and JPT-CHB clusters, respectively. Outliers are found for each
of the three reference populations and they are represented with the same, but
lighter color. Once again, the yellow points represent the center of each of the
cluster, when only considering the three reference panels.</p>
<div class="figure align-center" id="id3">
<span id="mds-plot-after-figure"></span><a class="reference internal image-reference" href="_images/ethnicity_after.png"><img alt="MDS plot after" src="_images/ethnicity_after.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-text">MDS plot after outlier detection</span></p>
</div>
<p>The last plot (Figure <a class="reference internal" href="#ethnicity-outliers-figure"><span>Ethnic outliers</span></a>) shows the outliers of
the selected reference population (CEU by default). Red, green and blue
represent the CEU, YRI and JPT-CHB samples, respectively. Orange represents the
individuals from the source panel who are part of the selected reference
population. Gray represents the outliers of the selected reference population.</p>
<div class="figure align-center" id="id4">
<span id="ethnicity-outliers-figure"></span><a class="reference internal image-reference" href="_images/ethnicity_outliers.png"><img alt="Ethnic outliers" src="_images/ethnicity_outliers.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-text">Ethnic outliers</span></p>
</div>
<div class="section" id="modifying-the-outlier-plot">
<h3>Modifying The Outlier Plot<a class="headerlink" href="#modifying-the-outlier-plot" title="Permalink to this headline">Â¶</a></h3>
<p>If you want to manually modify the above figures, have a look at the
<a class="reference internal" href="module_content/pyGenClean.PlinkUtils.html#module-pyGenClean.PlinkUtils.plot_MDS_standalone" title="pyGenClean.PlinkUtils.plot_MDS_standalone"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.PlinkUtils.plot_MDS_standalone</span></code></a> module. Here is the usage
of this script:</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> pyGenClean_plot_MDS --help
<span class="go">usage: pyGenClean_plot_MDS [-h] [-v] --file FILE --population-file FORMAT</span>
<span class="go">                           [--population-order STRING]</span>
<span class="go">                           [--population-colors STRING]</span>
<span class="go">                           [--population-sizes STRING]</span>
<span class="go">                           [--population-markers STRING]</span>
<span class="go">                           [--population-alpha STRING] [--format FORMAT]</span>
<span class="go">                           [--title STRING] [--xaxis STRING] [--xlabel STRING]</span>
<span class="go">                           [--yaxis STRING] [--ylabel STRING]</span>
<span class="go">                           [--legend-position STRING] [--legend-size INT]</span>
<span class="go">                           [--legend-ncol INT] [--legend-alpha FLOAT]</span>
<span class="go">                           [--title-fontsize INT] [--label-fontsize INT]</span>
<span class="go">                           [--axis-fontsize INT] [--adjust-left FLOAT]</span>
<span class="go">                           [--adjust-right FLOAT] [--adjust-top FLOAT]</span>
<span class="go">                           [--adjust-bottom FLOAT] [--out FILE]</span>

<span class="go">Creates a MDS plot.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>

<span class="go">Input File:</span>
<span class="go">  --file FILE           The MBS file.</span>
<span class="go">  --population-file FORMAT</span>
<span class="go">                        A file containing population information. There must</span>
<span class="go">                        be three columns: famID, indID and population</span>
<span class="go">                        information.</span>

<span class="go">Population Properties:</span>
<span class="go">  --population-order STRING</span>
<span class="go">                        The order to print the different populations.</span>
<span class="go">                        [default: CEU,YRI,JPT-CHB,SOURCE,OUTLIER]</span>
<span class="go">  --population-colors STRING</span>
<span class="go">                        The population point color in the plot [default:</span>
<span class="go">                        377eb8,4daf4a,984ea3,e41a1c,ff7f00]</span>
<span class="go">  --population-sizes STRING</span>
<span class="go">                        The population point size in the plot. [default:</span>
<span class="go">                        12,12,12,8,3]</span>
<span class="go">  --population-markers STRING</span>
<span class="go">                        The population point marker in the plot. [default:</span>
<span class="go">                        .,.,.,+,D]</span>
<span class="go">  --population-alpha STRING</span>
<span class="go">                        The population alpha value in the plot. [default:</span>
<span class="go">                        1.0,1.0,1.0,1.0,1.0]</span>

<span class="go">Graphical Properties:</span>
<span class="go">  --format FORMAT       The output file format (png, ps, pdf, or X11 formats</span>
<span class="go">                        are available). [default: png]</span>
<span class="go">  --title STRING        The title of the MDS plot. [default: C2 in function of</span>
<span class="go">                        C1 - MDS]</span>
<span class="go">  --xaxis STRING        The component to print on the X axis. [default: C1]</span>
<span class="go">  --xlabel STRING       The label of the X axis. [default: C1]</span>
<span class="go">  --yaxis STRING        The component to print on the Y axis. [default: C2]</span>
<span class="go">  --ylabel STRING       The label of the Y axis. [default: C2]</span>
<span class="go">  --legend-position STRING</span>
<span class="go">                        The position of the legend. [default: best]</span>
<span class="go">  --legend-size INT     The size of the legend. [default: 10]</span>
<span class="go">  --legend-ncol INT     The number of column for the legend. [default: 1]</span>
<span class="go">  --legend-alpha FLOAT  The alpha value of the legend frame. [default: 1.0]</span>
<span class="go">  --title-fontsize INT  The font size of the title. [default: 15]</span>
<span class="go">  --label-fontsize INT  The font size of the X and Y labels. [default: 12]</span>
<span class="go">  --axis-fontsize INT   The font size of the X and Y axis. [Default: 12]</span>
<span class="go">  --adjust-left FLOAT   Adjust the left margin. [Default: 0.12]</span>
<span class="go">  --adjust-right FLOAT  Adjust the right margin. [Default: 0.90]</span>
<span class="go">  --adjust-top FLOAT    Adjust the top margin. [Default: 0.90]</span>
<span class="go">  --adjust-bottom FLOAT</span>
<span class="go">                        Adjust the bottom margin. [Default: 0.10]</span>

<span class="go">Output File:</span>
<span class="go">  --out FILE            The prefix of the output files. [default: mds]</span>
</pre></div>
</div>
<p>And here is an example of usage (for a MDS and a population file named
<code class="docutils literal"><span class="pre">ethnicity.mds.mds</span></code> and <code class="docutils literal"><span class="pre">ethnicity.population_file_outliers</span></code>, respectively),
producing the Figure <a class="reference internal" href="#ethnicity-outliers-modified-figure"><span>Ethnic outliers modified</span></a>.</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> pyGenClean_plot_MDS <span class="se">\</span>
<span class="gp">&gt;</span>     --file ethnicity.mds.mds <span class="se">\</span>
<span class="gp">&gt;</span>     --population-file ethnicity.population_file_outliers <span class="se">\</span>
<span class="gp">&gt;</span>     --population-order SOURCE,CEU,YRI,JPT-CHB,OUTLIER <span class="se">\</span>
<span class="gp">&gt;</span>     --population-colors e41a1c,377eb8,4daf4a,984ea3,000000 <span class="se">\</span>
<span class="gp">&gt;</span>     --population-markers .,.,.,.,+ <span class="se">\</span>
<span class="gp">&gt;</span>     --population-sizes 8,8,8,8,8 <span class="se">\</span>
<span class="gp">&gt;</span>     --axis-fontsize <span class="m">18</span> <span class="se">\</span>
<span class="gp">&gt;</span>     --label-fontsize <span class="m">18</span> <span class="se">\</span>
<span class="gp">&gt;</span>     --title-fontsize <span class="m">24</span> <span class="se">\</span>
<span class="gp">&gt;</span>     --adjust-bottom 0.11 <span class="se">\</span>
<span class="gp">&gt;</span>     --adjust-left 0.15 <span class="se">\</span>
<span class="gp">&gt;</span>     --adjust-right 0.96 <span class="se">\</span>
<span class="gp">&gt;</span>     --legend-size <span class="m">14</span> <span class="se">\</span>
<span class="gp">&gt;</span>     --legend-position lower-right
</pre></div>
</div>
<div class="figure align-center" id="id5">
<span id="ethnicity-outliers-modified-figure"></span><a class="reference internal image-reference" href="_images/ethnicity_outliers_modified.png"><img alt="Ethnic outliers modified" src="_images/ethnicity_outliers_modified.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-text">Ethnic outliers modified</span></p>
</div>
</div>
</div>
<div class="section" id="finding-outliers">
<span id="ethnicity-find-outliers"></span><h2>Finding Outliers<a class="headerlink" href="#finding-outliers" title="Permalink to this headline">Â¶</a></h2>
<p>If the multiplier of the cluster standard deviation was too stringent (or not
stringent enough), there is no need to run the module from the start. A
standalone script was created for this exact purpose, and it will find the
outliers using the <code class="docutils literal"><span class="pre">MDS</span></code> and population file previously created. Just modify
the <code class="docutils literal"><span class="pre">--multiplier</span></code> option and restart the analysis (which takes about a couple
of seconds).</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> pyGenClean_find_outliers --help
<span class="go">usage: pyGenClean_find_outliers [-h] [-v] --mds FILE --population-file FILE</span>
<span class="go">                                [--outliers-of POP] [--multiplier FLOAT]</span>
<span class="go">                                [--xaxis COMPONENT] [--yaxis COMPONENT]</span>
<span class="go">                                [--format FORMAT] [--out FILE]</span>

<span class="go">Finds outliers in SOURCE from CEU samples.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>

<span class="go">Input File:</span>
<span class="go">  --mds FILE            The MDS file from Plink</span>
<span class="go">  --population-file FILE</span>
<span class="go">                        A population file containing the following columns</span>
<span class="go">                        (without a header): FID, IID and POP. POP should be</span>
<span class="go">                        one of &#39;CEU&#39;, &#39;JPT-CHB&#39;, &#39;YRI&#39; and SOURCE.</span>

<span class="go">Options:</span>
<span class="go">  --outliers-of POP     Finds the outliers of this population. [default: CEU]</span>
<span class="go">  --multiplier FLOAT    To find the outliers, we look for more than x times</span>
<span class="go">                        the cluster standard deviation. [default: 1.9]</span>
<span class="go">  --xaxis COMPONENT     The component to use for the X axis. [default: C1]</span>
<span class="go">  --yaxis COMPONENT     The component to use for the Y axis. [default: C2]</span>
<span class="go">  --format FORMAT       The output file format (png, ps, or pdf formats are</span>
<span class="go">                        available). [default: png]</span>

<span class="go">Output File:</span>
<span class="go">  --out FILE            The prefix of the output files. [default: ethnicity]</span>
</pre></div>
</div>
</div>
<div class="section" id="the-algorithm">
<h2>The Algorithm<a class="headerlink" href="#the-algorithm" title="Permalink to this headline">Â¶</a></h2>
<p>For more information about the actual algorithms and source codes, refer to the
following pages.</p>
<ul class="simple">
<li><a class="reference internal" href="module_content/pyGenClean.Ethnicity.html#module-pyGenClean.Ethnicity.check_ethnicity" title="pyGenClean.Ethnicity.check_ethnicity"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.Ethnicity.check_ethnicity</span></code></a></li>
<li><a class="reference internal" href="module_content/pyGenClean.Ethnicity.html#module-pyGenClean.Ethnicity.find_outliers" title="pyGenClean.Ethnicity.find_outliers"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.Ethnicity.find_outliers</span></code></a></li>
<li><a class="reference internal" href="module_content/pyGenClean.PlinkUtils.html#module-pyGenClean.PlinkUtils.plot_MDS_standalone" title="pyGenClean.PlinkUtils.plot_MDS_standalone"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.PlinkUtils.plot_MDS_standalone</span></code></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="flag_maf_zero.html" title="Minor Allele Frequency of Zero Module"
             >next</a> |</li>
        <li class="right" >
          <a href="find_related_samples.html" title="Related Samples Module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>