<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Duplicated Markers Module &mdash; pyGenClean 1.8.1 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.8.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.8.1 documentation" href="index.html" />
    <link rel="next" title="Clean No Call and Only Heterozygous Markers Module" href="clean_noCall_hetero_snps.html" />
    <link rel="prev" title="Duplicated Samples Module" href="duplicated_samples.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="index.html"><img src="_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="clean_noCall_hetero_snps.html" title="Clean No Call and Only Heterozygous Markers Module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="duplicated_samples.html" title="Duplicated Samples Module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.1 documentation</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Duplicated Markers Module</a><ul>
<li><a class="reference internal" href="#input-files">Input Files</a></li>
<li><a class="reference internal" href="#procedure">Procedure</a></li>
<li><a class="reference internal" href="#output-files">Output Files</a></li>
<li><a class="reference internal" href="#the-algorithm">The Algorithm</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="duplicated_samples.html"
                        title="previous chapter">Duplicated Samples Module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="clean_noCall_hetero_snps.html"
                        title="next chapter">Clean No Call and Only Heterozygous Markers Module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/duplicated_markers.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="duplicated-markers-module">
<span id="dup-markers-label"></span><h1>Duplicated Markers Module<a class="headerlink" href="#duplicated-markers-module" title="Permalink to this headline">¶</a></h1>
<p>The usage of the standalone module is shown below:</p>
<div class="highlight-console"><div class="highlight"><pre><span class="gp">$</span> pyGenClean_duplicated_snps --help
<span class="go">usage: pyGenClean_duplicated_snps [-h] [-v] --tfile FILE</span>
<span class="go">                                  [--snp-completion-threshold FLOAT]</span>
<span class="go">                                  [--snp-concordance-threshold FLOAT]</span>
<span class="go">                                  [--frequency_difference FLOAT] [--out FILE]</span>

<span class="go">Extracts and merges duplicated markers.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>

<span class="go">Input File:</span>
<span class="go">  --tfile FILE          The input file prefix (will find the tped and tfam</span>
<span class="go">                        file by appending the prefix to .tped and .tfam,</span>
<span class="go">                        respectively. A .map file is also required.</span>

<span class="go">Options:</span>
<span class="go">  --snp-completion-threshold FLOAT</span>
<span class="go">                        The completion threshold to consider a replicate when</span>
<span class="go">                        choosing the best replicates and for composite</span>
<span class="go">                        creation. [default: 0.9]</span>
<span class="go">  --snp-concordance-threshold FLOAT</span>
<span class="go">                        The concordance threshold to consider a replicate when</span>
<span class="go">                        choosing the best replicates and for composite</span>
<span class="go">                        creation. [default: 0.98]</span>
<span class="go">  --frequency_difference FLOAT</span>
<span class="go">                        The maximum difference in frequency between duplicated</span>
<span class="go">                        markers [default: 0.05]</span>

<span class="go">Output File:</span>
<span class="go">  --out FILE            The prefix of the output files. [default: dup_snps]</span>
</pre></div>
</div>
<div class="section" id="input-files">
<h2>Input Files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h2>
<p>This module uses PLINK&#8217;s transposed pedfile format (<code class="docutils literal"><span class="pre">tped</span></code> and <code class="docutils literal"><span class="pre">tfam</span></code>
files). It also requires a <code class="docutils literal"><span class="pre">map</span></code> file to speed up the process of finding the
duplicated markers, so that the <code class="docutils literal"><span class="pre">tped</span></code> file is not read.</p>
</div>
<div class="section" id="procedure">
<h2>Procedure<a class="headerlink" href="#procedure" title="Permalink to this headline">¶</a></h2>
<p>Here are the steps performed by the module:</p>
<ol class="arabic simple">
<li>Reads the <code class="docutils literal"><span class="pre">map</span></code> file to gather marker&#8217;s position.</li>
<li>Reads the <code class="docutils literal"><span class="pre">tfam</span></code> file.</li>
<li>Finds the unique markers using the <code class="docutils literal"><span class="pre">map</span></code> file.</li>
<li>Process the <code class="docutils literal"><span class="pre">tped</span></code> file, finding unique and duplicated markers according
to chromosomal positions.</li>
<li>If there are no duplicated markers, stop here.</li>
<li>If there are duplicated markers, print a <code class="docutils literal"><span class="pre">tped</span></code> and <code class="docutils literal"><span class="pre">tfam</span></code> file
containing the duplicated markers.</li>
<li>Computes the frequency of the duplicated markers (using Plink) and read
the output file.</li>
<li>Computes the concordance and pairwise completion of each of the
duplicated markers.</li>
<li>Prints the problematic duplicated markers with a file containing the
summary of the statistics (completion and pairwise concordance).</li>
<li>Print the pairwise concordance in a file (matrices).</li>
<li>Choose the best duplicated markers using concordance and completion.</li>
<li>Completes the chosen markers with the remaining duplicated markers.</li>
<li>Creates the final <code class="docutils literal"><span class="pre">tped</span></code> file, containing the unique markers, the
chosen duplicated markers that were completed, and the problematic
duplicated markers (for further analysis). This set excludes markers
that were used for completing the chosen ones.</li>
</ol>
</div>
<div class="section" id="output-files">
<h2>Output Files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The output files of each of the steps described above are as follow (note that
the output prefix shown is the one by default [<em>i.e.</em> dup_snps]):</p>
<ol class="arabic simple">
<li>If the marker names are not unique, one file is created:<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.duplicated_marker_names</span></code>: a list of marker names and
chromosomal positions for each marker with duplicated names. This file
is not created if there are no markers with duplicated names.</li>
</ul>
</li>
<li>No files are created.</li>
<li>No files are created.</li>
<li>One set of transposed pedfiles.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.unique_snps</span></code>: the transposed pedfiles containing the unique
markers (according to chromosomal positions).</li>
</ul>
</li>
<li>If there are no duplicated markers (according to chromosomal positions), the
transposed pedfiles created at the previous step are copied to a new set of
transposed pedfiles.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.final</span></code>: the final transposed pedfiles.</li>
</ul>
</li>
<li>One set of transposed pedfiles.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.duplicated_snps</span></code>: the transposed pedfiles containing the
duplicated markers (according to chromosomal positions).</li>
</ul>
</li>
<li>One set of PLINK&#8217;s result file.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.duplicated_snps</span></code>: the file with the <code class="docutils literal"><span class="pre">frq</span></code> extension
contains the frequency of each duplicated markers.</li>
</ul>
</li>
<li>No files are created.</li>
<li>Multiple files are created.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.summary</span></code>: contains the completion and pairwise concordance
between duplicated markers.</li>
<li><code class="docutils literal"><span class="pre">dup_snps.problems</span></code>: contains the list of markers with &#8220;problems&#8221; that
can&#8217;t be used for further completion of the duplicated markers.
(either a difference in MAF [<code class="docutils literal"><span class="pre">diff_frequency</span></code>], a difference in the
minor allele [<code class="docutils literal"><span class="pre">diff_minor_allele</span></code>], two homozygous markers where one
is flipped [<code class="docutils literal"><span class="pre">homo_flip</span></code>], markers with flipped alleles [<code class="docutils literal"><span class="pre">flip</span></code>], one
marker is homozygous, the other is heterozygous [<code class="docutils literal"><span class="pre">homo_hetero</span></code>], one
marker is homozygous, the other is heterozygous but one is flipped
[<code class="docutils literal"><span class="pre">homo_hetero_flip</span></code>] or any other problem [<code class="docutils literal"><span class="pre">problem</span></code>].</li>
</ul>
</li>
<li>One output file is created.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.concordance</span></code>: a matrix containing a pairwise concordance
comparison for each duplicated markers.</li>
</ul>
</li>
<li>Two output files are created.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.chosen_snps.info</span></code>: the list of duplicated markers that were
chosen for completion with the other markers (the best of the duplicated
markers, according to concordance and completion).</li>
<li><code class="docutils literal"><span class="pre">dup_snps.not_chosen_snps.info</span></code>: the list of duplicated markers that
were not chosen for completion with the other markers.</li>
</ul>
</li>
<li>Multiple output files are created along with a set of transposed pedfiles.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.zeroed_out</span></code>: the list of genotypes that were zeroed out
while completing the chosen duplicated markers with the others. Each
line contains the id of the sample and the name of the marker that was
zeroed out.</li>
<li><code class="docutils literal"><span class="pre">dup_snps.not_good_enough</span></code>: the list of markers that were not good
enough (according to concordance and completion) to complete the best of
the duplicated markers.</li>
<li><code class="docutils literal"><span class="pre">dup_snps.removed_duplicates</span></code>: the list of markers that were used to
complete the chosen duplicated markers. Those markers were removed from
the dataset.</li>
<li><code class="docutils literal"><span class="pre">dup_snps.chosen_snps</span></code>: the transposed pedfiles containing the
completed chosen duplicated markers (a composite of all the duplicated
markers that were good enough).</li>
</ul>
</li>
<li>On set of transposed pedfiles.<ul>
<li><code class="docutils literal"><span class="pre">dup_snps.final</span></code>: the final dataset, containing the unique markers,
the chosen duplicated markers that were complete (composite) and the
duplicated markers that weren&#8217;t completed because of various problems.</li>
</ul>
</li>
</ol>
</div>
<div class="section" id="the-algorithm">
<h2>The Algorithm<a class="headerlink" href="#the-algorithm" title="Permalink to this headline">¶</a></h2>
<p>For more information about the actual algorithms and source codes, refer to the
following page.</p>
<ul class="simple">
<li><a class="reference internal" href="module_content/pyGenClean.DupSNPs.html#module-pyGenClean.DupSNPs.duplicated_snps" title="pyGenClean.DupSNPs.duplicated_snps"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.DupSNPs.duplicated_snps</span></code></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="clean_noCall_hetero_snps.html" title="Clean No Call and Only Heterozygous Markers Module"
             >next</a> |</li>
        <li class="right" >
          <a href="duplicated_samples.html" title="Duplicated Samples Module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>