
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Index &mdash; pyGenClean 1.7.0 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.7.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.7.0 documentation" href="index.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="index.html"><img src="_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.7.0 documentation</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">

   

<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#A"><strong>A</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#E"><strong>E</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#K"><strong>K</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 
</div>
<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.add_custom_options">add_custom_options() (in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.addToTPEDandTFAM">addToTPEDandTFAM() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.all_files_exist">all_files_exist() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.allFileExists">allFileExists() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.check_args">check_args() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.check_fam_for_samples">check_fam_for_samples() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.check_file_names">check_file_names() (in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.check_input_files">check_input_files() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.checkArgs">checkArgs() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.checkArgs">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.checkArgs">(in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.checkArgs">(in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

        
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.checkArgs">(in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

        
  <dt><a href="flag_maf_zero.html#pyGenClean.FlagMAF.flag_maf_zero.checkArgs">(in module pyGenClean.FlagMAF.flag_maf_zero)</a>
  </dt>

        
  <dt><a href="remove_heterozygous_haploid.html#pyGenClean.HeteroHap.remove_heterozygous_haploid.checkArgs">(in module pyGenClean.HeteroHap.remove_heterozygous_haploid)</a>
  </dt>

        
  <dt><a href="marker_missingness.html#pyGenClean.MarkerMissingness.snp_missingness.checkArgs">(in module pyGenClean.MarkerMissingness.snp_missingness)</a>
  </dt>

        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.checkArgs">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.clean_noCall_hetero_snps.checkArgs">(in module pyGenClean.NoCallHetero.clean_noCall_hetero_snps)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.checkArgs">(in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

        
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.checkArgs">(in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.compare_bim.checkArgs">(in module pyGenClean.PlinkUtils.compare_bim)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.checkArgs">(in module pyGenClean.PlinkUtils.plot_MDS_standalone)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.subset_data.checkArgs">(in module pyGenClean.PlinkUtils.subset_data)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.checkArgs">(in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.merge_related_samples.checkArgs">(in module pyGenClean.RelatedSamples.merge_related_samples)</a>
  </dt>

        
  <dt><a href="sample_missingness.html#pyGenClean.SampleMissingness.sample_missingness.checkArgs">(in module pyGenClean.SampleMissingness.sample_missingness)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.checkArgs">(in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.checkArgs">(in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.checkArgs">(in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.checkBim">checkBim() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.checkNumberOfSNP">checkNumberOfSNP() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.chooseBestDuplicates">chooseBestDuplicates() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.chooseBestSnps">chooseBestSnps() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.combinePlinkBinaryFiles">combinePlinkBinaryFiles() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      
  <dt><a href="marker_missingness.html#pyGenClean.MarkerMissingness.snp_missingness.compareBIM">compareBIM() (in module pyGenClean.MarkerMissingness.snp_missingness)</a>
  </dt>

      
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.compareBIMfiles">compareBIMfiles() (in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

      
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.compare_bim.compareSNPs">compareSNPs() (in module pyGenClean.PlinkUtils.compare_bim)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.compute_eigenvalues">compute_eigenvalues() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.compute_heterozygosity">compute_heterozygosity() (in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

      
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.compute_nb_samples">compute_nb_samples() (in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.compute_statistics">compute_statistics() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.computeFrequency">computeFrequency() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.computeFrequency">(in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

        
  <dt><a href="flag_maf_zero.html#pyGenClean.FlagMAF.flag_maf_zero.computeFrequency">(in module pyGenClean.FlagMAF.flag_maf_zero)</a>
  </dt>

        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.computeFrequency">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      </dl></dd>
  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.computeFrequencyOfSignificantSNPs">computeFrequencyOfSignificantSNPs() (in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.computeHeteroPercentage">computeHeteroPercentage() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.computeHWE">computeHWE() (in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.computeNoCall">computeNoCall() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.computeNumberOfMarkers">computeNumberOfMarkers() (in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.computeStatistics">computeStatistics() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.computeStatistics">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.count_markers_samples">count_markers_samples() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.create_scree_plot">create_scree_plot() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.createAndCleanTPED">createAndCleanTPED() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.createAndCleanTPED">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.createFinalTPEDandTFAM">createFinalTPEDandTFAM() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.createGenderPlot">createGenderPlot() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.createLrrBafPlot">createLrrBafPlot() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.createMDSFile">createMDSFile() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.createPedChr23UsingPlink">createPedChr23UsingPlink() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.createPedChr24UsingPlink">createPedChr24UsingPlink() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.createPopulationFile">createPopulationFile() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.createRowFromPlinkSpacedOutput">createRowFromPlinkSpacedOutput() (in module pyGenClean.PlinkUtils)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="E">E</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.encode_chr">encode_chr() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.encode_chromosome">encode_chromosome() (in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.encode_gender">encode_gender() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.exclude_SNPs_samples">exclude_SNPs_samples() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.excludeSNPs">excludeSNPs() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.executePlateBiasAnalysis">executePlateBiasAnalysis() (in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.extractData">extractData() (in module pyGenClean.PlinkUtils.plot_MDS_standalone)</a>
  </dt>

      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.extractRelatedIndividuals">extractRelatedIndividuals() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

      
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.extractSignificantSNPs">extractSignificantSNPs() (in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.extractSNPs">extractSNPs() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.extractSNPs">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.extractSNPs">(in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

      </dl></dd>
  </dl></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.find_outliers">find_outliers() (in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.find_ref_centers">find_ref_centers() (in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.find_the_outliers">find_the_outliers() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.findDuplicates">findDuplicates() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.findFlippedSNPs">findFlippedSNPs() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.findFlippedSNPs">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.findOverlappingSNPsWithGoldStandard">findOverlappingSNPsWithGoldStandard() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.findOverlappingSNPsWithReference">findOverlappingSNPsWithReference() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      
  <dt><a href="flag_maf_zero.html#pyGenClean.FlagMAF.flag_maf_zero.findSnpWithMaf0">findSnpWithMaf0() (in module pyGenClean.FlagMAF.flag_maf_zero)</a>
  </dt>

      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.findUniques">findUniques() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.flipGenotype">flipGenotype() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.flipSNPs">flipSNPs() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.flipSNPs">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      </dl></dd>
  </dl></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.get_version">get_version() (in module pyGenClean.PlinkUtils)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.getIndexOfHeteroMen">getIndexOfHeteroMen() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.illumina_to_snp">illumina_to_snp() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.is_heterozygous">is_heterozygous() (in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="K">K</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.keepSamples">keepSamples() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.main">main() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.main">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.main">(in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.main">(in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

        
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.main">(in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

        
  <dt><a href="flag_maf_zero.html#pyGenClean.FlagMAF.flag_maf_zero.main">(in module pyGenClean.FlagMAF.flag_maf_zero)</a>
  </dt>

        
  <dt><a href="remove_heterozygous_haploid.html#pyGenClean.HeteroHap.remove_heterozygous_haploid.main">(in module pyGenClean.HeteroHap.remove_heterozygous_haploid)</a>
  </dt>

        
  <dt><a href="marker_missingness.html#pyGenClean.MarkerMissingness.snp_missingness.main">(in module pyGenClean.MarkerMissingness.snp_missingness)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.clean_noCall_hetero_snps.main">(in module pyGenClean.NoCallHetero.clean_noCall_hetero_snps)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.main">(in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

        
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.main">(in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.compare_bim.main">(in module pyGenClean.PlinkUtils.compare_bim)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.main">(in module pyGenClean.PlinkUtils.plot_MDS_standalone)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.subset_data.main">(in module pyGenClean.PlinkUtils.subset_data)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.main">(in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.merge_related_samples.main">(in module pyGenClean.RelatedSamples.merge_related_samples)</a>
  </dt>

        
  <dt><a href="sample_missingness.html#pyGenClean.SampleMissingness.sample_missingness.main">(in module pyGenClean.SampleMissingness.sample_missingness)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.main">(in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.main">(in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.main">(in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

        
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.main">(in module pyGenClean.run_data_clean_up)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.merge_related_samples.merge_related_samples">merge_related_samples() (in module pyGenClean.RelatedSamples.merge_related_samples)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.mergeGenomeLogFiles">mergeGenomeLogFiles() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.parse_args">parse_args() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.parseArgs">parseArgs() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.parseArgs">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.parseArgs">(in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.parseArgs">(in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

        
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.parseArgs">(in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

        
  <dt><a href="flag_maf_zero.html#pyGenClean.FlagMAF.flag_maf_zero.parseArgs">(in module pyGenClean.FlagMAF.flag_maf_zero)</a>
  </dt>

        
  <dt><a href="remove_heterozygous_haploid.html#pyGenClean.HeteroHap.remove_heterozygous_haploid.parseArgs">(in module pyGenClean.HeteroHap.remove_heterozygous_haploid)</a>
  </dt>

        
  <dt><a href="marker_missingness.html#pyGenClean.MarkerMissingness.snp_missingness.parseArgs">(in module pyGenClean.MarkerMissingness.snp_missingness)</a>
  </dt>

        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.parseArgs">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.clean_noCall_hetero_snps.parseArgs">(in module pyGenClean.NoCallHetero.clean_noCall_hetero_snps)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.parseArgs">(in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

        
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.parseArgs">(in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.compare_bim.parseArgs">(in module pyGenClean.PlinkUtils.compare_bim)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.parseArgs">(in module pyGenClean.PlinkUtils.plot_MDS_standalone)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.subset_data.parseArgs">(in module pyGenClean.PlinkUtils.subset_data)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.parseArgs">(in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.merge_related_samples.parseArgs">(in module pyGenClean.RelatedSamples.merge_related_samples)</a>
  </dt>

        
  <dt><a href="sample_missingness.html#pyGenClean.SampleMissingness.sample_missingness.parseArgs">(in module pyGenClean.SampleMissingness.sample_missingness)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.parseArgs">(in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.parseArgs">(in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.parseArgs">(in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.plot_baf_lrr">plot_baf_lrr() (in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.plot_gender">plot_gender() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.plot_heterozygosity">plot_heterozygosity() (in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.plot_related_data">plot_related_data() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.plotMDS">plotMDS() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.plotMDS">(in module pyGenClean.PlinkUtils.plot_MDS_standalone)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.print_data_to_file">print_data_to_file() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.printConcordance">printConcordance() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.printConcordance">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.printDuplicatedTPEDandTFAM">printDuplicatedTPEDandTFAM() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.printDuplicatedTPEDandTFAM">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.printProblems">printProblems() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.printStatistics">printStatistics() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.printUniqueTFAM">printUniqueTFAM() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.processTPED">processTPED() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.processTPED">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.clean_noCall_hetero_snps.processTPEDandTFAM">processTPEDandTFAM() (in module pyGenClean.NoCallHetero.clean_noCall_hetero_snps)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.ProgramError">ProgramError</a>, <a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.ProgramError">[1]</a>, <a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.ProgramError">[2]</a>, <a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.clean_noCall_hetero_snps.ProgramError">[3]</a>, <a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.ProgramError">[4]</a>, <a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.ProgramError">[5]</a>, <a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.ProgramError">[6]</a>, <a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.ProgramError">[7]</a>, <a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.ProgramError">[8]</a>, <a href="find_related_samples.html#pyGenClean.RelatedSamples.merge_related_samples.ProgramError">[9]</a>, <a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.ProgramError">[10]</a>, <a href="flag_maf_zero.html#pyGenClean.FlagMAF.flag_maf_zero.ProgramError">[11]</a>, <a href="marker_missingness.html#pyGenClean.MarkerMissingness.snp_missingness.ProgramError">[12]</a>, <a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.ProgramError">[13]</a>, <a href="plink_utils.html#pyGenClean.PlinkUtils.compare_bim.ProgramError">[14]</a>, <a href="plink_utils.html#pyGenClean.PlinkUtils.subset_data.ProgramError">[15]</a>, <a href="remove_heterozygous_haploid.html#pyGenClean.HeteroHap.remove_heterozygous_haploid.ProgramError">[16]</a>, <a href="sample_missingness.html#pyGenClean.SampleMissingness.sample_missingness.ProgramError">[17]</a>, <a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.ProgramError">[18]</a>, <a href="sex_check.html#pyGenClean.SexCheck.gender_plot.ProgramError">[19]</a>, <a href="sex_check.html#pyGenClean.SexCheck.sex_check.ProgramError">[20]</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#module-pyGenClean.DupSamples.duplicated_samples">pyGenClean.DupSamples.duplicated_samples (module)</a>
  </dt>

      
  <dt><a href="duplicated_markers.html#module-pyGenClean.DupSNPs.duplicated_snps">pyGenClean.DupSNPs.duplicated_snps (module)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#module-pyGenClean.Ethnicity.check_ethnicity">pyGenClean.Ethnicity.check_ethnicity (module)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#module-pyGenClean.Ethnicity.find_outliers">pyGenClean.Ethnicity.find_outliers (module)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="flag_hw.html#module-pyGenClean.FlagHW.flag_hw">pyGenClean.FlagHW.flag_hw (module)</a>
  </dt>

      
  <dt><a href="flag_maf_zero.html#module-pyGenClean.FlagMAF.flag_maf_zero">pyGenClean.FlagMAF.flag_maf_zero (module)</a>
  </dt>

      
  <dt><a href="remove_heterozygous_haploid.html#module-pyGenClean.HeteroHap.remove_heterozygous_haploid">pyGenClean.HeteroHap.remove_heterozygous_haploid (module)</a>
  </dt>

      
  <dt><a href="marker_missingness.html#module-pyGenClean.MarkerMissingness.snp_missingness">pyGenClean.MarkerMissingness.snp_missingness (module)</a>
  </dt>

      
  <dt><a href="compare_gold_standard.html#module-pyGenClean.Misc.compare_gold_standard">pyGenClean.Misc.compare_gold_standard (module)</a>
  </dt>

      
  <dt><a href="clean_noCall_hetero_snps.html#module-pyGenClean.NoCallHetero.clean_noCall_hetero_snps">pyGenClean.NoCallHetero.clean_noCall_hetero_snps (module)</a>
  </dt>

      
  <dt><a href="clean_noCall_hetero_snps.html#module-pyGenClean.NoCallHetero.heterozygosity_plot">pyGenClean.NoCallHetero.heterozygosity_plot (module)</a>
  </dt>

      
  <dt><a href="plate_bias.html#module-pyGenClean.PlateBias.plate_bias">pyGenClean.PlateBias.plate_bias (module)</a>
  </dt>

      
  <dt><a href="plink_utils.html#module-pyGenClean.PlinkUtils">pyGenClean.PlinkUtils (module)</a>
  </dt>

      
  <dt><a href="plink_utils.html#module-pyGenClean.PlinkUtils.compare_bim">pyGenClean.PlinkUtils.compare_bim (module)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#module-pyGenClean.PlinkUtils.plot_MDS_standalone">pyGenClean.PlinkUtils.plot_MDS_standalone (module)</a>
  </dt>

      
  <dt><a href="plink_utils.html#module-pyGenClean.PlinkUtils.subset_data">pyGenClean.PlinkUtils.subset_data (module)</a>
  </dt>

      
  <dt><a href="find_related_samples.html#module-pyGenClean.RelatedSamples.find_related_samples">pyGenClean.RelatedSamples.find_related_samples (module)</a>
  </dt>

      
  <dt><a href="find_related_samples.html#module-pyGenClean.RelatedSamples.merge_related_samples">pyGenClean.RelatedSamples.merge_related_samples (module)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#module-pyGenClean.run_data_clean_up">pyGenClean.run_data_clean_up (module)</a>
  </dt>

      
  <dt><a href="sample_missingness.html#module-pyGenClean.SampleMissingness.sample_missingness">pyGenClean.SampleMissingness.sample_missingness (module)</a>
  </dt>

      
  <dt><a href="sex_check.html#module-pyGenClean.SexCheck.baf_lrr_plot">pyGenClean.SexCheck.baf_lrr_plot (module)</a>
  </dt>

      
  <dt><a href="sex_check.html#module-pyGenClean.SexCheck.gender_plot">pyGenClean.SexCheck.gender_plot (module)</a>
  </dt>

      
  <dt><a href="sex_check.html#module-pyGenClean.SexCheck.sex_check">pyGenClean.SexCheck.sex_check (module)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.read_bim">read_bim() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.read_config_file">read_config_file() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.read_fam">read_fam() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.read_intensities">read_intensities() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.read_mds_file">read_mds_file() (in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.read_population_file">read_population_file() (in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.read_problematic_samples">read_problematic_samples() (in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.read_same_samples_file">read_same_samples_file() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.read_sex_problems">read_sex_problems() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.read_source_alleles">read_source_alleles() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.read_source_manifest">read_source_manifest() (in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.read_summarized_intensities">read_summarized_intensities() (in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

      
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.compare_bim.readBIM">readBIM() (in module pyGenClean.PlinkUtils.compare_bim)</a>
  </dt>

      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.readCheckSexFile">readCheckSexFile() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.readMAP">readMAP() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.readPopulations">readPopulations() (in module pyGenClean.PlinkUtils.plot_MDS_standalone)</a>
  </dt>

      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.readTFAM">readTFAM() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.readTFAM">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.renameSNPs">renameSNPs() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.renameSNPs">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_check_ethnicity">run_check_ethnicity() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_command">run_command() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_compare_gold_standard">run_compare_gold_standard() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_duplicated_samples">run_duplicated_samples() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_duplicated_snps">run_duplicated_snps() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_find_related_samples">run_find_related_samples() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_flag_hw">run_flag_hw() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_flag_maf_zero">run_flag_maf_zero() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_noCall_hetero_snps">run_noCall_hetero_snps() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_plate_bias">run_plate_bias() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_remove_heterozygous_haploid">run_remove_heterozygous_haploid() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_sample_missingness">run_sample_missingness() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_sex_check">run_sex_check() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_snp_missingness">run_snp_missingness() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.run_subset_data">run_subset_data() (in module pyGenClean.run_data_clean_up)</a>
  </dt>

      
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.runCommand">runCommand() (in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.runCommand">(in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

        
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.runCommand">(in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.runCommand">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

        
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.runCommand">(in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.subset_data.runCommand">(in module pyGenClean.PlinkUtils.subset_data)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.runCommand">(in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.runCommand">(in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.runGenome">runGenome() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.runGenomeSGE">runGenomeSGE() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

      
  <dt><a href="remove_heterozygous_haploid.html#pyGenClean.HeteroHap.remove_heterozygous_haploid.runPlink">runPlink() (in module pyGenClean.HeteroHap.remove_heterozygous_haploid)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="marker_missingness.html#pyGenClean.MarkerMissingness.snp_missingness.runPlink">(in module pyGenClean.MarkerMissingness.snp_missingness)</a>
  </dt>

        
  <dt><a href="sample_missingness.html#pyGenClean.SampleMissingness.sample_missingness.runPlink">(in module pyGenClean.SampleMissingness.sample_missingness)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.runPlinkSexCheck">runPlinkSexCheck() (in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

      
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.runRelatedness">runRelatedness() (in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="duplicated_samples.html#pyGenClean.DupSamples.duplicated_samples.safe_main">safe_main() (in module pyGenClean.DupSamples.duplicated_samples)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="duplicated_markers.html#pyGenClean.DupSNPs.duplicated_snps.safe_main">(in module pyGenClean.DupSNPs.duplicated_snps)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.check_ethnicity.safe_main">(in module pyGenClean.Ethnicity.check_ethnicity)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.Ethnicity.find_outliers.safe_main">(in module pyGenClean.Ethnicity.find_outliers)</a>
  </dt>

        
  <dt><a href="flag_hw.html#pyGenClean.FlagHW.flag_hw.safe_main">(in module pyGenClean.FlagHW.flag_hw)</a>
  </dt>

        
  <dt><a href="flag_maf_zero.html#pyGenClean.FlagMAF.flag_maf_zero.safe_main">(in module pyGenClean.FlagMAF.flag_maf_zero)</a>
  </dt>

        
  <dt><a href="remove_heterozygous_haploid.html#pyGenClean.HeteroHap.remove_heterozygous_haploid.safe_main">(in module pyGenClean.HeteroHap.remove_heterozygous_haploid)</a>
  </dt>

        
  <dt><a href="marker_missingness.html#pyGenClean.MarkerMissingness.snp_missingness.safe_main">(in module pyGenClean.MarkerMissingness.snp_missingness)</a>
  </dt>

        
  <dt><a href="compare_gold_standard.html#pyGenClean.Misc.compare_gold_standard.safe_main">(in module pyGenClean.Misc.compare_gold_standard)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.clean_noCall_hetero_snps.safe_main">(in module pyGenClean.NoCallHetero.clean_noCall_hetero_snps)</a>
  </dt>

        
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.safe_main">(in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

        
  <dt><a href="plate_bias.html#pyGenClean.PlateBias.plate_bias.safe_main">(in module pyGenClean.PlateBias.plate_bias)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.compare_bim.safe_main">(in module pyGenClean.PlinkUtils.compare_bim)</a>
  </dt>

        
  <dt><a href="check_ethnicity.html#pyGenClean.PlinkUtils.plot_MDS_standalone.safe_main">(in module pyGenClean.PlinkUtils.plot_MDS_standalone)</a>
  </dt>

        
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.subset_data.safe_main">(in module pyGenClean.PlinkUtils.subset_data)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.safe_main">(in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

        
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.merge_related_samples.safe_main">(in module pyGenClean.RelatedSamples.merge_related_samples)</a>
  </dt>

        
  <dt><a href="sample_missingness.html#pyGenClean.SampleMissingness.sample_missingness.safe_main">(in module pyGenClean.SampleMissingness.sample_missingness)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.baf_lrr_plot.safe_main">(in module pyGenClean.SexCheck.baf_lrr_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.gender_plot.safe_main">(in module pyGenClean.SexCheck.gender_plot)</a>
  </dt>

        
  <dt><a href="sex_check.html#pyGenClean.SexCheck.sex_check.safe_main">(in module pyGenClean.SexCheck.sex_check)</a>
  </dt>

        
  <dt><a href="data_clean_up_module.html#pyGenClean.run_data_clean_up.safe_main">(in module pyGenClean.run_data_clean_up)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="clean_noCall_hetero_snps.html#pyGenClean.NoCallHetero.heterozygosity_plot.save_heterozygosity">save_heterozygosity() (in module pyGenClean.NoCallHetero.heterozygosity_plot)</a>
  </dt>

      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.selectSNPsAccordingToLD">selectSNPsAccordingToLD() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="find_related_samples.html#pyGenClean.RelatedSamples.find_related_samples.splitFile">splitFile() (in module pyGenClean.RelatedSamples.find_related_samples)</a>
  </dt>

      
  <dt><a href="plink_utils.html#pyGenClean.PlinkUtils.subset_data.subset_data">subset_data() (in module pyGenClean.PlinkUtils.subset_data)</a>
  </dt>

  </dl></td>
</tr></table>



          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.7.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>