<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Result Summary Table &mdash; pyGenClean 1.8.1 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.8.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.8.1 documentation" href="index.html" />
    <link rel="prev" title="Plink Utils" href="plink_utils.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="index.html"><img src="_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="plink_utils.html" title="Plink Utils"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.1 documentation</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="plink_utils.html"
                        title="previous chapter">Plink Utils</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/result_table.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="result-summary-table">
<h1>Result Summary Table<a class="headerlink" href="#result-summary-table" title="Permalink to this headline">¶</a></h1>
<p>This table summurized  information available from output files produced by
<em>pyGenClean</em> during the data clean up procedure. Numbers correspond to number of
lines in output files see <a class="reference internal" href="protocol.html#proposed-protocol-label"><span>Proposed Protocol</span></a> for details. Only
removed SNPs and IDs are indicated in the column SNPs and IDs, flagged SNPs or
IDs are present in the <span class="math">\(n\)</span> column.</p>
<table border="1" class="docutils" id="id2">
<span id="id1"></span><caption><span class="caption-text">Summary information of the data clean up procedure.</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="54%" />
<col width="21%" />
<col width="15%" />
<col width="10%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Description</th>
<th class="head"><span class="math">\(n\)</span></th>
<th class="head">SNPs</th>
<th class="head">IDs</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Total number of SNPs in file received</td>
<td>2,379,855</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Total number of samples</td>
<td>494</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of duplicate samples</td>
<td>0</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of individuals with no
genotype (failed)</td>
<td>0</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of SNPs with no physical
position (chromosome and physical
position = 0)</td>
<td>7,239</td>
<td>-7,239</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of INDEL</td>
<td>43</td>
<td>-43</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of replicate controls</td>
<td>5</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of replicate samples</td>
<td>0</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of duplicate SNPs (by
chromosome and physical position)</td>
<td>5,643</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Duplicated SNPs by chromosome and
physical position with the same
allele (merge)</td>
<td>5,417</td>
<td>-5,147</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of duplicated SNP with &lt;98%
concordance</td>
<td>22</td>
<td>-22</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Completely failed SNPs</td>
<td>1</td>
<td>-1</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>All heterozogous SNPs</td>
<td>0</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of individuals removed because
they have more than 10% missing
genotypes</td>
<td>5</td>
<td>&nbsp;</td>
<td>-5</td>
</tr>
<tr class="row-even"><td>Number of SNPs removed because they
have more than 2% missing value</td>
<td>128,562</td>
<td>-128,562</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of individuals removed because
they have more than 2% missing
genotypes</td>
<td>7</td>
<td>&nbsp;</td>
<td>-7</td>
</tr>
<tr class="row-even"><td>Number of individuals with gender
problem</td>
<td>1</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of SNPs with plate bias test P
value below threshold of
<span class="math">\(1\times10^{-7}\)</span></td>
<td>19</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of SNPs used for IBS analysis</td>
<td>73,651</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of duplicates pairs or twin</td>
<td>1</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of related pairs
(including twins)</td>
<td>2</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of SNPs used for MDS analysis</td>
<td>80,262</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of individuals with ethnicity
other than Caucasian as detected by
MDS analysis</td>
<td>20</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of gender problems</td>
<td>1</td>
<td>&nbsp;</td>
<td>-1</td>
</tr>
<tr class="row-even"><td>Number of related pairs</td>
<td>2</td>
<td>&nbsp;</td>
<td>-2</td>
</tr>
<tr class="row-odd"><td>Number of caucasian outliers</td>
<td>20</td>
<td>&nbsp;</td>
<td>-20</td>
</tr>
<tr class="row-even"><td>Number of controls</td>
<td>5</td>
<td>&nbsp;</td>
<td>-5</td>
</tr>
<tr class="row-odd"><td>Number of heterozygote haploid
genotypes set to missing (after
correction of gender problems)</td>
<td>277,206</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of SNPs with MAF=0</td>
<td>602,480</td>
<td>-602,480</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>Number of SNPS with HWE test P Value
below threshold of
<span class="math">\(1\times10^{-4}\)</span> and higher
than Bonferroni threshold</td>
<td>603</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>Number of SNPS with HWE test below
Bonferroni threshold</td>
<td>162</td>
<td>-162</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td><strong>Total number of SNPs</strong></td>
<td><strong>1,635,931</strong></td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td><strong>Total number of samples</strong></td>
<td><strong>454</strong></td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="plink_utils.html" title="Plink Utils"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>