<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Sex Check Module &#8212; pyGenClean 1.8.3 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.8.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Plate Bias Module" href="plate_bias.html" />
    <link rel="prev" title="Marker Missingness Module" href="marker_missingness.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="index.html"><img src="_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="plate_bias.html" title="Plate Bias Module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="marker_missingness.html" title="Marker Missingness Module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.3 documentation</a> &#187;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Sex Check Module</a><ul>
<li><a class="reference internal" href="#input-files">Input Files</a></li>
<li><a class="reference internal" href="#procedure">Procedure</a></li>
<li><a class="reference internal" href="#output-files">Output Files</a></li>
<li><a class="reference internal" href="#the-plots">The Plots</a></li>
<li><a class="reference internal" href="#the-algorithm">The Algorithm</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="marker_missingness.html"
                        title="previous chapter">Marker Missingness Module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="plate_bias.html"
                        title="next chapter">Plate Bias Module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/sex_check.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="sex-check-module">
<span id="sexcheck-module-lable"></span><h1>Sex Check Module<a class="headerlink" href="#sex-check-module" title="Permalink to this headline">¶</a></h1>
<p>The usage of the standalone module is shown below:</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> pyGenClean_sex_check --help
<span class="go">usage: pyGenClean_sex_check [-h] [-v] --bfile FILE [--femaleF FLOAT]</span>
<span class="go">                            [--maleF FLOAT] [--nbChr23 INT] [--gender-plot]</span>
<span class="go">                            [--sex-chr-intensities FILE]</span>
<span class="go">                            [--gender-plot-format FORMAT] [--lrr-baf]</span>
<span class="go">                            [--lrr-baf-raw-dir DIR] [--lrr-baf-format FORMAT]</span>
<span class="go">                            [--out FILE]</span>

<span class="go">Check sample&#39;s gender using Plink.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>

<span class="go">Input File:</span>
<span class="go">  --bfile FILE          The input file prefix (will find the Plink binary</span>
<span class="go">                        files by appending the prefix to the .bed, .bim, and</span>
<span class="go">                        .fam files, respectively.</span>

<span class="go">Options:</span>
<span class="go">  --femaleF FLOAT       The female F threshold. [default: &lt; 0.300000]</span>
<span class="go">  --maleF FLOAT         The male F threshold. [default: &gt; 0.700000]</span>
<span class="go">  --nbChr23 INT         The minimum number of markers on chromosome 23 before</span>
<span class="go">                        computing Plink&#39;s sex check [default: 50]</span>

<span class="go">Gender Plot:</span>
<span class="go">  --gender-plot         Create the gender plot (summarized chr Y intensities</span>
<span class="go">                        in function of summarized chr X intensities) for</span>
<span class="go">                        problematic samples.</span>
<span class="go">  --sex-chr-intensities FILE</span>
<span class="go">                        A file containing alleles intensities for each of the</span>
<span class="go">                        markers located on the X and Y chromosome for the</span>
<span class="go">                        gender plot.</span>
<span class="go">  --gender-plot-format FORMAT</span>
<span class="go">                        The output file format for the gender plot (png, ps,</span>
<span class="go">                        pdf, or X11 formats are available). [default: png]</span>

<span class="go">LRR and BAF Plot:</span>
<span class="go">  --lrr-baf             Create the LRR and BAF plot for problematic samples</span>
<span class="go">  --lrr-baf-raw-dir DIR</span>
<span class="go">                        Directory or list of directories containing</span>
<span class="go">                        information about every samples (BAF and LRR).</span>
<span class="go">  --lrr-baf-format FORMAT</span>
<span class="go">                        The output file format for the LRR and BAF plot (png,</span>
<span class="go">                        ps, pdf, or X11 formats are available). [default: png]</span>

<span class="go">Output File:</span>
<span class="go">  --out FILE            The prefix of the output files (which will be a Plink</span>
<span class="go">                        binary file. [default: sexcheck]</span>
</pre></div>
</div>
<div class="section" id="input-files">
<h2>Input Files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h2>
<p>This module uses PLINK&#8217;s binary file format (<code class="docutils literal"><span class="pre">bed</span></code>, <code class="docutils literal"><span class="pre">bim</span></code> and <code class="docutils literal"><span class="pre">fam</span></code> files)
for the source data set (the data of interest).</p>
<p>If the option of generating the gender plot is used, a file containing
intensities information about each markers on the sexual chromosomes is
required. This file (which could be gzipped) should contain at least the
following columns:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">Sample</span> <span class="pre">ID</span></code>: the unique sample id for each individual.</li>
<li><code class="docutils literal"><span class="pre">SNP</span> <span class="pre">Name</span></code>: the unique name of each markers.</li>
<li><code class="docutils literal"><span class="pre">Chr</span></code>: the name of the chromosome on which each marker is located.</li>
<li><code class="docutils literal"><span class="pre">X</span></code>: the intensities of the first allele of each marker.</li>
<li><code class="docutils literal"><span class="pre">Y</span></code>: the intensities of the second allele of each marker.</li>
</ul>
</div></blockquote>
<p>If the options of generating the BAF and LRR values is used, the name of a
directory containing intensities file for each sample is required. The name of
each file should be the unique sample id. This file (which could be gzipped)
should contain at least the following columns:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">Chr</span></code>: the name of the chromosome on which each marker is located.</li>
<li><code class="docutils literal"><span class="pre">Position</span></code>: the position of the marker on the chromosome.</li>
<li><code class="docutils literal"><span class="pre">B</span> <span class="pre">Allele</span> <span class="pre">Freq</span></code>: the BAF value of each marker.</li>
<li><code class="docutils literal"><span class="pre">Log</span> <span class="pre">R</span> <span class="pre">Ratio</span></code>: the LRR value of each marker.</li>
</ul>
</div></blockquote>
<p>If the two plotting module is used alone, one more file is required per module:
a list of samples with gender problem and explanation for
<a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.gender_plot" title="pyGenClean.SexCheck.gender_plot"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.gender_plot</span></code></a>, and only the list of samples with
gender problem for <a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.baf_lrr_plot" title="pyGenClean.SexCheck.baf_lrr_plot"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.baf_lrr_plot</span></code></a> (both files are
provided by the <a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.sex_check" title="pyGenClean.SexCheck.sex_check"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.sex_check</span></code></a> module).</p>
</div>
<div class="section" id="procedure">
<h2>Procedure<a class="headerlink" href="#procedure" title="Permalink to this headline">¶</a></h2>
<p>Here are the steps performed by the module:</p>
<ol class="arabic simple">
<li>Checks if there are enough markers on the chromosome 23. If not, the module
stops here.</li>
<li>Runs the sex check analysis using Plink.</li>
<li>If there are no sex problems, the module quits.</li>
<li>Creates the recoded file for the chromosome 23.</li>
<li>Computes the heterozygosity percentage on the chromosome 23.</li>
<li>If there are enough markers on chromosome 24 (at least 1), creates the
recoded file for this chromosome.</li>
<li>Computes the number of no call on the chromosome 24.</li>
<li>If required, plots the gender plot.<ol class="lowerroman">
<li>If there are <code class="docutils literal"><span class="pre">summarized_intensities</span></code> provided, reads the files
and skips to step vi.</li>
<li>Reads the <code class="docutils literal"><span class="pre">bim</span></code> file to get markers on the sexual chromosomes.</li>
<li>Reads the <code class="docutils literal"><span class="pre">fam</span></code> file to get individual&#8217;s gender.</li>
<li>Reads the file containing samples with sex problems.</li>
<li>Reads the intensities and summarizes them.</li>
<li>Plots the summarized intensities.</li>
</ol>
</li>
<li>If required, plots the BAF and LRR plots.<ol class="lowerroman">
<li>Reads the problematic samples.</li>
<li>Finds and checks the raw files for each of the problematic samples.</li>
<li>Plots the BAF and LRR plots.</li>
</ol>
</li>
</ol>
</div>
<div class="section" id="output-files">
<h2>Output Files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The output files of each of the steps described above are as follow (note that
the output prefix shown is the one by default [<em>i.e</em> <code class="docutils literal"><span class="pre">sexcheck</span></code>]):</p>
<ol class="arabic simple">
<li>No output files created.</li>
<li>One set of PLINK&#8217;s result files:<ul>
<li><code class="docutils literal"><span class="pre">sexcheck</span></code>: the result of the sex check procedure from Plink.</li>
</ul>
</li>
<li>Two files are created if there are sex problems:<ul>
<li><code class="docutils literal"><span class="pre">sexcheck.list_problem_sex</span></code>: a summary of samples with sex problem.</li>
<li><code class="docutils literal"><span class="pre">sexcheck.list_problem_sex_ids</span></code>: the list of sample ids with sex
problem.</li>
</ul>
</li>
<li>One set of Plink&#8217;s files:<ul>
<li><code class="docutils literal"><span class="pre">sexcheck.chr23_recodeA</span></code>: the recoded file for the chromosome 23.</li>
</ul>
</li>
<li>One custom output file:<ul>
<li><code class="docutils literal"><span class="pre">sexcheck.chr23_recodeA.raw.hetero</span></code>: the heterozygosity percentage on
the chromosome 23. The file includes the following columns: <code class="docutils literal"><span class="pre">PED</span></code> (the
pedigree ID), <code class="docutils literal"><span class="pre">ID</span></code> (the individual ID), <code class="docutils literal"><span class="pre">SEX</span></code> (the gender) and
<code class="docutils literal"><span class="pre">HETERO</span></code> (the heterozygosity).</li>
</ul>
</li>
<li>One set of Plink&#8217;s files:<ul>
<li><code class="docutils literal"><span class="pre">sexcheck.chr24_recodeA</span></code>: the recoded file for the chromosome 24.</li>
</ul>
</li>
<li>One custom output file:<ul>
<li><code class="docutils literal"><span class="pre">sexcheck.chr24_recodeA.raw.noCall</span></code>: the number of no call on the
chromosome 24. The file includes the following columns: <code class="docutils literal"><span class="pre">PED</span></code> (the
pedigree ID), <code class="docutils literal"><span class="pre">ID</span></code> (the individual ID), <code class="docutils literal"><span class="pre">SEX</span></code> (the gender),
<code class="docutils literal"><span class="pre">nbGeno</span></code> (the number of genotypes on the chromosome 24) and
<code class="docutils literal"><span class="pre">nbNoCall</span></code> (the number of genotypes that were not called on chromosome
24).</li>
</ul>
</li>
<li>Multiple files and one plot. The files are created so that the plot can be
generated again with different parameters (since the summarized intensities
for each sample are really long to compute).<ul>
<li><code class="docutils literal"><span class="pre">sexcheck.png</span></code>: the gender plot (see Figure <a class="reference internal" href="#gender-plot-figure"><span class="std std-ref">Gender plot</span></a>).</li>
<li><code class="docutils literal"><span class="pre">sexcheck.ok_females.txt</span></code>: the list of females without sex problem,
including their summarized intensities on chromosome 23 and 24.</li>
<li><code class="docutils literal"><span class="pre">sexcheck.ok_males.txt</span></code>: the list of males without sex problem,
including their summarized intensities on chromosome 23 and 24.</li>
<li><code class="docutils literal"><span class="pre">sexcheck.ok_unknowns.txt</span></code>: the list of unknown gender without sex
problem, including their summarized intensities on chromosome 23 and 24.</li>
<li><code class="docutils literal"><span class="pre">sexcheck.problematic_females.txt</span></code>: the list of females with sex
problem, including their summarized intensities on chromosome 23 and 24.</li>
<li><code class="docutils literal"><span class="pre">sexcheck.problematic_males.txt</span></code>: the list of males with sex problem,
including their summarized intensities on chromosome 23 and 24.</li>
<li><code class="docutils literal"><span class="pre">sexcheck.problematic_unknowns.txt</span></code>: the list of unknown gender with
sex problem, including their summarized intensities on chromosome 23 and
24. When this file is created by the
<a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.sex_check" title="pyGenClean.SexCheck.sex_check"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.sex_check</span></code></a> module, it is empty.</li>
</ul>
</li>
<li>A directory containing one file per individual with gender problem (see
Figure <a class="reference internal" href="#baf-lrr-plot-figure"><span class="std std-ref">BAF and LRR plot</span></a>).</li>
</ol>
</div>
<div class="section" id="the-plots">
<span id="sex-check-plots"></span><h2>The Plots<a class="headerlink" href="#the-plots" title="Permalink to this headline">¶</a></h2>
<p>The plot generated by the <a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.gender_plot" title="pyGenClean.SexCheck.gender_plot"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.gender_plot</span></code></a> module (the
<a class="reference internal" href="#gender-plot-figure"><span class="std std-ref">Gender plot</span></a> figure) represents the summarized intensities of each
sample of the data set. The color code represent the gender of each sample. Blue
and red dots represent the males and females without gender problem,
respectively. The green and purple triangles represent the males and females
with gender problem. The gray dots and triangles represent the samples with
unknown gender, with and without gender problem, respectively. This plot makes
possible to find samples with sexual chromosomes abnormalities, such as males
which are XXY or females with only one copy of the X chromosome (X0). Males that
appear to be females and vice versa might in fact be sample mix up and would
require further analysis.</p>
<div class="figure align-center" id="id1">
<span id="gender-plot-figure"></span><a class="reference internal image-reference" href="_images/sexcheck.png"><img alt="Gender plot" src="_images/sexcheck.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-text">Gender plot</span></p>
</div>
<p>The <a class="reference internal" href="#gender-plot-figure"><span class="std std-ref">Gender plot</span></a> figure can also be manually created after the data
clean up pipeline, using its results and this following standalone script:</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> pyGenClean_gender_plot --help
<span class="go">usage: pyGenClean_gender_plot [-h] [-v] --bfile FILE [--intensities FILE]</span>
<span class="go">                              [--summarized-intensities FILE] --sex-problems</span>
<span class="go">                              FILE [--format FORMAT] [--xlabel STRING]</span>
<span class="go">                              [--ylabel STRING] [--out FILE]</span>

<span class="go">Plots the gender using X and Y chromosomes&#39; intensities</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>

<span class="go">Input File:</span>
<span class="go">  --bfile FILE          The plink binary file containing information about</span>
<span class="go">                        markers and individuals. Must be specified if</span>
<span class="go">                        &#39;--summarized-intensities&#39; is not.</span>
<span class="go">  --intensities FILE    A file containing alleles intensities for each of the</span>
<span class="go">                        markers located on the X and Y chromosome. Must be</span>
<span class="go">                        specified if &#39;--summarized-intensities&#39; is not.</span>
<span class="go">  --summarized-intensities FILE</span>
<span class="go">                        The prefix of six files (prefix.ok_females.txt,</span>
<span class="go">                        prefix.ok_males.txt, prefix.ok_unknowns.txt,</span>
<span class="go">                        problematic_females.txt, problematic_males.txt and</span>
<span class="go">                        problematic_unknowns.txt) containing summarized chr23</span>
<span class="go">                        and chr24 intensities. Must be specified if &#39;--bfile&#39;</span>
<span class="go">                        and &#39;--intensities&#39; are not.</span>
<span class="go">  --sex-problems FILE   The file containing individuals with sex problems.</span>
<span class="go">                        This file is not read if the option &#39;summarized-</span>
<span class="go">                        intensities&#39; is used.</span>

<span class="go">Options:</span>
<span class="go">  --format FORMAT       The output file format (png, ps, pdf, or X11 formats</span>
<span class="go">                        are available). [default: png]</span>
<span class="go">  --xlabel STRING       The label of the X axis. [default: X intensity]</span>
<span class="go">  --ylabel STRING       The label of the Y axis. [default: Y intensity]</span>

<span class="go">Output File:</span>
<span class="go">  --out FILE            The prefix of the output files (which will be a Plink</span>
<span class="go">                        binary file. [default: sexcheck]</span>
</pre></div>
</div>
<p>The log R ratio (LRR) for a sample is the log ratio of the normalized R value
for the marker divided by the expected normalized R value. Hence, a value of 0
means 2 copies. A drop in the LRR shows a loss of a copy, while an increasing
LRR shows a gain of a copy. Expected LRR values on chromosome 23 for female and
female are 0 and [-0.5, -1], respectively. The LRR values of each markers on
both the X and Y chromosomes are shown in the <a class="reference internal" href="#baf-lrr-plot-figure"><span class="std std-ref">BAF and LRR plot</span></a>
figure.</p>
<p>The B allele frequency (BAF) for a sample shows the theta value for a marker,
corrected for cluster position. For a normal number of copies, markers should
have a BAF around 1 (homozygous for the B allele), 0.5 (heterozygous) or 0
(homozygous for A allele). Normal females should have the three lines across the
chromosome. Normal males should only have two lines, located near 1 or 0. The
BAF values of each markers on both the X and Y chromosomes are shown in the
<a class="reference internal" href="#baf-lrr-plot-figure"><span class="std std-ref">BAF and LRR plot</span></a> figure.</p>
<div class="figure align-center" id="id2">
<span id="baf-lrr-plot-figure"></span><a class="reference internal image-reference" href="_images/lrr_baf_plot.png"><img alt="BAF and LRR plot" src="_images/lrr_baf_plot.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-text">BAF and LRR plot</span></p>
</div>
<p>The <a class="reference internal" href="#baf-lrr-plot-figure"><span class="std std-ref">BAF and LRR plot</span></a> figure can also be manually created after the data
clean up pipeline, using this following standalone script:</p>
<div class="highlight-console"><div class="highlight"><pre><span></span><span class="gp">$</span> pyGenClean_baf_lrr_plot --help
<span class="go">usage: pyGenClean_baf_lrr_plot [-h] [-v] --problematic-samples FILE</span>
<span class="go">                               [--use-full-ids] [--full-ids-delimiter CHAR]</span>
<span class="go">                               --raw-dir DIR [--format FORMAT] [--out FILE]</span>

<span class="go">Plots the BAF and LRR of problematic samples.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>

<span class="go">Input File:</span>
<span class="go">  --problematic-samples FILE</span>
<span class="go">                        A file containing the list of samples with sex</span>
<span class="go">                        problems (family and individual ID required, separated</span>
<span class="go">                        by a single tabulation). Uses only the individual ID</span>
<span class="go">                        by default, unless --use-full-ids is used.</span>
<span class="go">  --use-full-ids        Use this options to use full sample IDs (famID and</span>
<span class="go">                        indID). Otherwise, only the indID will be use.</span>
<span class="go">  --full-ids-delimiter CHAR</span>
<span class="go">                        The delimiter between famID and indID for the raw file</span>
<span class="go">                        names. [default: _]</span>
<span class="go">  --raw-dir DIR         Directory containing information about every samples</span>
<span class="go">                        (BAF and LRR).</span>

<span class="go">Options:</span>
<span class="go">  --format FORMAT       The output file format (png, ps, pdf, or X11 formats</span>
<span class="go">                        are available). [default: png]</span>

<span class="go">Output File:</span>
<span class="go">  --out FILE            The prefix of the output files. [default: sexcheck]</span>
</pre></div>
</div>
</div>
<div class="section" id="the-algorithm">
<h2>The Algorithm<a class="headerlink" href="#the-algorithm" title="Permalink to this headline">¶</a></h2>
<p>For more information about the actual algorithms and source codes, refer to the
following pages.</p>
<ul class="simple">
<li><a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.sex_check" title="pyGenClean.SexCheck.sex_check"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.sex_check</span></code></a></li>
<li><a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.gender_plot" title="pyGenClean.SexCheck.gender_plot"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.gender_plot</span></code></a></li>
<li><a class="reference internal" href="module_content/pyGenClean.SexCheck.html#module-pyGenClean.SexCheck.baf_lrr_plot" title="pyGenClean.SexCheck.baf_lrr_plot"><code class="xref py py-mod docutils literal"><span class="pre">pyGenClean.SexCheck.baf_lrr_plot</span></code></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="plate_bias.html" title="Plate Bias Module"
             >next</a> |</li>
        <li class="right" >
          <a href="marker_missingness.html" title="Marker Missingness Module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pyGenClean 1.8.3 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>