<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pyGenClean.SexCheck.baf_lrr_plot &#8212; pyGenClean 1.8.3 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.8.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="../../../index.html"><img src="../../../_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.3 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" accesskey="U">pyGenClean</a> &#187;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for pyGenClean.SexCheck.baf_lrr_plot</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python2.7</span>

<span class="c1"># This file is part of pyGenClean.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is free software: you can redistribute it and/or modify it under</span>
<span class="c1"># the terms of the GNU General Public License as published by the Free Software</span>
<span class="c1"># Foundation, either version 3 of the License, or (at your option) any later</span>
<span class="c1"># version.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is distributed in the hope that it will be useful, but WITHOUT ANY</span>
<span class="c1"># WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR</span>
<span class="c1"># A PARTICULAR PURPOSE.  See the GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1"># You should have received a copy of the GNU General Public License along with</span>
<span class="c1"># pyGenClean.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">..</span> <span class="k">import</span> <span class="n">__version__</span>


<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;baf_llr_plot&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;The main function of this module.</span>

<span class="sd">    :param argString: the options.</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    These are the steps:</span>

<span class="sd">    1. Prints the options.</span>
<span class="sd">    2. Reads the problematic samples (:py:func:`read_problematic_samples`).</span>
<span class="sd">    3. Finds and checks the raw files for each of the problematic samples</span>
<span class="sd">       (:py:func:`check_file_names`).</span>
<span class="sd">    4. Plots the BAF and LRR (:py:func:`plot_baf_lrr`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Getting and checking the options</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>
    <span class="n">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="c1"># Reading the problematic samples</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">read_problematic_samples</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">problematic_samples</span><span class="p">)</span>

    <span class="c1"># Checking for file names</span>
    <span class="n">file_names</span> <span class="o">=</span> <span class="n">check_file_names</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">raw_dir</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

    <span class="c1"># Plotting the BAF and LRR for chromosome X and Y</span>
    <span class="n">plot_baf_lrr</span><span class="p">(</span><span class="n">file_names</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span></div>


<div class="viewcode-block" id="check_file_names"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.check_file_names">[docs]</a><span class="k">def</span> <span class="nf">check_file_names</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">raw_dir</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check if all files are present.</span>

<span class="sd">    :param samples: a list of tuples with the family ID as first element (str)</span>
<span class="sd">                    and sample ID as last element (str).</span>
<span class="sd">    :param raw_dir: the directory containing the raw files.</span>
<span class="sd">    :param options: the options.</span>

<span class="sd">    :type samples: list of tuples</span>
<span class="sd">    :type raw_dir: str</span>
<span class="sd">    :type options: argparse.Namespace</span>

<span class="sd">    :returns: a dict containing samples as key (a tuple with the family ID as</span>
<span class="sd">              first element and sample ID as last element) and the name of the</span>
<span class="sd">              raw file as element.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file_names</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">:</span>
        <span class="n">the_sample</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">the_sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;problematic samples file should include both family and &quot;</span>
                   <span class="s2">&quot;individual IDs&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">use_full_ids</span><span class="p">:</span>
            <span class="n">the_sample</span> <span class="o">=</span> <span class="n">options</span><span class="o">.</span><span class="n">full_ids_delimiter</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

        <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">raw_dir</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">the_sample</span><span class="p">))</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
            <span class="n">file_name</span> <span class="o">+=</span> <span class="s2">&quot;.gz&quot;</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;can&#39;t find file for sample </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">the_sample</span><span class="p">)</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="n">file_names</span><span class="p">[</span><span class="n">the_sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>

    <span class="k">return</span> <span class="n">file_names</span></div>


<div class="viewcode-block" id="read_problematic_samples"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.read_problematic_samples">[docs]</a><span class="k">def</span> <span class="nf">read_problematic_samples</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads a file with sample IDs.</span>

<span class="sd">    :param file_name: the name of the file containing problematic samples after</span>
<span class="sd">                      sex check.</span>

<span class="sd">    :type file_name: str</span>

<span class="sd">    :returns: a set of problematic samples (tuple containing the family ID as</span>
<span class="sd">              first element and the sample ID as last element).</span>

<span class="sd">    Reads a file containing problematic samples after sex check. The file is</span>
<span class="sd">    provided by the module :py:mod:`pyGenClean.SexCheck.sex_check`. This file</span>
<span class="sd">    contains two columns, the first one being the family ID and the second one,</span>
<span class="sd">    the sample ID.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">problematic_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">open_func</span> <span class="o">=</span> <span class="nb">open</span>
    <span class="k">if</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">):</span>
        <span class="n">open_func</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span>
    <span class="k">with</span> <span class="n">open_func</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="n">problematic_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span>
            <span class="nb">tuple</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">input_file</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="p">])</span>

    <span class="k">return</span> <span class="n">problematic_samples</span></div>


<div class="viewcode-block" id="encode_chromosome"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.encode_chromosome">[docs]</a><span class="k">def</span> <span class="nf">encode_chromosome</span><span class="p">(</span><span class="n">chromosome</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Encodes chromosomes.</span>

<span class="sd">    :param chromosome: the chromosome to encode.</span>

<span class="sd">    :type chromosome: str</span>

<span class="sd">    :returns: the encoded chromosome.</span>

<span class="sd">    Encodes the sexual chromosomes, from ``23`` and ``24`` to ``X`` and ``Y``,</span>
<span class="sd">    respectively.</span>

<span class="sd">    .. note::</span>
<span class="sd">        Only the sexual chromosomes are encoded.</span>

<span class="sd">    .. testsetup::</span>

<span class="sd">        from pyGenClean.SexCheck.baf_lrr_plot import encode_chromosome</span>

<span class="sd">    .. doctest::</span>

<span class="sd">        &gt;&gt;&gt; encode_chromosome(&quot;23&quot;)</span>
<span class="sd">        &#39;X&#39;</span>
<span class="sd">        &gt;&gt;&gt; encode_chromosome(&quot;24&quot;)</span>
<span class="sd">        &#39;Y&#39;</span>
<span class="sd">        &gt;&gt;&gt; encode_chromosome(&quot;This is not a chromosome&quot;)</span>
<span class="sd">        &#39;This is not a chromosome&#39;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">chromosome</span> <span class="o">==</span> <span class="s2">&quot;23&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;X&quot;</span>
    <span class="k">if</span> <span class="n">chromosome</span> <span class="o">==</span> <span class="s2">&quot;24&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;Y&quot;</span>
    <span class="k">return</span> <span class="n">chromosome</span></div>


<div class="viewcode-block" id="plot_baf_lrr"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.plot_baf_lrr">[docs]</a><span class="k">def</span> <span class="nf">plot_baf_lrr</span><span class="p">(</span><span class="n">file_names</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Plot BAF and LRR for a list of files.</span>

<span class="sd">    :param file_names: contains the name of the input file for each sample.</span>
<span class="sd">    :param options: the options.</span>

<span class="sd">    :type file_names: dict</span>
<span class="sd">    :type options: argparse.Namespace</span>

<span class="sd">    Plots the BAF (B Allele Frequency) and LRR (Log R Ratio) of each samples.</span>
<span class="sd">    Only the sexual chromosome are shown.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># importing important stuff</span>
    <span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span> <span class="o">!=</span> <span class="s2">&quot;X11&quot;</span> <span class="ow">and</span> <span class="n">mpl</span><span class="o">.</span><span class="n">get_backend</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;agg&quot;</span><span class="p">:</span>
        <span class="n">mpl</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s2">&quot;Agg&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span> <span class="o">!=</span> <span class="s2">&quot;X11&quot;</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ioff</span><span class="p">()</span>

    <span class="c1"># For each of the sample/files</span>
    <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">file_name</span> <span class="ow">in</span> <span class="n">file_names</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Reading the file</span>
        <span class="n">open_func</span> <span class="o">=</span> <span class="nb">open</span>
        <span class="k">if</span> <span class="n">file_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">):</span>
            <span class="n">open_func</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span>
        <span class="k">with</span> <span class="n">open_func</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">header_index</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span>
                <span class="p">(</span><span class="n">col_name</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">col_name</span> <span class="ow">in</span>
                <span class="nb">enumerate</span><span class="p">(</span><span class="n">input_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>
            <span class="p">])</span>
            <span class="k">for</span> <span class="n">col_name</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;Chr&quot;</span><span class="p">,</span> <span class="s2">&quot;Position&quot;</span><span class="p">,</span> <span class="s2">&quot;B Allele Freq&quot;</span><span class="p">,</span>
                             <span class="s2">&quot;Log R Ratio&quot;</span><span class="p">}:</span>
                <span class="k">if</span> <span class="n">col_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header_index</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: no column named </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">col_name</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

            <span class="c1"># Reading the dat</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\r\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="c1"># We only need X and Y chromosomes</span>
                <span class="n">chromosome</span> <span class="o">=</span> <span class="n">encode_chromosome</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">header_index</span><span class="p">[</span><span class="s2">&quot;Chr&quot;</span><span class="p">]])</span>
                <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;Y&quot;</span><span class="p">}:</span>
                    <span class="k">continue</span>

                <span class="c1"># The position</span>
                <span class="n">position</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">header_index</span><span class="p">[</span><span class="s2">&quot;Position&quot;</span><span class="p">]]</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">position</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">position</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: impossible position </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span>
                                                              <span class="n">position</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

                <span class="c1"># The BAF</span>
                <span class="n">baf</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">header_index</span><span class="p">[</span><span class="s2">&quot;B Allele Freq&quot;</span><span class="p">]]</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">baf</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">baf</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: impossible baf </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">baf</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

                <span class="c1"># The LRR</span>
                <span class="n">lrr</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">header_index</span><span class="p">[</span><span class="s2">&quot;Log R Ratio&quot;</span><span class="p">]]</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">lrr</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">lrr</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: impossible lrr </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">lrr</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

                <span class="c1"># Saving the data</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">lrr</span><span class="p">,</span> <span class="n">baf</span><span class="p">))</span>

        <span class="c1"># Creating the numpy array</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;a1&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
                                     <span class="p">(</span><span class="s2">&quot;lrr&quot;</span><span class="p">,</span> <span class="nb">float</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;baf&quot;</span><span class="p">,</span> <span class="nb">float</span><span class="p">)])</span>

        <span class="c1"># Creating the figure and axes</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">0.97</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>

        <span class="c1"># Setting subplot properties</span>
        <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">():</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s2">&quot;bottom&quot;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;bottom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_position</span><span class="p">((</span><span class="s2">&quot;outward&quot;</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;left&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_position</span><span class="p">((</span><span class="s2">&quot;outward&quot;</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>

        <span class="c1"># Separating the axes</span>
        <span class="n">x_lrr_ax</span><span class="p">,</span> <span class="n">x_baf_ax</span><span class="p">,</span> <span class="n">y_lrr_ax</span><span class="p">,</span> <span class="n">y_baf_ax</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">order</span><span class="o">=</span><span class="s1">&#39;F&#39;</span><span class="p">)</span>

        <span class="c1"># Printing the X chromosome</span>
        <span class="n">curr_chr</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;X&quot;</span>
        <span class="n">x_lrr_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">]</span><span class="o">/</span><span class="mf">1000000.0</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;lrr&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">],</span>
                      <span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">ms</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">mec</span><span class="o">=</span><span class="s2">&quot;#0099CC&quot;</span><span class="p">,</span>
                      <span class="n">mfc</span><span class="o">=</span><span class="s2">&quot;#0099CC&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_clip_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">x_baf_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">]</span><span class="o">/</span><span class="mf">1000000.0</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;baf&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">],</span>
                      <span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">ms</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">mec</span><span class="o">=</span><span class="s2">&quot;#669900&quot;</span><span class="p">,</span>
                      <span class="n">mfc</span><span class="o">=</span><span class="s2">&quot;#669900&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_clip_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">x_lrr_ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#000000&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">1.2</span><span class="p">)</span>
        <span class="n">x_baf_ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#000000&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">1.2</span><span class="p">)</span>
        <span class="n">x_lrr_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;LRR&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
        <span class="n">x_baf_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;BAF&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
        <span class="n">x_baf_ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Position (Mb)&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
        <span class="n">x_lrr_ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Chromosome X&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>

        <span class="c1"># Printing the X chromosome</span>
        <span class="n">curr_chr</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Y&quot;</span>
        <span class="n">y_lrr_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">]</span><span class="o">/</span><span class="mf">1000000.0</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;lrr&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">],</span>
                      <span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">ms</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">mec</span><span class="o">=</span><span class="s2">&quot;#0099CC&quot;</span><span class="p">,</span>
                      <span class="n">mfc</span><span class="o">=</span><span class="s2">&quot;#0099CC&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_clip_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">y_baf_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">]</span><span class="o">/</span><span class="mf">1000000.0</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;baf&quot;</span><span class="p">][</span><span class="n">curr_chr</span><span class="p">],</span>
                      <span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">ms</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">mec</span><span class="o">=</span><span class="s2">&quot;#669900&quot;</span><span class="p">,</span>
                      <span class="n">mfc</span><span class="o">=</span><span class="s2">&quot;#669900&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_clip_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">y_lrr_ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#000000&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">1.2</span><span class="p">)</span>
        <span class="n">y_baf_ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#000000&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">1.2</span><span class="p">)</span>
        <span class="n">y_lrr_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;LRR&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
        <span class="n">y_baf_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;BAF&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
        <span class="n">y_baf_ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Position (Mb)&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
        <span class="n">y_lrr_ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Chromosome Y&quot;</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>

        <span class="c1"># Saving the figure</span>
        <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span> <span class="o">==</span> <span class="s2">&quot;X11&quot;</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span>
                <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_lrr_baf.</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span><span class="p">),</span>
                <span class="n">dpi</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">dpi</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="c1"># Closing the figure</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span></div>


<div class="viewcode-block" id="checkArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.checkArgs">[docs]</a><span class="k">def</span> <span class="nf">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    :param args: an object containing the options of the program.</span>

<span class="sd">    :type args: argparse.Namespace</span>

<span class="sd">    :returns: ``True`` if everything was OK.</span>

<span class="sd">    If there is a problem with an option, an exception is raised using the</span>
<span class="sd">    :py:class:`ProgramError` class, a message is printed to the</span>
<span class="sd">    :class:`sys.stderr` and the program exists with code 1.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Checking the input file</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">problematic_samples</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">problematic_samples</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Checking the raw directory</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">raw_dir</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: no such directory&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">raw_dir</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Checking the DPI value</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">dpi</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: DPI too low&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">dpi</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">return</span> <span class="kc">True</span></div>


<div class="viewcode-block" id="parseArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.parseArgs">[docs]</a><span class="k">def</span> <span class="nf">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>  <span class="c1"># pragma: no cover</span>
    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    :param argString: the options.</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    :returns: A :py:class:`argparse.Namespace` object created by the</span>
<span class="sd">              :py:mod:`argparse` module. It contains the values of the</span>
<span class="sd">              different options.</span>

<span class="sd">    ========================= ====== ==========================================</span>
<span class="sd">              Options          Type                     Description</span>
<span class="sd">    ========================= ====== ==========================================</span>
<span class="sd">    ``--problematic-samples`` string The list of sample with sex problems to</span>
<span class="sd">                                     plot</span>
<span class="sd">    ``--use-full-ids``        bool   Use full sample IDs (famID and indID).</span>
<span class="sd">    ``--full-ids-delimiter``  string The delimiter between famID and indID.</span>
<span class="sd">    ``--raw-dir``             string Directory containing information about</span>
<span class="sd">                                     every samples (BAF and LRR).</span>
<span class="sd">    ``--format``              string The output file format (png, ps, pdf, or</span>
<span class="sd">                                     X11).</span>
<span class="sd">    ``--out``                 string The prefix of the output files.</span>
<span class="sd">    ========================= ====== ==========================================</span>

<span class="sd">    .. note::</span>
<span class="sd">        No option check is done here (except for the one automatically done by</span>
<span class="sd">        argparse). Those need to be done elsewhere (see :py:func:`checkArgs`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">argString</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">args</span></div>


<div class="viewcode-block" id="ProgramError"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.ProgramError">[docs]</a><span class="k">class</span> <span class="nc">ProgramError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;An :py:class:`Exception` raised in case of a problem.</span>

<span class="sd">    :param msg: the message to print to the user before exiting.</span>

<span class="sd">    :type msg: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Construction of the :py:class:`ProgramError` class.</span>

<span class="sd">        :param msg: the message to print to the user</span>

<span class="sd">        :type msg: str</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">message</span></div>

<span class="c1"># The parser object</span>
<span class="n">desc</span> <span class="o">=</span> <span class="s2">&quot;Plots the BAF and LRR of problematic samples.&quot;</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-v&quot;</span><span class="p">,</span> <span class="s2">&quot;--version&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;version&quot;</span><span class="p">,</span>
                    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;pyGenClean version </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>

<span class="c1"># The INPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Input File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--problematic-samples&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;A file containing the list of samples with sex &quot;</span>
                         <span class="s2">&quot;problems (family and individual ID required, &quot;</span>
                         <span class="s2">&quot;separated by a single tabulation). Uses &quot;</span>
                         <span class="s2">&quot;only the individual ID by default, unless &quot;</span>
                         <span class="s2">&quot;--use-full-ids is used.&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--use-full-ids&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;Use this options to use full sample IDs (famID and &quot;</span>
                         <span class="s2">&quot;indID). Otherwise, only the indID will be use.&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--full-ids-delimiter&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;CHAR&quot;</span><span class="p">,</span>
                   <span class="n">default</span><span class="o">=</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The delimiter between famID and indID &quot;</span>
                                      <span class="s2">&quot;for the raw file names. [default: &quot;</span>
                                      <span class="s2">&quot;</span><span class="si">%(default)s</span><span class="s2">]&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--raw-dir&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;DIR&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;Directory containing information about every &quot;</span>
                         <span class="s2">&quot;samples (BAF and LRR).&quot;</span><span class="p">))</span>
<span class="c1"># The options</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Options&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--format&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FORMAT&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;png&quot;</span><span class="p">,</span>
                   <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;png&quot;</span><span class="p">,</span> <span class="s2">&quot;ps&quot;</span><span class="p">,</span> <span class="s2">&quot;pdf&quot;</span><span class="p">,</span> <span class="s2">&quot;X11&quot;</span><span class="p">],</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The output file format (png, ps, pdf, or X11 &quot;</span>
                         <span class="s2">&quot;formats are available). [default: </span><span class="si">%(default)s</span><span class="s2">]&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--dpi&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;DPI&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The pixel density of the figure(s) (DPI). &quot;</span>
                         <span class="s2">&quot;[default: </span><span class="si">%(default)d</span><span class="s2">]&quot;</span><span class="p">))</span>
<span class="c1"># The OUTPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Output File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">default</span><span class="o">=</span><span class="s2">&quot;sexcheck&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The prefix of the output files. [default: &quot;</span>
                         <span class="s2">&quot;</span><span class="si">%(default)s</span><span class="s2">]&quot;</span><span class="p">))</span>


<div class="viewcode-block" id="safe_main"><a class="viewcode-back" href="../../../module_content/pyGenClean.SexCheck.html#pyGenClean.SexCheck.baf_lrr_plot.safe_main">[docs]</a><span class="k">def</span> <span class="nf">safe_main</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;A safe version of the main function (that catches ProgramError).&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">main</span><span class="p">()</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">safe_main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.3 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" >pyGenClean</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>