<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pyGenClean.Misc package &#8212; pyGenClean 1.8.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.8.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="pyGenClean.NoCallHetero package" href="pyGenClean.NoCallHetero.html" />
    <link rel="prev" title="pyGenClean.MarkerMissingness package" href="pyGenClean.MarkerMissingness.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="../index.html"><img src="../_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pyGenClean.NoCallHetero.html" title="pyGenClean.NoCallHetero package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pyGenClean.MarkerMissingness.html" title="pyGenClean.MarkerMissingness package"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pyGenClean 1.8.3 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="modules.html" >Main pyGenClean pipeline</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="pyGenClean.html" accesskey="U">pyGenClean package</a> &#187;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pyGenClean.Misc package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-pyGenClean.Misc.compare_gold_standard">pyGenClean.Misc.compare_gold_standard module</a></li>
<li><a class="reference internal" href="#module-pyGenClean.Misc">Module contents</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pyGenClean.MarkerMissingness.html"
                        title="previous chapter">pyGenClean.MarkerMissingness package</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pyGenClean.NoCallHetero.html"
                        title="next chapter">pyGenClean.NoCallHetero package</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/module_content/pyGenClean.Misc.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pygenclean-misc-package">
<h1>pyGenClean.Misc package<a class="headerlink" href="#pygenclean-misc-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-pyGenClean.Misc.compare_gold_standard">
<span id="pygenclean-misc-compare-gold-standard-module"></span><h2>pyGenClean.Misc.compare_gold_standard module<a class="headerlink" href="#module-pyGenClean.Misc.compare_gold_standard" title="Permalink to this headline">¶</a></h2>
<dl class="exception">
<dt id="pyGenClean.Misc.compare_gold_standard.ProgramError">
<em class="property">exception </em><code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">ProgramError</code><span class="sig-paren">(</span><em>msg</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#ProgramError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.ProgramError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/2/library/exceptions.html#exceptions.Exception" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></code></a></p>
<p>An <code class="xref py py-class docutils literal"><span class="pre">Exception</span></code> raised in case of a problem.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>msg</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the message to print to the user before exiting.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.checkArgs">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">checkArgs</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#checkArgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.checkArgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the arguments and options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/2/library/argparse.html#argparse.Namespace" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">argparse.Namespace</span></code></a>) &#8211; a <code class="xref py py-class docutils literal"><span class="pre">Namespace</span></code> object containing the options of the
program.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="docutils literal"><span class="pre">True</span></code> if everything was OK.</td>
</tr>
</tbody>
</table>
<p>If there is a problem with an option, an exception is raised using the
<a class="reference internal" href="#pyGenClean.Misc.compare_gold_standard.ProgramError" title="pyGenClean.Misc.compare_gold_standard.ProgramError"><code class="xref py py-class docutils literal"><span class="pre">ProgramError</span></code></a> class, a message is printed
to the <code class="xref py py-class docutils literal"><span class="pre">sys.stderr</span></code> and the program exists with code 1.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.check_fam_for_samples">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">check_fam_for_samples</code><span class="sig-paren">(</span><em>required_samples</em>, <em>source</em>, <em>gold</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#check_fam_for_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.check_fam_for_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Check fam files for required_samples.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.computeFrequency">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">computeFrequency</code><span class="sig-paren">(</span><em>prefix</em>, <em>outPrefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#computeFrequency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.computeFrequency" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the frequency using Plink.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.compute_statistics">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">compute_statistics</code><span class="sig-paren">(</span><em>out_dir</em>, <em>gold_prefix</em>, <em>source_prefix</em>, <em>same_samples</em>, <em>use_sge</em>, <em>final_out_prefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#compute_statistics"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.compute_statistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the statistics.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.exclude_SNPs_samples">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">exclude_SNPs_samples</code><span class="sig-paren">(</span><em>inPrefix</em>, <em>outPrefix</em>, <em>exclusionSNP=None</em>, <em>keepSample=None</em>, <em>transpose=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#exclude_SNPs_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.exclude_SNPs_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Exclude some SNPs and keep some samples using Plink.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.extractSNPs">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">extractSNPs</code><span class="sig-paren">(</span><em>prefixes</em>, <em>snpToExtractFileNames</em>, <em>outPrefixes</em>, <em>runSGE</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#extractSNPs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.extractSNPs" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a list of SNPs using Plink.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.findFlippedSNPs">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">findFlippedSNPs</code><span class="sig-paren">(</span><em>goldFrqFile1</em>, <em>sourceAlleles</em>, <em>outPrefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#findFlippedSNPs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.findFlippedSNPs" title="Permalink to this definition">¶</a></dt>
<dd><p>Find flipped SNPs and flip them in the data1.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.findOverlappingSNPsWithGoldStandard">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">findOverlappingSNPsWithGoldStandard</code><span class="sig-paren">(</span><em>prefix</em>, <em>gold_prefixe</em>, <em>out_prefix</em>, <em>use_marker_names=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#findOverlappingSNPsWithGoldStandard"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.findOverlappingSNPsWithGoldStandard" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the overlapping SNPs in 4 different data sets.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.flipSNPs">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">flipSNPs</code><span class="sig-paren">(</span><em>inPrefix</em>, <em>outPrefix</em>, <em>flipFileName</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#flipSNPs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.flipSNPs" title="Permalink to this definition">¶</a></dt>
<dd><p>Flip SNPs using Plink.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.illumina_to_snp">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">illumina_to_snp</code><span class="sig-paren">(</span><em>strand</em>, <em>snp</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#illumina_to_snp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.illumina_to_snp" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the TOP strand of the marker.</p>
<p>Function that takes a strand (TOP or BOT) and a SNP (e.g. : [A/C]) and
returns a space separated AlleleA[space]AlleleB string.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>strand</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Either &#8220;TOP&#8221; or &#8220;BOT&#8221;</li>
<li><strong>snp</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; [A/C], [A/T], [G/C], [T/C], [A/G], [C/G], [T/A] or [T/G].</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The nucleotide for allele A and the nucleotide for allele B
(space separated)</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)">str</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.keepSamples">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">keepSamples</code><span class="sig-paren">(</span><em>prefixes</em>, <em>samplesToExtractFileNames</em>, <em>outPrefixes</em>, <em>runSGE</em>, <em>transpose=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#keepSamples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.keepSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a list of SNPs using Plink.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.main">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">main</code><span class="sig-paren">(</span><em>argString=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.parseArgs">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">parseArgs</code><span class="sig-paren">(</span><em>argString=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#parseArgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.parseArgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the command line options and arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A <code class="xref py py-class docutils literal"><span class="pre">numpy.Namespace</span></code> object created by
the <a class="reference external" href="https://docs.python.org/2/library/argparse.html#module-argparse" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">argparse</span></code></a> module. It contains the values of the
different options.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="30%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Options</th>
<th class="head">Type</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--bfile</span></code></td>
<td>string</td>
<td>The input file prefix (will find the plink
binary files by appending the prefix to the
<code class="docutils literal"><span class="pre">.bim</span></code>, <code class="docutils literal"><span class="pre">.bed</span></code> and <code class="docutils literal"><span class="pre">.fam</span></code> files,
respectively).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--gold-bfile</span></code></td>
<td>string</td>
<td>The input file prefix (will find the plink
binary files by appending the prefix to the
<code class="docutils literal"><span class="pre">.bim</span></code>, <code class="docutils literal"><span class="pre">.bed</span></code> and <code class="docutils literal"><span class="pre">.fam</span></code> files,
respectively) for the Gold Standard.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--same-samples</span></code></td>
<td>string</td>
<td>A file containing samples which are present
in both the gold standard and the source
panel. One line by identity and tab
separated. For each row, first sample is
Gold Standard, second is source panel.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--source-manifest</span></code></td>
<td>string</td>
<td>The illumina marker manifest. This file
should have tabs as field separator. There
should be no lines before the main header
line. There should be no lines after the
last data line.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--source-alleles</span></code></td>
<td>string</td>
<td>A file containing the source alleles (TOP).
Two columns (separated by tabulation, one
with the marker name, the other with the
alleles (separated by space). No header.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--sge</span></code></td>
<td>boolean</td>
<td>Use SGE for parallelization.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--do-not-flip</span></code></td>
<td>boolean</td>
<td>Do not flip SNPs. WARNING: only use this
option only if the Gold Standard was
generated using the same chip (hence,
flipping is unnecessary).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--use-marker-names</span></code></td>
<td>boolean</td>
<td>Use marker names instead of (chr, position).
WARNING: only use this options only if the
Gold Standard was generated using the same
chip (hence, they have the same marker
names).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--out</span></code></td>
<td>string</td>
<td>The prefix of the output files.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">No option check is done here (except for the one automatically done by
argparse). Those need to be done elsewhere (see <a class="reference internal" href="#pyGenClean.Misc.compare_gold_standard.checkArgs" title="pyGenClean.Misc.compare_gold_standard.checkArgs"><code class="xref py py-func docutils literal"><span class="pre">checkArgs()</span></code></a>).</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.read_same_samples_file">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">read_same_samples_file</code><span class="sig-paren">(</span><em>filename</em>, <em>out_prefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#read_same_samples_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.read_same_samples_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a file containing same samples.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.read_source_alleles">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">read_source_alleles</code><span class="sig-paren">(</span><em>file_name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#read_source_alleles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.read_source_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads an allele file.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.read_source_manifest">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">read_source_manifest</code><span class="sig-paren">(</span><em>file_name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#read_source_manifest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.read_source_manifest" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads Illumina manifest.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.renameSNPs">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">renameSNPs</code><span class="sig-paren">(</span><em>inPrefix</em>, <em>updateFileName</em>, <em>outPrefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#renameSNPs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.renameSNPs" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the name of the SNPs using Plink.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.runCommand">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">runCommand</code><span class="sig-paren">(</span><em>command</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#runCommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.runCommand" title="Permalink to this definition">¶</a></dt>
<dd><p>Run a command.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.Misc.compare_gold_standard.safe_main">
<code class="descclassname">pyGenClean.Misc.compare_gold_standard.</code><code class="descname">safe_main</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/Misc/compare_gold_standard.html#safe_main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.Misc.compare_gold_standard.safe_main" title="Permalink to this definition">¶</a></dt>
<dd><p>A safe version of the main function (that catches ProgramError).</p>
</dd></dl>

</div>
<div class="section" id="module-pyGenClean.Misc">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pyGenClean.Misc" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pyGenClean.NoCallHetero.html" title="pyGenClean.NoCallHetero package"
             >next</a> |</li>
        <li class="right" >
          <a href="pyGenClean.MarkerMissingness.html" title="pyGenClean.MarkerMissingness package"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pyGenClean 1.8.3 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="modules.html" >Main pyGenClean pipeline</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="pyGenClean.html" >pyGenClean package</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>