<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pyGenClean.DupSNPs package &mdash; pyGenClean 1.8.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.8.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.8.2 documentation" href="../index.html" />
    <link rel="up" title="pyGenClean package" href="pyGenClean.html" />
    <link rel="next" title="pyGenClean.DupSamples package" href="pyGenClean.DupSamples.html" />
    <link rel="prev" title="pyGenClean.Contamination package" href="pyGenClean.Contamination.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="../index.html"><img src="../_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pyGenClean.DupSamples.html" title="pyGenClean.DupSamples package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pyGenClean.Contamination.html" title="pyGenClean.Contamination package"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="modules.html" >Main pyGenClean pipeline</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="pyGenClean.html" accesskey="U">pyGenClean package</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pyGenClean.DupSNPs package</a><ul>
<li><a class="reference internal" href="#module-pyGenClean.DupSNPs">Module contents</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a><ul>
<li><a class="reference internal" href="#module-pyGenClean.DupSNPs.duplicated_snps">pyGenClean.DupSNPs.duplicated_snps module</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pyGenClean.Contamination.html"
                        title="previous chapter">pyGenClean.Contamination package</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pyGenClean.DupSamples.html"
                        title="next chapter">pyGenClean.DupSamples package</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/module_content/pyGenClean.DupSNPs.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pygenclean-dupsnps-package">
<h1>pyGenClean.DupSNPs package<a class="headerlink" href="#pygenclean-dupsnps-package" title="Permalink to this headline">¶</a></h1>
<p>For more information about how to use this module, refer to the
<a class="reference internal" href="../duplicated_markers.html#dup-markers-label"><span>Duplicated Markers Module</span></a>.</p>
<div class="section" id="module-pyGenClean.DupSNPs">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pyGenClean.DupSNPs" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
<div class="section" id="module-pyGenClean.DupSNPs.duplicated_snps">
<span id="pygenclean-dupsnps-duplicated-snps-module"></span><h3>pyGenClean.DupSNPs.duplicated_snps module<a class="headerlink" href="#module-pyGenClean.DupSNPs.duplicated_snps" title="Permalink to this headline">¶</a></h3>
<dl class="exception">
<dt id="pyGenClean.DupSNPs.duplicated_snps.ProgramError">
<em class="property">exception </em><code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">ProgramError</code><span class="sig-paren">(</span><em>msg</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#ProgramError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.ProgramError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/exceptions.html#exceptions.Exception" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></code></a></p>
<p>An <code class="xref py py-class docutils literal"><span class="pre">Exception</span></code> raised in case of a problem.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>msg</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the message to print to the user before exiting.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.checkArgs">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">checkArgs</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#checkArgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.checkArgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the arguments and options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="http://docs.python.org/library/argparse.html#argparse.Namespace" title="(in Python v2.7)"><em>argparse.Namespace</em></a>) &#8211; an object containing the options of the program.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><code class="docutils literal"><span class="pre">True</span></code> if everything was OK.</td>
</tr>
</tbody>
</table>
<p>If there is a problem with an option, an exception is raised using the
<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.ProgramError" title="pyGenClean.DupSNPs.duplicated_snps.ProgramError"><code class="xref py py-class docutils literal"><span class="pre">ProgramError</span></code></a> class, a message is printed to the
<code class="xref py py-class docutils literal"><span class="pre">sys.stderr</span></code> and the program exists with code 1.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.chooseBestSnps">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">chooseBestSnps</code><span class="sig-paren">(</span><em>tped</em>, <em>snps</em>, <em>trueCompletion</em>, <em>trueConcordance</em>, <em>prefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#chooseBestSnps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.chooseBestSnps" title="Permalink to this definition">¶</a></dt>
<dd><p>Choose the best duplicates according to the completion and concordance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>tped</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tped</span></code> of duplicated markers.</li>
<li><strong>snps</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the position of the duplicated markers in the <code class="docutils literal"><span class="pre">tped</span></code>.</li>
<li><strong>trueCompletion</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; the completion of each markers.</li>
<li><strong>trueConcordance</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the pairwise concordance of each markers.</li>
<li><strong>prefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of the output files.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple containing the chosen indexes (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a>) as the
first element, the completion (<code class="xref py py-class docutils literal"><span class="pre">numpy.array</span></code>) as the
second element, and the concordance (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a>) as last
element.</p>
</td>
</tr>
</tbody>
</table>
<p>It creates two output files: <code class="docutils literal"><span class="pre">prefix.chosen_snps.info</span></code> and
<code class="docutils literal"><span class="pre">prefix.not_chosen_snps.info</span></code>. The first one contains the markers that
were chosen for completion, and the second one, the markers that weren&#8217;t.</p>
<p>It starts by computing the completion of each markers (dividing the number
of calls divided by the total number of genotypes). Then, for each of the
duplicated markers, we choose the best one according to completion and
concordance (see explanation in
<code class="xref py py-func docutils literal"><span class="pre">DupSamples.duplicated_samples.chooseBestDuplicates()</span></code> for more
details).</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.computeFrequency">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">computeFrequency</code><span class="sig-paren">(</span><em>prefix</em>, <em>outPrefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#computeFrequency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.computeFrequency" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the frequency of the SNPs using Plink.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of the input files.</li>
<li><strong>outPrefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of the output files.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a <a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> containing the frequency of each marker.</p>
</td>
</tr>
</tbody>
</table>
<p>Start by computing the frequency of all markers using Plink. Then, it reads
the output file, and saves the frequency and allele information.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.computeStatistics">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">computeStatistics</code><span class="sig-paren">(</span><em>tped</em>, <em>tfam</em>, <em>snps</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#computeStatistics"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.computeStatistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the completion and concordance of each SNPs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>tped</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tped</span></code>.</li>
<li><strong>tfam</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tfam</span></code></li>
<li><strong>snps</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the position of the duplicated markers in the <code class="docutils literal"><span class="pre">tped</span></code>.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple containing the completion of duplicated markers
(<code class="xref py py-class docutils literal"><span class="pre">numpy.array</span></code>) as first element, and the concordance
(<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a>) of duplicated markers, as last element.</p>
</td>
</tr>
</tbody>
</table>
<p>A marker&#8217;s completion is compute using this formula (where <span class="math">\(G_i\)</span> is
the set of genotypes for the marker <span class="math">\(i\)</span>):</p>
<div class="math">
\[Completion_i = \frac{||g \in G_i \textrm{ where } g \neq 0||}
                     {||G_i||}\]</div>
<p>The pairwise concordance between duplicated markers is compute as follow
(where <span class="math">\(G_i\)</span> and <span class="math">\(G_j\)</span> are the sets of genotypes for markers
<span class="math">\(i\)</span> and <span class="math">\(j\)</span>, respectively):</p>
<div class="math">
\[Concordance_{i,j} = \frac{
    ||g \in G_i \cup G_j \textrm{ where } g_i = g_j \neq 0||
}{
    ||g \in G_i \cup G_j \textrm{ where } g \neq 0||
}\]</div>
<p>Hence, we only computes the numerators and denominators of the completion
and concordance, for future reference.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When the genotypes are not comparable, the function tries to flip one
of the genotype to see if it becomes comparable.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.createAndCleanTPED">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">createAndCleanTPED</code><span class="sig-paren">(</span><em>tped</em>, <em>tfam</em>, <em>snps</em>, <em>prefix</em>, <em>chosenSNPs</em>, <em>completion</em>, <em>concordance</em>, <em>snpsToComplete</em>, <em>tfamFileName</em>, <em>completionT</em>, <em>concordanceT</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#createAndCleanTPED"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.createAndCleanTPED" title="Permalink to this definition">¶</a></dt>
<dd><p>Complete a TPED for duplicated SNPs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>tped</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tped</span></code> of duplicated markers.</li>
<li><strong>tfam</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tfam</span></code>.</li>
<li><strong>snps</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the position of duplicated markers in the <code class="docutils literal"><span class="pre">tped</span></code>.</li>
<li><strong>prefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of the output files.</li>
<li><strong>chosenSNPs</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the markers that were chosen for completion (including
problems).</li>
<li><strong>completion</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; the completion of each of the duplicated markers.</li>
<li><strong>concordance</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the pairwise concordance of the duplicated markers.</li>
<li><strong>snpsToComplete</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#set" title="(in Python v2.7)"><em>set</em></a>) &#8211; the markers that will be completed (excluding
problems).</li>
<li><strong>tfamFileName</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the name of the original <code class="docutils literal"><span class="pre">tfam</span></code> file.</li>
<li><strong>completionT</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; the completion threshold.</li>
<li><strong>concordanceT</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; the concordance threshold.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple containing the new <code class="docutils literal"><span class="pre">tped</span></code> after completion
(<code class="xref py py-class docutils literal"><span class="pre">numpy.array</span></code> as the first element, and the index of
the markers that will need to be rid of (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#set" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">set</span></code></a>) as the
last element.</p>
</td>
</tr>
</tbody>
</table>
<p>It creates three different files:</p>
<ul>
<li><dl class="first docutils">
<dt><code class="docutils literal"><span class="pre">prefix.zeroed_out</span></code>: contains information about markers and samples</dt>
<dd><p class="first last">where the genotyped was zeroed out.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><code class="docutils literal"><span class="pre">prefix.not_good_enough</span></code>: contains information about markers that were</dt>
<dd><p class="first last">not good enough to help in completing the
chosen markers (because of concordance or
completion).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><code class="docutils literal"><span class="pre">prefix.removed_duplicates</span></code>: the list of markers that where used for</dt>
<dd><p class="first last">completing the chosen one, hence they will
be removed from the final data
set.</p>
</dd>
</dl>
</li>
</ul>
<p>Cycling through every genotypes of every samples of every duplicated
markers, checks if the genotypes are all the same. If the chosen one was
not called, but the other ones were, then we complete the chosen one with
the genotypes for the others (assuming that they are all the same). If
there is a difference between the genotypes, it is zeroed out for the
chosen marker.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.createFinalTPEDandTFAM">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">createFinalTPEDandTFAM</code><span class="sig-paren">(</span><em>tped</em>, <em>toReadPrefix</em>, <em>prefix</em>, <em>snpToRemove</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#createFinalTPEDandTFAM"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.createFinalTPEDandTFAM" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates the final TPED and TFAM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>tped</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tped</span></code> of duplicated markers.</li>
<li><strong>toReadPrefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of the unique files.</li>
<li><strong>prefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of the output files.</li>
<li><strong>snpToRemove</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#set" title="(in Python v2.7)"><em>set</em></a>) &#8211; the markers to remove.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Starts by copying the unique markers&#8217; <code class="docutils literal"><span class="pre">tfam</span></code> file to
<code class="docutils literal"><span class="pre">prefix.final.tfam</span></code>. Then, it copies the unique markers&#8217; <code class="docutils literal"><span class="pre">tped</span></code> file,
in which the chosen markers will be appended.</p>
<p>The final data set will include the unique markers, the chosen markers
which were completed, and the problematic duplicated markers (for further
analysis). The markers that were used to complete the chosen ones are not
present in the final data set.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.findUniques">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">findUniques</code><span class="sig-paren">(</span><em>mapF</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#findUniques"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.findUniques" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the unique markers in a MAP.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>mapF</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; representation of a <code class="docutils literal"><span class="pre">map</span></code> file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a <a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> containing unique markers (according to their
genomic localisation).</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.flipGenotype">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">flipGenotype</code><span class="sig-paren">(</span><em>genotype</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#flipGenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.flipGenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Flips a genotype.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>genotype</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#set" title="(in Python v2.7)"><em>set</em></a>) &#8211; the genotype to flip.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the new flipped genotype (as a <a class="reference external" href="http://docs.python.org/library/stdtypes.html#set" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">set</span></code></a>)</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">flipGenotype</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">})</span>
<span class="go">set([&#39;A&#39;, &#39;T&#39;])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">flipGenotype</span><span class="p">({</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">})</span>
<span class="go">set([&#39;A&#39;, &#39;G&#39;])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">flipGenotype</span><span class="p">({</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">})</span>
<span class="go">set([&#39;A&#39;, &#39;C&#39;])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">flipGenotype</span><span class="p">({</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">})</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ProgramError</span>: <span class="n">0: unkown allele</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">flipGenotype</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;N&quot;</span><span class="p">})</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ProgramError</span>: <span class="n">N: unkown allele</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.getIndexOfHeteroMen">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">getIndexOfHeteroMen</code><span class="sig-paren">(</span><em>genotypes</em>, <em>menIndex</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#getIndexOfHeteroMen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.getIndexOfHeteroMen" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the indexes of heterozygous men.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>genotypes</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; the genotypes of everybody.</li>
<li><strong>menIndex</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; the indexes of the men (for the genotypes).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a <code class="xref py py-class docutils literal"><span class="pre">numpy.array</span></code> containing the indexes of the genotypes
to remove.</p>
</td>
</tr>
</tbody>
</table>
<p>Finds the mean that have a heterozygous genotype for this current marker.
Usually used on sexual chromosomes.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.main">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">main</code><span class="sig-paren">(</span><em>argString=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.main" title="Permalink to this definition">¶</a></dt>
<dd><p>The main function of the module..</p>
<p>Here are the steps for duplicated samples:</p>
<ol class="arabic simple">
<li>Prints the options.</li>
<li>Reads the <code class="docutils literal"><span class="pre">map</span></code> file to gather marker&#8217;s position
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.readMAP" title="pyGenClean.DupSNPs.duplicated_snps.readMAP"><code class="xref py py-func docutils literal"><span class="pre">readMAP()</span></code></a>).</li>
<li>Reads the <code class="docutils literal"><span class="pre">tfam</span></code> file (<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.readTFAM" title="pyGenClean.DupSNPs.duplicated_snps.readTFAM"><code class="xref py py-func docutils literal"><span class="pre">readTFAM()</span></code></a>).</li>
<li>Finds the unique markers using the <code class="docutils literal"><span class="pre">map</span></code> file
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.findUniques" title="pyGenClean.DupSNPs.duplicated_snps.findUniques"><code class="xref py py-func docutils literal"><span class="pre">findUniques()</span></code></a>).</li>
<li>Process the <code class="docutils literal"><span class="pre">tped</span></code> file. Write a file containing unique markers in
<code class="docutils literal"><span class="pre">prefix.unique_snps</span></code> (<code class="docutils literal"><span class="pre">tfam</span></code> and <code class="docutils literal"><span class="pre">tped</span></code>). Keep in memory
information about the duplicated markers (<code class="docutils literal"><span class="pre">tped</span></code>)
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.processTPED" title="pyGenClean.DupSNPs.duplicated_snps.processTPED"><code class="xref py py-func docutils literal"><span class="pre">processTPED()</span></code></a>).</li>
<li>If there are no duplicated markers, the file <code class="docutils literal"><span class="pre">prefix.unique_snps</span></code>
(<code class="docutils literal"><span class="pre">tped</span></code> and <code class="docutils literal"><span class="pre">tfam</span></code>) are copied to <code class="docutils literal"><span class="pre">prefix.final</span></code>.</li>
<li>If there are duplicated markers, print a <code class="docutils literal"><span class="pre">tped</span></code> and <code class="docutils literal"><span class="pre">tfam</span></code> file
containing the duplicated markers
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.printDuplicatedTPEDandTFAM" title="pyGenClean.DupSNPs.duplicated_snps.printDuplicatedTPEDandTFAM"><code class="xref py py-func docutils literal"><span class="pre">printDuplicatedTPEDandTFAM()</span></code></a>).</li>
<li>Computes the frequency of the duplicated markers (using Plink) and read
the output file (<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.computeFrequency" title="pyGenClean.DupSNPs.duplicated_snps.computeFrequency"><code class="xref py py-func docutils literal"><span class="pre">computeFrequency()</span></code></a>).</li>
<li>Computes the concordance and pairwise completion of each of the
duplicated markers (<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.computeStatistics" title="pyGenClean.DupSNPs.duplicated_snps.computeStatistics"><code class="xref py py-func docutils literal"><span class="pre">computeStatistics()</span></code></a>).</li>
<li>Prints the problematic duplicated markers with a file containing the
summary of the statistics (completion and pairwise concordance)
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.printProblems" title="pyGenClean.DupSNPs.duplicated_snps.printProblems"><code class="xref py py-func docutils literal"><span class="pre">printProblems()</span></code></a>).</li>
<li>Print the pairwise concordance in a file (matrices)
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.printConcordance" title="pyGenClean.DupSNPs.duplicated_snps.printConcordance"><code class="xref py py-func docutils literal"><span class="pre">printConcordance()</span></code></a>).</li>
<li>Choose the best duplicated markers using concordance and completion
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.chooseBestSnps" title="pyGenClean.DupSNPs.duplicated_snps.chooseBestSnps"><code class="xref py py-func docutils literal"><span class="pre">chooseBestSnps()</span></code></a>).</li>
<li>Completes the chosen markers with the remaining duplicated markers
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.createAndCleanTPED" title="pyGenClean.DupSNPs.duplicated_snps.createAndCleanTPED"><code class="xref py py-func docutils literal"><span class="pre">createAndCleanTPED()</span></code></a>).</li>
<li>Creates the final <code class="docutils literal"><span class="pre">tped</span></code> file, containing the unique markers, the
chosen duplicated markers that were completed, and the problematic
duplicated markers (for further analysis). This set excludes markers
that were used for completing the chosen ones
(<a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.createFinalTPEDandTFAM" title="pyGenClean.DupSNPs.duplicated_snps.createFinalTPEDandTFAM"><code class="xref py py-func docutils literal"><span class="pre">createFinalTPEDandTFAM()</span></code></a>).</li>
</ol>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.parseArgs">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">parseArgs</code><span class="sig-paren">(</span><em>argString=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#parseArgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.parseArgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the command line options and arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>argString</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; the options.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">A <a class="reference external" href="http://docs.python.org/library/argparse.html#argparse.Namespace" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">argparse.Namespace</span></code></a> object created by the
<a class="reference external" href="http://docs.python.org/library/argparse.html#module-argparse" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">argparse</span></code></a> module. It contains the values of the
different options.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="42%" />
<col width="8%" />
<col width="49%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Options</th>
<th class="head">Type</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--tfile</span></code></td>
<td>string</td>
<td>The input file prefix (Plink tfile).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--snp-completion-threshold</span></code></td>
<td>float</td>
<td>The completion threshold to consider
a replicate when choosing the best
replicates.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--snp-concordance-threshold</span></code></td>
<td>float</td>
<td>The concordance threshold to
consider a replicate when choosing
the best replicates.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--frequency_difference</span></code></td>
<td>float</td>
<td>The maximum difference in frequency
between duplicated markers.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--out</span></code></td>
<td>string</td>
<td>The prefix of the output files.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">No option check is done here (except for the one automatically done by
argparse). Those need to be done elsewhere (see <a class="reference internal" href="#pyGenClean.DupSNPs.duplicated_snps.checkArgs" title="pyGenClean.DupSNPs.duplicated_snps.checkArgs"><code class="xref py py-func docutils literal"><span class="pre">checkArgs()</span></code></a>).</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.printConcordance">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">printConcordance</code><span class="sig-paren">(</span><em>concordance</em>, <em>prefix</em>, <em>tped</em>, <em>snps</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#printConcordance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.printConcordance" title="Permalink to this definition">¶</a></dt>
<dd><p>Print the concordance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>concordance</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the concordance.</li>
<li><strong>prefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix if the output files.</li>
<li><strong>tped</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tped</span></code> of duplicated markers.</li>
<li><strong>snps</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the position of the duplicated markers in the <code class="docutils literal"><span class="pre">tped</span></code>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Prints the concordance in a file, in the format of a matrix. For each
duplicated markers, the first line (starting with the <cite>#</cite> signs) contains
the name of all the markers in the duplicated markers set. Then a <span class="math">\(N
\times N\)</span> matrix is printed to file (where <span class="math">\(N\)</span> is the number of
markers in the duplicated marker list), containing the pairwise
concordance.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.printDuplicatedTPEDandTFAM">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">printDuplicatedTPEDandTFAM</code><span class="sig-paren">(</span><em>tped</em>, <em>tfamFileName</em>, <em>outPrefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#printDuplicatedTPEDandTFAM"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.printDuplicatedTPEDandTFAM" title="Permalink to this definition">¶</a></dt>
<dd><p>Print the duplicated SNPs TPED and TFAM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>tped</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tped</span></code> of duplicated markers.</li>
<li><strong>tfamFileName</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the name of the original <code class="docutils literal"><span class="pre">tfam</span></code> file.</li>
<li><strong>outPrefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the output prefix.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>First, it copies the original <code class="docutils literal"><span class="pre">tfam</span></code> into
<code class="docutils literal"><span class="pre">outPrefix.duplicated_snps.tfam</span></code>. Then, it prints the <code class="docutils literal"><span class="pre">tped</span></code> of
duplicated markers in <code class="docutils literal"><span class="pre">outPrefix.duplicated_snps.tped</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.printProblems">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">printProblems</code><span class="sig-paren">(</span><em>completion</em>, <em>concordance</em>, <em>tped</em>, <em>snps</em>, <em>frequencies</em>, <em>prefix</em>, <em>diffFreq</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#printProblems"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.printProblems" title="Permalink to this definition">¶</a></dt>
<dd><p>Print the statistics.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>completion</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; the completion of each duplicated markers.</li>
<li><strong>concordance</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the pairwise concordance between duplicated markers.</li>
<li><strong>tped</strong> (<a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.10)"><em>numpy.array</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">tped</span></code> of duplicated markers.</li>
<li><strong>snps</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the positions of the duplicated markers in the <code class="docutils literal"><span class="pre">tped</span></code></li>
<li><strong>frequencies</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the frequency of each of the duplicated markers.</li>
<li><strong>prefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of the output files.</li>
<li><strong>diffFreq</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; the frequency difference threshold.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a <a class="reference external" href="http://docs.python.org/library/stdtypes.html#set" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">set</span></code></a> containing duplicated markers to complete.</p>
</td>
</tr>
</tbody>
</table>
<p>Creates a summary file (<code class="docutils literal"><span class="pre">prefix.summary</span></code>) containing information about
duplicated markers: chromosome, position, name, alleles, status, completion
percentage, completion number and mean concordance.</p>
<p>The frequency and the minor allele are used to be certain that two
duplicated markers are exactly the same marker (and not a tri-allelic one,
for example).</p>
<p>For each duplicated markers:</p>
<ol class="arabic simple">
<li>Constructs the set of available alleles for the first marker.</li>
<li>Constructs the set of available alleles for the second marker.</li>
<li>If the two sets are different, but the number of alleles is the same, we
try to flip one of the marker. If the two sets are the same, but the
number of alleles is 1, we set the status to <code class="docutils literal"><span class="pre">homo_flip</span></code>. If the
markers are heterozygous, we set the status to <code class="docutils literal"><span class="pre">flip</span></code>.</li>
<li>If there is a difference in the number of alleles (one is homozygous,
the other, heterozygous), and that there is on allele in common, we set
the status to <code class="docutils literal"><span class="pre">homo_hetero</span></code>. If there are no allele in common, we try
to flip one. If the new sets have one allele in common, we set the
status to <code class="docutils literal"><span class="pre">homo_hetero_flip</span></code>.</li>
<li>If the sets of available alleles are the same (without flip), we check
the frequency and the minor alleles. If the minor allele is different,
we set the status to <code class="docutils literal"><span class="pre">diff_minor_allele</span></code>. If the difference in
frequencies is higher than a threshold, we set the status to
<code class="docutils literal"><span class="pre">diff_frequency</span></code>.</li>
<li>If all of the above fail, we set the status to <code class="docutils literal"><span class="pre">problem</span></code>.</li>
</ol>
<p>Problems are written in the <code class="docutils literal"><span class="pre">prefix.problems</span></code> file, and contains the
following columns: chromosome, position, name and status. This file
contains all the markers with a status, as explained above.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.processTPED">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">processTPED</code><span class="sig-paren">(</span><em>uniqueSNPs</em>, <em>mapF</em>, <em>fileName</em>, <em>tfam</em>, <em>prefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#processTPED"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.processTPED" title="Permalink to this definition">¶</a></dt>
<dd><p>Process the TPED file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>uniqueSNPs</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; the unique markers.</li>
<li><strong>mapF</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; a representation of the <code class="docutils literal"><span class="pre">map</span></code> file.</li>
<li><strong>fileName</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the name of the <code class="docutils literal"><span class="pre">tped</span></code> file.</li>
<li><strong>tfam</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the name of the <code class="docutils literal"><span class="pre">tfam</span></code> file.</li>
<li><strong>prefix</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the prefix of all the files.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tuple with the representation of the <code class="docutils literal"><span class="pre">tped</span></code> file
(<code class="xref py py-class docutils literal"><span class="pre">numpy.array</span></code>) as first element, and the updated
position of the duplicated markers in the <code class="docutils literal"><span class="pre">tped</span></code>
representation.</p>
</td>
</tr>
</tbody>
</table>
<p>Copies the <code class="docutils literal"><span class="pre">tfam</span></code> file into <code class="docutils literal"><span class="pre">prefix.unique_snps.tfam</span></code>. While reading
the <code class="docutils literal"><span class="pre">tped</span></code> file, creates a new one (<code class="docutils literal"><span class="pre">prefix.unique_snps.tped</span></code>)
containing only unique markers.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.readMAP">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">readMAP</code><span class="sig-paren">(</span><em>fileName</em>, <em>prefix</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#readMAP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.readMAP" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the MAP file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fileName</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the name of the <code class="docutils literal"><span class="pre">map</span></code> file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a list of tuples, representing the <code class="docutils literal"><span class="pre">map</span></code> file.</td>
</tr>
</tbody>
</table>
<p>While reading the <code class="docutils literal"><span class="pre">map</span></code> file, it saves a file
(<code class="docutils literal"><span class="pre">prefix.duplicated_marker_names</span></code>) containing the name of the unique
duplicated markers.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.readTFAM">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">readTFAM</code><span class="sig-paren">(</span><em>fileName</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#readTFAM"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.readTFAM" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the TFAM file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fileName</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the name of the <code class="docutils literal"><span class="pre">tfam</span></code> file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a representation the <code class="docutils literal"><span class="pre">tfam</span></code> file (<code class="xref py py-class docutils literal"><span class="pre">numpy.array</span></code>).</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.runCommand">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">runCommand</code><span class="sig-paren">(</span><em>command</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#runCommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.runCommand" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the command in Plink.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>command</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; the command to run.</td>
</tr>
</tbody>
</table>
<p>Tries to run a command using <a class="reference external" href="http://docs.python.org/library/subprocess.html#module-subprocess" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">subprocess</span></code></a>.</p>
</dd></dl>

<dl class="function">
<dt id="pyGenClean.DupSNPs.duplicated_snps.safe_main">
<code class="descclassname">pyGenClean.DupSNPs.duplicated_snps.</code><code class="descname">safe_main</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pyGenClean/DupSNPs/duplicated_snps.html#safe_main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pyGenClean.DupSNPs.duplicated_snps.safe_main" title="Permalink to this definition">¶</a></dt>
<dd><p>A safe version of the main function (that catches ProgramError).</p>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pyGenClean.DupSamples.html" title="pyGenClean.DupSamples package"
             >next</a> |</li>
        <li class="right" >
          <a href="pyGenClean.Contamination.html" title="pyGenClean.Contamination package"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pyGenClean 1.8.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="modules.html" >Main pyGenClean pipeline</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="pyGenClean.html" >pyGenClean package</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>