<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pyGenClean.PlateBias.plate_bias &mdash; pyGenClean 1.8.1 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.8.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.8.1 documentation" href="../../../index.html" />
    <link rel="up" title="pyGenClean" href="../../pyGenClean.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="../../../index.html"><img src="../../../_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" accesskey="U">pyGenClean</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for pyGenClean.PlateBias.plate_bias</h1><div class="highlight"><pre>
<span class="ch">#!/usr/bin/env python2.7</span>

<span class="c1"># This file is part of pyGenClean.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is free software: you can redistribute it and/or modify it under</span>
<span class="c1"># the terms of the GNU General Public License as published by the Free Software</span>
<span class="c1"># Foundation, either version 3 of the License, or (at your option) any later</span>
<span class="c1"># version.</span>
<span class="c1">#</span>
<span class="c1"># pyGenClean is distributed in the hope that it will be useful, but WITHOUT ANY</span>
<span class="c1"># WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR</span>
<span class="c1"># A PARTICULAR PURPOSE.  See the GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1"># You should have received a copy of the GNU General Public License along with</span>
<span class="c1"># pyGenClean.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">namedtuple</span>

<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">__version__</span>
<span class="kn">from</span> <span class="nn">..PlinkUtils</span> <span class="kn">import</span> <span class="n">createRowFromPlinkSpacedOutput</span>


<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;plate_bias&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;The main function of this module.</span>

<span class="sd">    :param argString: the options.</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    These are the steps:</span>

<span class="sd">    1. Runs a plate bias analysis using Plink</span>
<span class="sd">       (:py:func:`executePlateBiasAnalysis`).</span>
<span class="sd">    2. Extracts the list of significant markers after plate bias analysis</span>
<span class="sd">       (:py:func:`extractSignificantSNPs`).</span>
<span class="sd">    3. Computes the frequency of all significant markers after plate bias</span>
<span class="sd">       analysis (:py:func:`computeFrequencyOfSignificantSNPs`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Getting and checking the options</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>
    <span class="n">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Options used:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;  --{} {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">),</span> <span class="n">value</span><span class="p">))</span>

    <span class="c1"># Run plink</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running Plink to check the plate bias&quot;</span><span class="p">)</span>
    <span class="n">executePlateBiasAnalysis</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="c1"># Extract significant SNPs</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Extracting significant SNPs&quot;</span><span class="p">)</span>
    <span class="n">assocResults</span> <span class="o">=</span> <span class="n">extractSignificantSNPs</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

    <span class="c1"># Remove significant SNPs using plink</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing frequency of significant SNPs&quot;</span><span class="p">)</span>
    <span class="n">maf</span> <span class="o">=</span> <span class="n">computeFrequencyOfSignificantSNPs</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="c1"># Create the final summary file</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating the summary file&quot;</span><span class="p">)</span>
    <span class="n">createSummaryFile</span><span class="p">(</span><span class="n">assocResults</span><span class="p">,</span> <span class="n">maf</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span></div>


<div class="viewcode-block" id="createSummaryFile"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.createSummaryFile">[docs]</a><span class="k">def</span> <span class="nf">createSummaryFile</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">maf</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Creat the final summary file containing plate bias results.</span>

<span class="sd">    :param results: the list of all the significant results.</span>
<span class="sd">    :param maf: the minor allele frequency of the significant results.</span>
<span class="sd">    :param prefix: the prefix of all the files.</span>

<span class="sd">    :type results: list</span>
<span class="sd">    :type maf: dict</span>
<span class="sd">    :type prefix: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">o_filename</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.significant_SNPs.summary&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">o_filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o_file</span><span class="p">:</span>
            <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">o_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">,</span> <span class="s2">&quot;p&quot;</span><span class="p">,</span>
                                       <span class="s2">&quot;odds&quot;</span><span class="p">,</span> <span class="s2">&quot;plate&quot;</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
                <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">o_file</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="n">maf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;N/A&quot;</span><span class="p">),</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">p</span><span class="p">,</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">odds</span><span class="p">,</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">plate</span><span class="p">,</span>
                <span class="p">))</span>

    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: cannot write file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">o_filename</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="extractSignificantSNPs"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.extractSignificantSNPs">[docs]</a><span class="k">def</span> <span class="nf">extractSignificantSNPs</span><span class="p">(</span><span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract significant SNPs in the fisher file.</span>

<span class="sd">    :param prefix: the prefix of the input file.</span>

<span class="sd">    :type prefix: str</span>

<span class="sd">    Reads a list of significant markers (``prefix.assoc.fisher``) after plate</span>
<span class="sd">    bias analysis with Plink. Writes a file (``prefix.significant_SNPs.txt``)</span>
<span class="sd">    containing those significant markers.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The list of all assoc files</span>
    <span class="n">fileNames</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.*.assoc.fisher&quot;</span><span class="p">)</span>

    <span class="c1"># The object to save an assoc row</span>
    <span class="n">AssocRow</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span><span class="s2">&quot;AssocRow&quot;</span><span class="p">,</span>
                          <span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;p&quot;</span><span class="p">,</span> <span class="s2">&quot;odds&quot;</span><span class="p">,</span> <span class="s2">&quot;plate&quot;</span><span class="p">])</span>

    <span class="n">snpNames</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">assocResults</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">fileName</span> <span class="ow">in</span> <span class="n">fileNames</span><span class="p">:</span>
        <span class="c1"># Getting the plate name</span>
        <span class="n">plateName</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">r&quot;/plate_bias\.(\S+)\.assoc\.fisher$&quot;</span><span class="p">,</span> <span class="n">fileName</span><span class="p">)</span>
        <span class="n">plateName</span> <span class="o">=</span> <span class="n">plateName</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">plateName</span> <span class="k">else</span> <span class="s2">&quot;unknown&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fileName</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
                <span class="n">headerIndex</span> <span class="o">=</span> <span class="bp">None</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inputFile</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">createRowFromPlinkSpacedOutput</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

                    <span class="k">if</span> <span class="n">headerIndex</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                        <span class="c1"># This is the header line</span>
                        <span class="n">headerIndex</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span>
                            <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">))</span>
                        <span class="p">])</span>
                        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span> <span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="s2">&quot;BP&quot;</span><span class="p">,</span> <span class="s2">&quot;P&quot;</span><span class="p">,</span> <span class="s2">&quot;OR&quot;</span><span class="p">):</span>
                            <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">headerIndex</span><span class="p">:</span>
                                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: missing column {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                    <span class="n">fileName</span><span class="p">,</span>
                                    <span class="n">name</span><span class="p">,</span>
                                <span class="p">)</span>
                                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">snpName</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">]]</span>
                        <span class="n">snpNames</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">snpName</span><span class="p">)</span>
                        <span class="n">assocResults</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">AssocRow</span><span class="p">(</span>
                            <span class="n">chrom</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;CHR&quot;</span><span class="p">]],</span>
                            <span class="n">pos</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;BP&quot;</span><span class="p">]],</span>
                            <span class="n">name</span><span class="o">=</span><span class="n">snpName</span><span class="p">,</span>
                            <span class="n">p</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;P&quot;</span><span class="p">]],</span>
                            <span class="n">odds</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="n">headerIndex</span><span class="p">[</span><span class="s2">&quot;OR&quot;</span><span class="p">]],</span>
                            <span class="n">plate</span><span class="o">=</span><span class="n">plateName</span><span class="p">,</span>
                        <span class="p">))</span>

        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(fileName)s</span><span class="s2">: no such file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># The output file</span>
    <span class="n">outputFileName</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.significant_SNPs.txt&quot;</span>
    <span class="n">outputFile</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">outputFile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputFileName</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(outputFileName)s</span><span class="s2">: can&#39;t write file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">snpNames</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">outputFile</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">snpNames</span><span class="p">)</span>

    <span class="c1"># Closing the file</span>
    <span class="n">outputFile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">assocResults</span></div>


<div class="viewcode-block" id="computeFrequencyOfSignificantSNPs"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.computeFrequencyOfSignificantSNPs">[docs]</a><span class="k">def</span> <span class="nf">computeFrequencyOfSignificantSNPs</span><span class="p">(</span><span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the frequency of the significant markers.</span>

<span class="sd">    :param options: the options.</span>

<span class="sd">    :type options: argparse.Namespace</span>

<span class="sd">    Extract a list of markers (significant after plate bias analysis) and</span>
<span class="sd">    computes their frequencies.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The plink command</span>
    <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">bfile</span><span class="p">,</span> <span class="s2">&quot;--extract&quot;</span><span class="p">,</span>
                    <span class="n">options</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.significant_SNPs.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;--freq&quot;</span><span class="p">,</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span>
                    <span class="n">options</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.significant_SNPs&quot;</span><span class="p">]</span>
    <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span>

    <span class="c1"># Reading the frequency file</span>
    <span class="n">maf</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;.significant_SNPs.frq&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="n">header</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">name</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span>
            <span class="nb">enumerate</span><span class="p">(</span><span class="n">createRowFromPlinkSpacedOutput</span><span class="p">(</span><span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()))</span>
        <span class="p">}</span>
        <span class="k">for</span> <span class="n">required_col</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="s2">&quot;MAF&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">required_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;{}: missing column {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">script_prefix</span> <span class="o">+</span> <span class="s2">&quot;.significant_SNPs.frq&quot;</span><span class="p">,</span>
                    <span class="n">required_col</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">createRowFromPlinkSpacedOutput</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
            <span class="n">maf</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="n">header</span><span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">]]]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">header</span><span class="p">[</span><span class="s2">&quot;MAF&quot;</span><span class="p">]]</span>

    <span class="k">return</span> <span class="n">maf</span></div>


<div class="viewcode-block" id="executePlateBiasAnalysis"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.executePlateBiasAnalysis">[docs]</a><span class="k">def</span> <span class="nf">executePlateBiasAnalysis</span><span class="p">(</span><span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Execute the plate bias analysis with Plink.</span>

<span class="sd">    :param options: the options.</span>

<span class="sd">    :type options: argparse.Namespace</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The plink command</span>
    <span class="n">plinkCommand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span> <span class="s2">&quot;--noweb&quot;</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">bfile</span><span class="p">,</span>
                    <span class="s2">&quot;--loop-assoc&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">loop_assoc</span><span class="p">,</span> <span class="s2">&quot;--fisher&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;--pfilter&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">pfilter</span><span class="p">),</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">out</span><span class="p">]</span>
    <span class="n">runCommand</span><span class="p">(</span><span class="n">plinkCommand</span><span class="p">)</span></div>


<div class="viewcode-block" id="runCommand"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.runCommand">[docs]</a><span class="k">def</span> <span class="nf">runCommand</span><span class="p">(</span><span class="n">command</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Run a command.</span>

<span class="sd">    :param command: the command to run.</span>

<span class="sd">    :type command: list</span>

<span class="sd">    Tries to run a command. If it fails, raise a :py:class:`ProgramError`. This</span>
<span class="sd">    function uses the :py:mod:`subprocess` module.</span>

<span class="sd">    .. warning::</span>
<span class="sd">        The variable ``command`` should be a list of strings (no other type).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">command</span><span class="p">,</span>
                                         <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;couldn&#39;t run command</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="checkArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.checkArgs">[docs]</a><span class="k">def</span> <span class="nf">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    :param args: an object containing the options of the program.</span>

<span class="sd">    :type args: argparse.Namespace</span>

<span class="sd">    :returns: ``True`` if everything was OK.</span>

<span class="sd">    If there is a problem with an option, an exception is raised using the</span>
<span class="sd">    :py:class:`ProgramError` class, a message is printed to the</span>
<span class="sd">    :class:`sys.stderr` and the program exists with code 1.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check if we have the tped and the tfam files</span>
    <span class="k">for</span> <span class="n">fileName</span> <span class="ow">in</span> <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">bfile</span> <span class="o">+</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;.bim&quot;</span><span class="p">,</span> <span class="s2">&quot;.fam&quot;</span><span class="p">]]:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fileName</span><span class="p">):</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%(fileName)s</span><span class="s2">: no such file&quot;</span> <span class="o">%</span> <span class="nb">locals</span><span class="p">()</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c1"># Check the plate bias file</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">loop_assoc</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">: no such file&quot;</span> <span class="o">%</span> <span class="n">args</span><span class="o">.</span><span class="n">loop_assoc</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">return</span> <span class="bp">True</span></div>


<div class="viewcode-block" id="parseArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.parseArgs">[docs]</a><span class="k">def</span> <span class="nf">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>  <span class="c1"># pragma: no cover</span>
    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    :param argString: the options.</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    :returns: A :py:class:`argparse.Namespace` object created by the</span>
<span class="sd">              :py:mod:`argparse` module. It contains the values of the</span>
<span class="sd">              different options.</span>

<span class="sd">    ================ ====== ==================================================</span>
<span class="sd">        Options       Type                    Description</span>
<span class="sd">    ================ ====== ==================================================</span>
<span class="sd">    ``--bfile``      string The input file prefix (Plink binary).</span>
<span class="sd">    ``--loop-assoc`` string The file containing the plate organization of each</span>
<span class="sd">                            samples.</span>
<span class="sd">    ``--pfilter``    float  The significance threshold used for the plate</span>
<span class="sd">                            effect.</span>
<span class="sd">    ``--out``        string The prefix of the output files.</span>
<span class="sd">    ================ ====== ==================================================</span>

<span class="sd">    .. note::</span>
<span class="sd">        No option check is done here (except for the one automatically done by</span>
<span class="sd">        argparse). Those need to be done elsewhere (see :py:func:`checkArgs`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">argString</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">args</span></div>


<div class="viewcode-block" id="ProgramError"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.ProgramError">[docs]</a><span class="k">class</span> <span class="nc">ProgramError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;An :py:class:`Exception` raised in case of a problem.</span>

<span class="sd">    :param msg: the message to print to the user before exiting.</span>

<span class="sd">    :type msg: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Construction of the :py:class:`ProgramError` class.</span>

<span class="sd">        :param msg: the message to print to the user</span>

<span class="sd">        :type msg: str</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">message</span></div>


<span class="c1"># The parser object</span>
<span class="n">pretty_name</span> <span class="o">=</span> <span class="s2">&quot;Plate bias&quot;</span>
<span class="n">desc</span> <span class="o">=</span> <span class="s2">&quot;Checks for plate bias.&quot;</span>
<span class="n">long_desc</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;The script identifies markers with plate bias, based on the &quot;</span>
             <span class="s2">&quot;plates used to dilute DNA samples.&quot;</span><span class="p">)</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-v&quot;</span><span class="p">,</span> <span class="s2">&quot;--version&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;version&quot;</span><span class="p">,</span>
                    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;pyGenClean version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>

<span class="c1"># The INPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Input File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The input file prefix (will find the plink binary &quot;</span>
                         <span class="s2">&quot;files by appending the prefix to the .bim, .bed &quot;</span>
                         <span class="s2">&quot;and .fam files, respectively.&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--loop-assoc&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The file containing the plate organization of &quot;</span>
                         <span class="s2">&quot;each samples. Must contains three column (with no &quot;</span>
                         <span class="s2">&quot;header): famID, indID and plateName.&quot;</span><span class="p">))</span>
<span class="c1"># The options</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Options&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--pfilter&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FLOAT&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">1e-7</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The significance threshold used for the plate &quot;</span>
                         <span class="s2">&quot;effect. [default: </span><span class="si">%(default).1e</span><span class="s2">]&quot;</span><span class="p">))</span>
<span class="c1"># The OUTPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Output File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">default</span><span class="o">=</span><span class="s2">&quot;plate_bias&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;The prefix of the output files. [default: &quot;</span>
                         <span class="s2">&quot;</span><span class="si">%(default)s</span><span class="s2">]&quot;</span><span class="p">))</span>


<div class="viewcode-block" id="safe_main"><a class="viewcode-back" href="../../../module_content/pyGenClean.PlateBias.html#pyGenClean.PlateBias.plate_bias.safe_main">[docs]</a><span class="k">def</span> <span class="nf">safe_main</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;A safe version of the main function (that catches ProgramError).&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">main</span><span class="p">()</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span></div>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">safe_main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.8.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" >pyGenClean</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>