<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pyGenClean.NoCallHetero.heterozygosity_plot &mdash; pyGenClean 1.7.1 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.7.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="pyGenClean 1.7.1 documentation" href="../../../index.html" />
    <link rel="up" title="pyGenClean" href="../../pyGenClean.html" /> 
  </head>
  <body role="document">

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
<a href="../../../index.html"><img src="../../../_static/logo.png" border="0" alt="py4sci"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.7.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" accesskey="U">pyGenClean</a> &raquo;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for pyGenClean.NoCallHetero.heterozygosity_plot</h1><div class="highlight"><pre>
<span class="c">#!/usr/bin/env python2.7</span>

<span class="c"># This file is part of pyGenClean.</span>
<span class="c">#</span>
<span class="c"># pyGenClean is free software: you can redistribute it and/or modify it under</span>
<span class="c"># the terms of the GNU General Public License as published by the Free Software</span>
<span class="c"># Foundation, either version 3 of the License, or (at your option) any later</span>
<span class="c"># version.</span>
<span class="c">#</span>
<span class="c"># pyGenClean is distributed in the hope that it will be useful, but WITHOUT ANY</span>
<span class="c"># WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR</span>
<span class="c"># A PARTICULAR PURPOSE.  See the GNU General Public License for more details.</span>
<span class="c">#</span>
<span class="c"># You should have received a copy of the GNU General Public License along with</span>
<span class="c"># pyGenClean.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">__version__</span>


<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s">&quot;heterozygosity_plot&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;The main function of the module.</span>

<span class="sd">    :param argString: the options.</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    These are the steps:</span>

<span class="sd">    1. Prints the options.</span>
<span class="sd">    2. Checks the number of samples in the ``tfam`` file</span>
<span class="sd">       (:py:func:`compute_nb_samples`).</span>
<span class="sd">    3. Computes the heterozygosity rate (:py:func:`compute_heterozygosity`).</span>
<span class="sd">    4. Saves the heterozygosity data (in ``out.het``).</span>
<span class="sd">    5. Plots the heterozygosity rate (:py:func:`plot_heterozygosity`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Getting and checking the options</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>
    <span class="n">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="c"># Check the number of samples</span>
    <span class="n">nb_samples</span> <span class="o">=</span> <span class="n">compute_nb_samples</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">tfile</span><span class="p">)</span>

    <span class="c"># Compute the heterozygosity rate</span>
    <span class="n">heterozygosity</span><span class="p">,</span> <span class="n">samples</span> <span class="o">=</span> <span class="n">compute_heterozygosity</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">tfile</span><span class="p">,</span> <span class="n">nb_samples</span><span class="p">)</span>

    <span class="c"># Save heterozygosity data</span>
    <span class="n">save_heterozygosity</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

    <span class="c"># Plotting the heterozygosity rate distribution</span>
    <span class="n">plot_heterozygosity</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="save_heterozygosity"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.save_heterozygosity">[docs]</a><span class="k">def</span> <span class="nf">save_heterozygosity</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Saves the heterozygosity data.</span>

<span class="sd">    :param heterozygosity: the heterozygosity data.</span>
<span class="sd">    :param samples: the list of samples.</span>
<span class="sd">    :param out_prefix: the prefix of the output files.</span>

<span class="sd">    :type heterozygosity: numpy.array</span>
<span class="sd">    :type samples: list of tuples of str</span>
<span class="sd">    :type out_prefix: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The output file</span>
    <span class="n">o_file</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.het&quot;</span><span class="p">,</span> <span class="s">&#39;wb&#39;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;{}.het: can&#39;t write file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_prefix</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c"># Saving the data</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">fid</span><span class="p">,</span> <span class="n">iid</span><span class="p">),</span> <span class="n">het</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">heterozygosity</span><span class="p">):</span>
        <span class="k">print</span> <span class="o">&gt;&gt;</span><span class="n">o_file</span><span class="p">,</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">fid</span><span class="p">,</span> <span class="n">iid</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">het</span><span class="p">)])</span>

    <span class="c"># Closing the file</span>
    <span class="n">o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

</div>
<div class="viewcode-block" id="compute_nb_samples"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.compute_nb_samples">[docs]</a><span class="k">def</span> <span class="nf">compute_nb_samples</span><span class="p">(</span><span class="n">in_prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check the number of samples.</span>

<span class="sd">    :param in_prefix: the prefix of the input file.</span>

<span class="sd">    :type in_prefix: str</span>

<span class="sd">    :returns: the number of sample in ``prefix.fam``.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file_name</span> <span class="o">=</span> <span class="n">in_prefix</span> <span class="o">+</span> <span class="s">&quot;.tfam&quot;</span>
    <span class="n">nb</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="n">nb</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">input_file</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">nb</span>

</div>
<div class="viewcode-block" id="is_heterozygous"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.is_heterozygous">[docs]</a><span class="k">def</span> <span class="nf">is_heterozygous</span><span class="p">(</span><span class="n">genotype</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Tells if a genotype &quot;A B&quot; is heterozygous.</span>

<span class="sd">    :param genotype: the genotype to test for heterozygosity.</span>

<span class="sd">    :type genotype: str</span>

<span class="sd">    :returns: ``True`` if the genotype is heterozygous, ``False`` otherwise.</span>

<span class="sd">    The genotype must contain two alleles, separated by a space. It then</span>
<span class="sd">    compares the first allele (``genotype[0]``) with the last one</span>
<span class="sd">    (``genotype[-1]``).</span>

<span class="sd">    .. testsetup::</span>

<span class="sd">        from pyGenClean.NoCallHetero.heterozygosity_plot import is_heterozygous</span>

<span class="sd">    .. doctest::</span>

<span class="sd">        &gt;&gt;&gt; is_heterozygous(&quot;A A&quot;)</span>
<span class="sd">        False</span>
<span class="sd">        &gt;&gt;&gt; is_heterozygous(&quot;G C&quot;)</span>
<span class="sd">        True</span>
<span class="sd">        &gt;&gt;&gt; is_heterozygous(&quot;0 0&quot;) # No call is not heterozygous.</span>
<span class="sd">        False</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="ow">not</span> <span class="n">genotype</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">genotype</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

</div>
<div class="viewcode-block" id="compute_heterozygosity"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.compute_heterozygosity">[docs]</a><span class="k">def</span> <span class="nf">compute_heterozygosity</span><span class="p">(</span><span class="n">in_prefix</span><span class="p">,</span> <span class="n">nb_samples</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the heterozygosity ratio of samples (from tped).&quot;&quot;&quot;</span>
    <span class="n">tped_name</span> <span class="o">=</span> <span class="n">in_prefix</span> <span class="o">+</span> <span class="s">&quot;.tped&quot;</span>
    <span class="n">tfam_name</span> <span class="o">=</span> <span class="n">in_prefix</span> <span class="o">+</span> <span class="s">&quot;.tfam&quot;</span>

    <span class="c"># The function we want to use</span>
    <span class="n">check_heterozygosity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="n">is_heterozygous</span><span class="p">)</span>

    <span class="c"># The autosomes</span>
    <span class="n">autosomes</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)}</span>

    <span class="c"># The tfam</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tfam_name</span><span class="p">,</span> <span class="s">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="n">input_file</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\r\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)[:</span><span class="mi">2</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">]</span>

    <span class="n">heterozygosity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">nb_samples</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">nb_markers</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">nb_samples</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tped_name</span><span class="p">,</span> <span class="s">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
        <span class="c"># There is no header</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\r\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">))</span>

            <span class="n">chromosome</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">autosomes</span><span class="p">:</span>
                <span class="c"># This is not an autosome, so we skip</span>
                <span class="k">continue</span>

            <span class="c"># Getting the genotypes</span>
            <span class="n">genotypes</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>

            <span class="c"># Finding the heterozygous genotypes</span>
            <span class="n">heterozygosity</span> <span class="o">+=</span> <span class="n">check_heterozygosity</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)</span>

            <span class="c"># Adding to number of markers for each samples (excluding no calls)</span>
            <span class="n">nb_markers</span> <span class="o">+=</span> <span class="n">genotypes</span> <span class="o">!=</span> <span class="s">&quot;0 0&quot;</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">true_divide</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">,</span> <span class="n">nb_markers</span><span class="p">),</span> <span class="n">samples</span>

</div>
<div class="viewcode-block" id="plot_heterozygosity"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.plot_heterozygosity">[docs]</a><span class="k">def</span> <span class="nf">plot_heterozygosity</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Plots the heterozygosity rate distribution.</span>

<span class="sd">    :param heterozygosity: the heterozygosity data.</span>
<span class="sd">    :param options: the options.</span>

<span class="sd">    :type heterozygosity: numpy.array</span>
<span class="sd">    :type options: argparse.Namespace</span>

<span class="sd">    Plots a histogram or a boxplot of the heterozygosity distribution.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># importing important stuff</span>
    <span class="kn">import</span> <span class="nn">matplotlib</span> <span class="kn">as</span> <span class="nn">mpl</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span> <span class="o">!=</span> <span class="s">&quot;X11&quot;</span> <span class="ow">and</span> <span class="n">mpl</span><span class="o">.</span><span class="n">get_backend</span><span class="p">()</span> <span class="o">!=</span> <span class="s">&quot;agg&quot;</span><span class="p">:</span>
        <span class="n">mpl</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s">&quot;Agg&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span> <span class="o">!=</span> <span class="s">&quot;X11&quot;</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ioff</span><span class="p">()</span>

    <span class="c"># Creating the figure and ax</span>
    <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mf">13.5</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">boxplot</span><span class="p">:</span>
        <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mf">13.5</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>

    <span class="c"># Setting subplot properties</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s">&quot;bottom&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s">&quot;bottom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_position</span><span class="p">((</span><span class="s">&quot;outward&quot;</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s">&quot;left&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_position</span><span class="p">((</span><span class="s">&quot;outward&quot;</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>

    <span class="c"># The title and labels</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">((</span><span class="s">&quot;Heterozygosity rate distribution &quot;</span>
                  <span class="s">&quot;({:,} samples)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">))),</span> <span class="n">weight</span><span class="o">=</span><span class="s">&quot;bold&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&quot;Heterozygosity rate&quot;</span><span class="p">)</span>

    <span class="c"># Plotting the histogram</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">boxplot</span><span class="p">:</span>
        <span class="c"># Specific settings for the boxplot</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s">&quot;none&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s">&quot;left&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([])</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mf">0.18</span><span class="p">)</span>

        <span class="c"># The boxplot</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">,</span> <span class="n">notch</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">vert</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c"># Specific settings for the histogram</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&quot;Proportion&quot;</span><span class="p">)</span>

        <span class="c"># The histogram</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span>
            <span class="n">heterozygosity</span><span class="p">,</span>
            <span class="n">bins</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">bins</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="s">&quot;#0099CC&quot;</span><span class="p">,</span>
            <span class="n">histtype</span><span class="o">=</span><span class="s">&quot;stepfilled&quot;</span><span class="p">,</span>
            <span class="n">weights</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">)</span> <span class="o">+</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="c"># Plotting the mean, the median and the variance</span>
        <span class="n">the_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">)</span>
        <span class="n">the_median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">)</span>
        <span class="n">the_variance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">heterozygosity</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">mean_line</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">the_mean</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;#CC0000&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s">&quot;--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                               <span class="n">clip_on</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
        <span class="n">median_line</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">the_median</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;#FF8800&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s">&quot;--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                                 <span class="n">clip_on</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
        <span class="n">variance_line</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">the_variance</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;#FFFFFF&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s">&quot;--&quot;</span><span class="p">,</span>
                                   <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">clip_on</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span>
            <span class="p">[</span><span class="n">mean_line</span><span class="p">,</span> <span class="n">median_line</span><span class="p">,</span> <span class="n">variance_line</span><span class="p">],</span>
            <span class="p">[</span>
                <span class="s">&quot;Mean ({:.4})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">the_mean</span><span class="p">),</span>
                <span class="s">&quot;Median ({:.4})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">the_median</span><span class="p">),</span>
                <span class="s">&quot;Variance ({:.4})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">the_variance</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="n">loc</span><span class="o">=</span><span class="s">&quot;best&quot;</span><span class="p">,</span>
            <span class="n">prop</span><span class="o">=</span><span class="p">{</span><span class="s">&quot;size&quot;</span><span class="p">:</span> <span class="mi">11</span><span class="p">},</span>
        <span class="p">)</span>

        <span class="c"># The ylim</span>
        <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">ymax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">ymax</span><span class="p">)</span>

    <span class="c"># The xlim</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">xlim</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">xlim</span><span class="p">)</span>

    <span class="c"># Saving the figure</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span> <span class="o">==</span> <span class="s">&quot;X11&quot;</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="s">&quot;{}.{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">boxplot</span><span class="p">:</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="s">&quot;{}_boxplot.{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">format</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="checkArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.checkArgs">[docs]</a><span class="k">def</span> <span class="nf">checkArgs</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    :param args: an object containing the options of the program.</span>

<span class="sd">    :type args: argparse.Namespace</span>

<span class="sd">    :returns: ``True`` if everything was OK.</span>

<span class="sd">    If there is a problem with an option, an exception is raised using the</span>
<span class="sd">    :py:class:`ProgramError` class, a message is printed to the</span>
<span class="sd">    :class:`sys.stderr` and the program exists with code 1.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Checking the input file</span>
    <span class="k">for</span> <span class="n">file_name</span> <span class="ow">in</span> <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">tfile</span> <span class="o">+</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">{</span><span class="s">&quot;.tfam&quot;</span><span class="p">,</span> <span class="s">&quot;.tped&quot;</span><span class="p">}]:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c"># Checking the xlimit</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">xlim</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">xlim</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">args</span><span class="o">.</span><span class="n">xlim</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;invalid xlim&quot;</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="c"># Checking the ymax</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">ymax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">ymax</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s">&quot;invalid ymax (must be above 0)&quot;</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">return</span> <span class="bp">True</span>

</div>
<div class="viewcode-block" id="parseArgs"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.parseArgs">[docs]</a><span class="k">def</span> <span class="nf">parseArgs</span><span class="p">(</span><span class="n">argString</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>  <span class="c"># pragma: no cover</span>
    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    :param argString: the options.</span>

<span class="sd">    :type argString: list</span>

<span class="sd">    :returns: A :py:class:`argparse.Namespace` object created by the</span>
<span class="sd">              :py:mod:`argparse` module. It contains the values of the</span>
<span class="sd">              different options.</span>

<span class="sd">    ============= ====== ======================================</span>
<span class="sd">       Options     Type              Description</span>
<span class="sd">    ============= ====== ======================================</span>
<span class="sd">    ``--tfile``   string The prefix of the transposed file.</span>
<span class="sd">    ``--boxplot`` bool   Draw a boxplot instead of a histogram.</span>
<span class="sd">    ``--format``  string The output file format.</span>
<span class="sd">    ``--bins``    int    The number of bins for the histogram.</span>
<span class="sd">    ``--xlim``    float  The limit of the x axis.</span>
<span class="sd">    ``--ymax``    float  &quot;The maximal Y value.</span>
<span class="sd">    ``--out``     string The prefix of the output files.</span>
<span class="sd">    ============= ====== ======================================</span>

<span class="sd">    .. note::</span>
<span class="sd">        No option check is done here (except for the one automatically done by</span>
<span class="sd">        argparse). Those need to be done elsewhere (see :py:func:`checkArgs`).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">argString</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">argString</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">args</span>

</div>
<div class="viewcode-block" id="ProgramError"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.ProgramError">[docs]</a><span class="k">class</span> <span class="nc">ProgramError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;An :py:class:`Exception` raised in case of a problem.</span>

<span class="sd">    :param msg: the message to print to the user before exiting.</span>

<span class="sd">    :type msg: str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Construction of the :py:class:`ProgramError` class.</span>

<span class="sd">        :param msg: the message to print to the user</span>

<span class="sd">        :type msg: str</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">message</span>


<span class="c"># The parser object</span></div>
<span class="n">desc</span> <span class="o">=</span> <span class="s">&quot;Plots the distribution of the heterozygosity ratio.&quot;</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;-v&quot;</span><span class="p">,</span> <span class="s">&quot;--version&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&quot;version&quot;</span><span class="p">,</span>
                    <span class="n">version</span><span class="o">=</span><span class="s">&quot;pyGenClean version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>

<span class="c"># The INPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Input File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--tfile&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="s">&quot;The prefix of the transposed file&quot;</span><span class="p">)</span>
<span class="c"># The options</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Options&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--boxplot&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;Draw a boxplot instead of a histogram.&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--format&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FORMAT&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;png&quot;</span><span class="p">,</span>
                   <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;png&quot;</span><span class="p">,</span> <span class="s">&quot;ps&quot;</span><span class="p">,</span> <span class="s">&quot;pdf&quot;</span><span class="p">,</span> <span class="s">&quot;X11&quot;</span><span class="p">],</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;The output file format (png, ps, pdf, or X11 &quot;</span>
                         <span class="s">&quot;formats are available). [default: </span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--bins&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;INT&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;The number of bins for the histogram. [default: &quot;</span>
                         <span class="s">&quot;</span><span class="si">%(default)d</span><span class="s">]&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--xlim&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;The limit of the x axis (floats).&quot;</span><span class="p">))</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--ymax&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;The maximal Y value.&quot;</span><span class="p">))</span>
<span class="c"># The OUTPUT files</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Output File&quot;</span><span class="p">)</span>
<span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--out&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
                   <span class="n">default</span><span class="o">=</span><span class="s">&quot;heterozygosity&quot;</span><span class="p">,</span>
                   <span class="n">help</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;The prefix of the output files. [default: &quot;</span>
                         <span class="s">&quot;</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">))</span>


<div class="viewcode-block" id="safe_main"><a class="viewcode-back" href="../../../module_content/pyGenClean.NoCallHetero.html#pyGenClean.NoCallHetero.heterozygosity_plot.safe_main">[docs]</a><span class="k">def</span> <span class="nf">safe_main</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;A safe version of the main function (that catches ProgramError).&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">main</span><span class="p">()</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>

</div>
<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">safe_main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pyGenClean 1.7.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../pyGenClean.html" >pyGenClean</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Louis-Philippe Lemieux Perreault.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>